<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getMetaCycModelForOrganism</title>
  <meta name="keywords" content="getMetaCycModelForOrganism">
  <meta name="description" content="getMetaCycModelForOrganism">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">metacyc</a> &gt; getMetaCycModelForOrganism.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\metacyc&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getMetaCycModelForOrganism
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getMetaCycModelForOrganism</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function model=getMetaCycModelForOrganism(organismID,fastaFile,keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getMetaCycModelForOrganism
   Reconstructs a genome-scale metabolic model based on protein homology to the
   MetaCyc pathway database

   organismID          abbreviation of the query organism 
   fastaFile           a FASTA file that contains the protein sequences of
                       the organism for which to reconstruct a model
   keepTransportRxns   include transportation reactions, which often have identical
                       reactants and products that turn to be all-zero columns in
                       the S matrix (opt, default false)
   keepUnbalanced      include reactions cannot be unbalanced reactions, usually
                       because they are polymeric reactions or because of a
                       specific difficulty in balancing class structures
                       (opt, default false)
   keepUndetermined    include reactions that have substrates lack chemical
                       structures or with non-numerical coefficients (e.g. n+1)
                       (opt, default false)
   minScore            minimum Bit scores of BLASTp search (opt, default 100)
   minPositives        minimum Positives values of BLASTp search (opt, default 45 %)

   model               a model structure for the query organism

   Usage: model=getMetaCycModelForOrganism(organismID,fastaFile,...
    keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives)

   Eduard Kerkhoven, 2017-12-12</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getMetsFromMetaCyc.html" class="code" title="function metaCycMets=getMetsFromMetaCyc(metacycPath)">getMetsFromMetaCyc</a>	getMetsFromMetaCyc</li><li><a href="getModelFromMetaCyc.html" class="code" title="function metaCycModel=getModelFromMetaCyc(metacycPath,keepTransportRxns,keepUnbalanced,keepUndetermined)">getModelFromMetaCyc</a>	getModelFromMetaCyc</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=getMetaCycModelForOrganism(organismID,fastaFile,</a><span class="keyword">...</span>
0002     keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives)
0003 <span class="comment">% getMetaCycModelForOrganism</span>
0004 <span class="comment">%   Reconstructs a genome-scale metabolic model based on protein homology to the</span>
0005 <span class="comment">%   MetaCyc pathway database</span>
0006 <span class="comment">%</span>
0007 <span class="comment">%   organismID          abbreviation of the query organism</span>
0008 <span class="comment">%   fastaFile           a FASTA file that contains the protein sequences of</span>
0009 <span class="comment">%                       the organism for which to reconstruct a model</span>
0010 <span class="comment">%   keepTransportRxns   include transportation reactions, which often have identical</span>
0011 <span class="comment">%                       reactants and products that turn to be all-zero columns in</span>
0012 <span class="comment">%                       the S matrix (opt, default false)</span>
0013 <span class="comment">%   keepUnbalanced      include reactions cannot be unbalanced reactions, usually</span>
0014 <span class="comment">%                       because they are polymeric reactions or because of a</span>
0015 <span class="comment">%                       specific difficulty in balancing class structures</span>
0016 <span class="comment">%                       (opt, default false)</span>
0017 <span class="comment">%   keepUndetermined    include reactions that have substrates lack chemical</span>
0018 <span class="comment">%                       structures or with non-numerical coefficients (e.g. n+1)</span>
0019 <span class="comment">%                       (opt, default false)</span>
0020 <span class="comment">%   minScore            minimum Bit scores of BLASTp search (opt, default 100)</span>
0021 <span class="comment">%   minPositives        minimum Positives values of BLASTp search (opt, default 45 %)</span>
0022 <span class="comment">%</span>
0023 <span class="comment">%   model               a model structure for the query organism</span>
0024 <span class="comment">%</span>
0025 <span class="comment">%   Usage: model=getMetaCycModelForOrganism(organismID,fastaFile,...</span>
0026 <span class="comment">%    keepTransportRxns,keepUnbalanced,keepUndetermined,minScore,minPositives)</span>
0027 <span class="comment">%</span>
0028 <span class="comment">%   Eduard Kerkhoven, 2017-12-12</span>
0029 <span class="comment">%</span>
0030 
0031 <span class="keyword">if</span> nargin&lt;2
0032     EM=<span class="string">'No query protein fasta file is specified'</span>;
0033     dispEM(EM);
0034 <span class="keyword">end</span>
0035 <span class="keyword">if</span> nargin&lt;3
0036     keepTransportRxns=false;
0037 <span class="keyword">end</span>
0038 <span class="keyword">if</span> nargin&lt;4
0039     keepUnbalanced=false;
0040 <span class="keyword">end</span>
0041 <span class="keyword">if</span> nargin&lt;5
0042     keepUndetermined=false;
0043 <span class="keyword">end</span>
0044 <span class="keyword">if</span> nargin&lt;6
0045     minScore=100;
0046 <span class="keyword">end</span>
0047 <span class="keyword">if</span> nargin&lt;7
0048     minPositives=45;
0049 <span class="keyword">end</span>
0050 
0051 <span class="comment">%Check if query fasta exists</span>
0052 <span class="keyword">if</span> ~exist(fastaFile,<span class="string">'file'</span>)
0053     EM=<span class="string">'Cannot find the query protein fasta file'</span>;
0054     dispEM(EM,true);
0055 <span class="keyword">end</span>
0056 
0057 <span class="comment">%First generate the full MetaCyc model</span>
0058 metaCycModel=<a href="getModelFromMetaCyc.html" class="code" title="function metaCycModel=getModelFromMetaCyc(metacycPath,keepTransportRxns,keepUnbalanced,keepUndetermined)">getModelFromMetaCyc</a>([],keepTransportRxns,keepUnbalanced,keepUndetermined);
0059 fprintf(<span class="string">'The full MetaCyc model loaded\n'</span>);
0060 
0061 <span class="comment">%Create the draft model from MetaCyc super model</span>
0062 <span class="comment">%model=metaCycModel;</span>
0063 model.id=organismID;
0064 model.description=<span class="string">'Generated by homology with MetaCyc database'</span>;
0065 model.rxns=metaCycModel.rxns;
0066 model.rxnNames=metaCycModel.rxnNames;
0067 model.eccodes=metaCycModel.eccodes;
0068 model.subSystems=metaCycModel.subSystems;
0069 model.rxnMiriams=metaCycModel.rxnMiriams;
0070 model.rxnReferences=metaCycModel.rxnReferences;
0071 model.lb=metaCycModel.lb;
0072 model.ub=metaCycModel.ub;
0073 model.rev=metaCycModel.rev;
0074 model.c=metaCycModel.c;
0075 model.equations=metaCycModel.equations;
0076 
0077 <span class="comment">%Get the 'external' directory for RAVEN Toolbox.</span>
0078 [ST I]=dbstack(<span class="string">'-completenames'</span>);
0079 ravenPath=fileparts(fileparts(ST(I).file));
0080 
0081 <span class="comment">%Generate blast strcture (ensure location of MetaCyc functions are is $RAVEN/external/metacyc)</span>
0082 blastStruc=getBlast(organismID,fastaFile,{<span class="string">'MetaCyc'</span>},fullfile(ravenPath,<span class="string">'metacyc'</span>,<span class="string">'protseq.fsa'</span>));
0083 
0084 <span class="comment">%Only look the query</span>
0085 blastStructure=blastStruc(2);
0086 
0087 <span class="comment">%Remove all blast hits that are below the cutoffs</span>
0088 indexes=blastStructure.bitscore&gt;=minScore &amp; blastStructure.ppos&gt;=minPositives;
0089 blastStructure.toGenes(~indexes)=[];
0090 blastStructure.fromGenes(~indexes)=[];
0091 blastStructure.evalue(~indexes)=[];
0092 blastStructure.aligLen(~indexes)=[];
0093 blastStructure.identity(~indexes)=[];
0094 blastStructure.bitscore(~indexes)=[];
0095 blastStructure.ppos(~indexes)=[];
0096 fprintf(<span class="string">'Completed searching against MetaCyc protein sequences.\n'</span>);
0097 
0098 <span class="comment">% Get the qualified genes of query organism from blast structure</span>
0099 model.genes=cell(10000,1);
0100 model.proteins=cell(10000,1);
0101 model.bitscore=zeros(10000,1);
0102 model.ppos=zeros(10000,1);
0103 num=1;
0104 
0105 <span class="comment">%Go through the strucutre and find out the hit with the best bit score</span>
0106 <span class="keyword">for</span> i=1:numel(blastStructure.fromGenes)
0107     <span class="keyword">if</span> num==1
0108         model.genes(num)=blastStructure.fromGenes(i);
0109         model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0110         model.bitscore(num)=blastStructure.bitscore(i);
0111         model.ppos(num)=blastStructure.ppos(i);
0112         num=num+1;
0113         lastGene=blastStructure.fromGenes(i);
0114     <span class="keyword">else</span>
0115         <span class="keyword">if</span> ~isequal(lastGene,blastStructure.fromGenes(i))
0116             model.genes(num)=blastStructure.fromGenes(i);
0117             model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0118             model.bitscore(num)=blastStructure.bitscore(i);
0119             model.ppos(num)=blastStructure.ppos(i);
0120             num=num+1;
0121             lastGene=blastStructure.fromGenes(i);
0122         <span class="keyword">else</span>
0123             <span class="keyword">if</span> model.bitscore(num)&lt;blastStructure.bitscore(i)
0124                 model.bitscore(num)=blastStructure.bitscore(i);
0125                 model.proteins(num)=strrep(blastStructure.toGenes(i), <span class="string">'gnl|META|'</span>, <span class="string">''</span>);
0126                 model.ppos(num)=blastStructure.ppos(i);
0127             <span class="keyword">end</span>
0128         <span class="keyword">end</span>    
0129     <span class="keyword">end</span>
0130 <span class="keyword">end</span>
0131 model.genes=model.genes(1:num-1);
0132 model.proteins=model.proteins(1:num-1);
0133 model.bitscore=model.bitscore(1:num-1);
0134 model.ppos=model.ppos(1:num-1);
0135 
0136 <span class="comment">% Get the indexes of matched genes in the metaCycModel</span>
0137 [crap, indexes]=ismember(model.proteins,metaCycModel.genes);
0138 
0139 <span class="comment">% Go throught the indexes to get the rxnGeneMat matrix</span>
0140 model.rxnGeneMat=sparse(numel(model.rxns),num-1);
0141 <span class="keyword">for</span> i=1:num-1
0142     model.rxnGeneMat(:,i)=metaCycModel.rxnGeneMat(:,indexes(i));
0143 <span class="keyword">end</span>
0144 
0145 <span class="comment">%Remove all reactions without genes</span>
0146 hasGenes=any(model.rxnGeneMat,2);
0147 model=removeReactions(model,~hasGenes,true);
0148 
0149 <span class="comment">%Generate grRules, only consider the or relationship here</span>
0150 <span class="comment">%Matched enzymes are stored in field rxnScores,</span>
0151 rxnNum=numel(model.rxns);
0152 model.rxnConfidenceScores=cell(rxnNum,1);
0153 model.rxnConfidenceScores(:)={<span class="string">'2'</span>};
0154 model.grRules=cell(rxnNum,1);
0155 <span class="comment">%model.rxnScores=cell(rxnNum,1);</span>
0156 <span class="keyword">for</span> j=1:rxnNum
0157     model.grRules{j}=<span class="string">''</span>;
0158     I=find(model.rxnGeneMat(j,:));
0159     <span class="keyword">for</span> k=1:numel(I)
0160         <span class="keyword">if</span> isempty(model.grRules{j})
0161             model.grRules(j)=model.genes(I(k));
0162             <span class="comment">%model.rxnScores(j)=model.proteins(I(k));</span>
0163         <span class="keyword">else</span>
0164             model.grRules(j)=strcat(model.grRules(j),{<span class="string">' or '</span>},model.genes(I(k)));
0165             <span class="comment">%model.rxnScores(j)=strcat(model.rxnScores(j),{' or '},model.proteins(I(k)));</span>
0166         <span class="keyword">end</span>
0167     <span class="keyword">end</span>
0168 <span class="keyword">end</span>
0169 
0170 <span class="comment">%Construct the S matrix and list of metabolites</span>
0171 [S, mets, badRxns]=constructS(model.equations);
0172 model.S=S;
0173 model.mets=mets;
0174 
0175 <span class="comment">%model=removeReactions(model,badRxns,true,true);</span>
0176 
0177 <span class="comment">%Then match up metabolites</span>
0178 metaCycMets=<a href="getMetsFromMetaCyc.html" class="code" title="function metaCycMets=getMetsFromMetaCyc(metacycPath)">getMetsFromMetaCyc</a>([]);
0179 
0180 <span class="comment">%Add information about all metabolites to the model</span>
0181 [a, b]=ismember(model.mets,metaCycMets.mets);
0182 a=find(a);
0183 b=b(a);
0184 
0185 <span class="keyword">if</span> ~isfield(model,<span class="string">'metNames'</span>)
0186    model.metNames=cell(numel(model.mets),1);
0187    model.metNames(:)={<span class="string">''</span>};
0188 <span class="keyword">end</span>
0189 model.metNames(a)=metaCycMets.metNames(b);
0190 
0191 <span class="keyword">if</span> ~isfield(model,<span class="string">'metFormulas'</span>)
0192    model.metFormulas=cell(numel(model.mets),1);
0193    model.metFormulas(:)={<span class="string">''</span>};
0194 <span class="keyword">end</span>
0195 model.metFormulas(a)=metaCycMets.metFormulas(b);
0196 
0197 <span class="keyword">if</span> ~isfield(model,<span class="string">'metCharges'</span>)
0198    model.metCharges=zeros(numel(model.mets),1);
0199 <span class="keyword">end</span>
0200 model.metCharges(a)=metaCycMets.metCharges(b);
0201 
0202 <span class="keyword">if</span> ~isfield(model,<span class="string">'b'</span>)
0203    model.b=zeros(numel(model.mets),1);
0204 <span class="keyword">end</span>
0205 <span class="comment">%model.b(a)=metaCycMets.b(b);</span>
0206 
0207 <span class="keyword">if</span> ~isfield(model,<span class="string">'inchis'</span>)
0208    model.inchis=cell(numel(model.mets),1);
0209    model.inchis(:)={<span class="string">''</span>};
0210 <span class="keyword">end</span>
0211 model.inchis(a)=metaCycMets.inchis(b);
0212 
0213 <span class="keyword">if</span> ~isfield(model,<span class="string">'metMiriams'</span>)
0214    model.metMiriams=cell(numel(model.mets),1);
0215 <span class="keyword">end</span>
0216 model.metMiriams(a)=metaCycMets.metMiriams(b);
0217 
0218 <span class="comment">%Put all metabolites in one compartment called 's' (for system). This is</span>
0219 <span class="comment">%done just to be more compatible with the rest of the code</span>
0220 model.comps={<span class="string">'s'</span>};
0221 model.compNames={<span class="string">'System'</span>};
0222 model.metComps=ones(numel(model.mets),1);
0223 
0224 <span class="comment">%It could also be that the metabolite names are empty for some reason</span>
0225 <span class="comment">%In that case, use the ID instead</span>
0226 I=cellfun(@isempty,model.metNames);
0227 model.metNames(I)=model.mets(I);
0228 
0229 <span class="comment">%Remove additional fields</span>
0230 model=rmfield(model,{<span class="string">'proteins'</span>,<span class="string">'bitscore'</span>,<span class="string">'ppos'</span>});
0231 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 14-Mar-2018 21:08:28 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>