<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getBlastFromExcel</title>
  <meta name="keywords" content="getBlastFromExcel">
  <meta name="description" content="getBlastFromExcel">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getBlastFromExcel.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getBlastFromExcel
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getBlastFromExcel</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function blastStructure=getBlastFromExcel(models,blastFile,organismId) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getBlastFromExcel
   Retrieves gene homology information from Excel files. Used as
   input to getModelFromHomology.

   Input:
   models          a cell array of model structures
   blastFile       Excel file with homology information
   organismId      the id of the organism of interest (as described in the
                   Excel file)

   Output:
   blastStructure  structure containing the information in the Excel
                   sheets.

   The Excel file should contain a number of spreadsheets which in turn
   contain the bidirectional homology measurements between the genes in the
   organisms. The first and second column headers in each sheet is the
   &quot;to&quot; and &quot;from&quot; model ids (as defined in models or for the new organism).
   The entries should correspond to the gene names in those models. The third,
   fourth, fifth, sixth and seventh columns represent the E-value, alignment
   length, identity, bitscore and percentage of positive-scoring matches for
   each measurement (captions should be &quot;E-value&quot;, &quot;Alignment length&quot;,
   &quot;Identity&quot;, &quot;Bitscore&quot; and &quot;PPOS&quot;).

   Usage: blastStructure=getBlastFromExcel(models,blastFile,organismId)

   Eduard Kerkhoven, 2018-12-10</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function blastStructure=getBlastFromExcel(models,blastFile,organismId)</a>
0002 <span class="comment">% getBlastFromExcel</span>
0003 <span class="comment">%   Retrieves gene homology information from Excel files. Used as</span>
0004 <span class="comment">%   input to getModelFromHomology.</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   Input:</span>
0007 <span class="comment">%   models          a cell array of model structures</span>
0008 <span class="comment">%   blastFile       Excel file with homology information</span>
0009 <span class="comment">%   organismId      the id of the organism of interest (as described in the</span>
0010 <span class="comment">%                   Excel file)</span>
0011 <span class="comment">%</span>
0012 <span class="comment">%   Output:</span>
0013 <span class="comment">%   blastStructure  structure containing the information in the Excel</span>
0014 <span class="comment">%                   sheets.</span>
0015 <span class="comment">%</span>
0016 <span class="comment">%   The Excel file should contain a number of spreadsheets which in turn</span>
0017 <span class="comment">%   contain the bidirectional homology measurements between the genes in the</span>
0018 <span class="comment">%   organisms. The first and second column headers in each sheet is the</span>
0019 <span class="comment">%   &quot;to&quot; and &quot;from&quot; model ids (as defined in models or for the new organism).</span>
0020 <span class="comment">%   The entries should correspond to the gene names in those models. The third,</span>
0021 <span class="comment">%   fourth, fifth, sixth and seventh columns represent the E-value, alignment</span>
0022 <span class="comment">%   length, identity, bitscore and percentage of positive-scoring matches for</span>
0023 <span class="comment">%   each measurement (captions should be &quot;E-value&quot;, &quot;Alignment length&quot;,</span>
0024 <span class="comment">%   &quot;Identity&quot;, &quot;Bitscore&quot; and &quot;PPOS&quot;).</span>
0025 <span class="comment">%</span>
0026 <span class="comment">%   Usage: blastStructure=getBlastFromExcel(models,blastFile,organismId)</span>
0027 <span class="comment">%</span>
0028 <span class="comment">%   Eduard Kerkhoven, 2018-12-10</span>
0029 <span class="comment">%</span>
0030 
0031 <span class="keyword">if</span> ~(exist(blastFile,<span class="string">'file'</span>)==2)
0032     error(<span class="string">'BLAST result file %s cannot be found'</span>,string(blastFile));
0033 <span class="keyword">end</span>
0034 
0035 blastStructure=[];
0036 
0037 <span class="comment">%Get a list of model IDs</span>
0038 organisms=cell(numel(models)+1,1);
0039 organisms{1}=organismId;
0040 <span class="keyword">for</span> i=1:numel(models)
0041     organisms{i+1}=models{i}.id;
0042 <span class="keyword">end</span>
0043 
0044 <span class="comment">%Get all the spreadsheets in the file</span>
0045 [type, sheets]=xlsfinfo(blastFile);
0046 
0047 <span class="comment">%Check if the file is a Microsoft Excel Spreadsheet</span>
0048 <span class="keyword">if</span> ~any(regexp(type,<span class="string">'Excel Spreadsheet'</span>))
0049     EM=<span class="string">'The file is not a Microsoft Excel Spreadsheet'</span>;
0050     dispEM(EM);
0051 <span class="keyword">end</span>
0052 
0053 <span class="keyword">for</span> i=1:numel(sheets)
0054     <span class="comment">%Check if the sheet has the right header and deal with organisms that</span>
0055     <span class="comment">%are in &quot;models&quot;</span>
0056     [values,dataSheet]=xlsread(blastFile,i);
0057     labels=dataSheet(1,:);
0058     <span class="keyword">if</span> strcmpi(labels{3},<span class="string">'E-value'</span>) &amp;&amp; strcmpi(labels{4},<span class="string">'Alignment length'</span>) <span class="keyword">...</span>
0059             &amp;&amp; strcmpi(labels{5},<span class="string">'Identity'</span>) &amp;&amp; strcmpi(labels{6},<span class="string">'Bitscore'</span>) <span class="keyword">...</span>
0060             &amp;&amp; strcmpi(labels{7},<span class="string">'PPOS'</span>)
0061         <span class="comment">%At least one of the organisms must have a model</span>
0062         fromID=find(strcmpi(labels{1},organisms));
0063         toID=find(strcmpi(labels{2},organisms));
0064         <span class="comment">%Check that the organism ids exist and that one of them is the</span>
0065         <span class="comment">%organism of interest</span>
0066         <span class="keyword">if</span> any(fromID) &amp;&amp; any(toID) &amp;&amp; (toID==1 || fromID==1)
0067             <span class="comment">%Check that no gene ids are empty. This could for example be</span>
0068             <span class="comment">%the case if the gene names are wrongly formatted as numbers</span>
0069             <span class="comment">%instead of strings</span>
0070             emptyNames=cellfun(@isempty,dataSheet(2:<span class="keyword">end</span>,1)) | cellfun(@isempty,dataSheet(2:<span class="keyword">end</span>,2));
0071             <span class="keyword">if</span> any(emptyNames)
0072                 <span class="keyword">if</span> all(emptyNames)
0073                     EM=[<span class="string">'Only empty gene names in sheet from '</span> organisms{fromID} <span class="string">' to '</span> organisms{toID}];
0074                     dispEM(EM);
0075                 <span class="keyword">else</span>
0076                     EM=[<span class="string">'Empty gene names in sheet from '</span> organisms{fromID} <span class="string">' to '</span> organisms{toID} <span class="string">'. Ignoring genes with empty names'</span>];
0077                     dispEM(EM,false);
0078                 <span class="keyword">end</span>
0079             <span class="keyword">end</span>
0080             blastStructure(numel(blastStructure)+1).toId=organisms{toID};
0081             blastStructure(numel(blastStructure)).fromId=organisms{fromID};
0082             blastStructure(numel(blastStructure)).fromGenes=dataSheet(2:<span class="keyword">end</span>,1);
0083             blastStructure(numel(blastStructure)).toGenes=dataSheet(2:<span class="keyword">end</span>,2);
0084             blastStructure(numel(blastStructure)).evalue=values(:,1);
0085             blastStructure(numel(blastStructure)).aligLen=values(:,2);
0086             blastStructure(numel(blastStructure)).identity=values(:,3);
0087             blastStructure(numel(blastStructure)).bitscore=values(:,4);
0088             blastStructure(numel(blastStructure)).ppos=values(:,5);
0089             
0090             <span class="comment">%Remove matches where any of the values is NaN. This would have</span>
0091             <span class="comment">%been done anyways in getModelFromHomology, but it's neater to</span>
0092             <span class="comment">%do it here</span>
0093             I=isnan(blastStructure(end).evalue) | isnan(blastStructure(end).aligLen) | isnan(blastStructure(end).identity);
0094             blastStructure(end).fromGenes(I)=[];
0095             blastStructure(end).toGenes(I)=[];
0096             blastStructure(end).evalue(I)=[];
0097             blastStructure(end).aligLen(I)=[];
0098             blastStructure(end).identity(I)=[];
0099             blastStructure(end).bitscore(I)=[];
0100             blastStructure(end).ppos(I)=[];
0101         <span class="keyword">else</span>
0102             <span class="keyword">if</span> isempty(toID) || isempty(fromID)
0103                 EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' has no corresponding model. Ignoring sheet'</span>];
0104                 dispEM(EM,false);
0105             <span class="keyword">else</span>
0106                 EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' does not involve the organism of interest. Ignoring sheet'</span>];
0107                 dispEM(EM,false);
0108             <span class="keyword">end</span>
0109         <span class="keyword">end</span>
0110     <span class="keyword">else</span>
0111         EM=[<span class="string">'The data in sheet '</span> sheets{i} <span class="string">' is not correctly formatted. Ignoring sheet'</span>];
0112         dispEM(EM,false);
0113     <span class="keyword">end</span>
0114 <span class="keyword">end</span>
0115 
0116 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 06-Nov-2019 21:22:42 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>