<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getBlast</title>
  <meta name="keywords" content="getBlast">
  <meta name="description" content="getBlast">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getBlast.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getBlast
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getBlast</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function [blastStructure,blastReport]=getBlast(organismID,fastaFile,modelIDs,refFastaFiles,developMode,hideVerbose) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getBlast
   Performs a bidirectional BLAST between the organism of interest and a
   set of template organisms

   Input:
   organismID      the id of the organism of interest. This should also
                   match with the id supplied to getModelFromHomology
   fastaFile       a FASTA file with the protein sequences for the
                   organism of interest
   modelIDs        a cell array of model ids. These must match the
                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the
                   output is to be used with getModelFromHomology
   refFastaFiles   a cell array with the paths to the corresponding FASTA
                   files
   developMode     true if blastReport should be generated that is used
                   in the unit testing function for BLAST+ (opt, default
                   false)
   hideVerbose     true if no status messages should be printed (opt,
                   default false)

   Output:
   blastStructure  structure containing the bidirectional homology
                   measurements that can be used by getModelFromHomology
   blastReport     structure containing MD5 hashes for FASTA database
                   files and non-parsed BLAST output data. Will be blank
                   if developMode is false.

   NOTE: This function calls BLAST+ to perform a bidirectional homology
   test between the organism of interest and a set of other organisms
   using standard settings. The only filtering this function does is the
   removal of hits with an E-value higher than 10e-5. The other homology
   measurements can be implemented using getBlastFromExcel.

   Usage: [blastStructure,blastReport]=getBlast(organismID,fastaFile,...
    modelIDs,refFastaFiles,developMode,hideVerbose)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [blastStructure,blastReport]=getBlast(organismID,fastaFile,</a><span class="keyword">...</span>
0002     modelIDs,refFastaFiles,developMode,hideVerbose)
0003 <span class="comment">% getBlast</span>
0004 <span class="comment">%   Performs a bidirectional BLAST between the organism of interest and a</span>
0005 <span class="comment">%   set of template organisms</span>
0006 <span class="comment">%</span>
0007 <span class="comment">%   Input:</span>
0008 <span class="comment">%   organismID      the id of the organism of interest. This should also</span>
0009 <span class="comment">%                   match with the id supplied to getModelFromHomology</span>
0010 <span class="comment">%   fastaFile       a FASTA file with the protein sequences for the</span>
0011 <span class="comment">%                   organism of interest</span>
0012 <span class="comment">%   modelIDs        a cell array of model ids. These must match the</span>
0013 <span class="comment">%                   &quot;model.id&quot; fields in the &quot;models&quot; structure if the</span>
0014 <span class="comment">%                   output is to be used with getModelFromHomology</span>
0015 <span class="comment">%   refFastaFiles   a cell array with the paths to the corresponding FASTA</span>
0016 <span class="comment">%                   files</span>
0017 <span class="comment">%   developMode     true if blastReport should be generated that is used</span>
0018 <span class="comment">%                   in the unit testing function for BLAST+ (opt, default</span>
0019 <span class="comment">%                   false)</span>
0020 <span class="comment">%   hideVerbose     true if no status messages should be printed (opt,</span>
0021 <span class="comment">%                   default false)</span>
0022 <span class="comment">%</span>
0023 <span class="comment">%   Output:</span>
0024 <span class="comment">%   blastStructure  structure containing the bidirectional homology</span>
0025 <span class="comment">%                   measurements that can be used by getModelFromHomology</span>
0026 <span class="comment">%   blastReport     structure containing MD5 hashes for FASTA database</span>
0027 <span class="comment">%                   files and non-parsed BLAST output data. Will be blank</span>
0028 <span class="comment">%                   if developMode is false.</span>
0029 <span class="comment">%</span>
0030 <span class="comment">%   NOTE: This function calls BLAST+ to perform a bidirectional homology</span>
0031 <span class="comment">%   test between the organism of interest and a set of other organisms</span>
0032 <span class="comment">%   using standard settings. The only filtering this function does is the</span>
0033 <span class="comment">%   removal of hits with an E-value higher than 10e-5. The other homology</span>
0034 <span class="comment">%   measurements can be implemented using getBlastFromExcel.</span>
0035 <span class="comment">%</span>
0036 <span class="comment">%   Usage: [blastStructure,blastReport]=getBlast(organismID,fastaFile,...</span>
0037 <span class="comment">%    modelIDs,refFastaFiles,developMode,hideVerbose)</span>
0038 
0039 <span class="keyword">if</span> nargin&lt;5
0040     developMode = false;
0041 <span class="keyword">end</span>
0042 <span class="keyword">if</span> nargin&lt;6
0043     hideVerbose = false;
0044 <span class="keyword">end</span>
0045 
0046 <span class="comment">%Everything should be cell arrays</span>
0047 organismID=convertCharArray(organismID);
0048 fastaFile=convertCharArray(fastaFile);
0049 modelIDs=convertCharArray(modelIDs);
0050 refFastaFiles=convertCharArray(refFastaFiles);
0051 
0052 <span class="comment">%Create blank structures for results</span>
0053 blastStructure=[];
0054 blastReport.dbHashes.phr={};
0055 blastReport.dbHashes.pot={};
0056 blastReport.dbHashes.psq={};
0057 blastReport.dbHashes.pto={};
0058 blastReport.blastTxtOutput={};
0059 
0060 <span class="comment">%Get the directory for RAVEN Toolbox</span>
0061 ravenPath=findRAVENroot();
0062 
0063 <span class="comment">%Generate temporary names for BLAST databases and output files</span>
0064 tmpDB=tempname;
0065 outFile=tempname;
0066 
0067 <span class="comment">%Check for existence of files. If no full path is specified for a file,</span>
0068 <span class="comment">%assume that it is in the current folder</span>
0069 <span class="keyword">if</span> isrow(refFastaFiles)
0070     files=horzcat(fastaFile,refFastaFiles);
0071 <span class="keyword">else</span>
0072     files=vertcat(fastaFile,refFastaFiles);
0073 <span class="keyword">end</span>
0074 
0075 files=checkFileExistence(files,2); <span class="comment">%Copy files to temp dir</span>
0076 fastaFile = files(1);
0077 refFastaFiles = files(2:end);
0078 
0079 <span class="comment">%Identify the operating system</span>
0080 <span class="keyword">if</span> isunix
0081     <span class="keyword">if</span> ismac
0082         binEnd=<span class="string">'.mac'</span>;
0083     <span class="keyword">else</span>
0084         binEnd=<span class="string">''</span>;
0085     <span class="keyword">end</span>
0086 <span class="keyword">elseif</span> ispc
0087     binEnd=<span class="string">'.exe'</span>;
0088     setenv(<span class="string">'BLASTDB_LMDB_MAP_SIZE'</span>,<span class="string">'1000000'</span>);
0089 <span class="keyword">else</span>
0090     dispEM(<span class="string">'Unknown OS, exiting.'</span>)
0091     <span class="keyword">return</span>
0092 <span class="keyword">end</span>
0093 
0094 <span class="comment">%Run BLAST multi-threaded to use all logical cores assigned to MATLAB</span>
0095 cores = evalc(<span class="string">'feature(''numcores'')'</span>);
0096 cores = strsplit(cores, <span class="string">'MATLAB was assigned: '</span>);
0097 cores = regexp(cores{2},<span class="string">'^\d*'</span>,<span class="string">'match'</span>);
0098 cores = cores{1};
0099 
0100 <span class="comment">%Create a database for the new organism and blast each of the refFastaFiles</span>
0101 <span class="comment">%against it</span>
0102 [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast+'</span>,[<span class="string">'makeblastdb'</span> binEnd]) <span class="string">'&quot; -in '</span> fastaFile{1} <span class="string">' -out &quot;'</span> fullfile(tmpDB, <span class="string">'tmpDB'</span>) <span class="string">'&quot; -dbtype prot'</span>]);
0103 <span class="keyword">if</span> developMode
0104     blastReport.dbHashes.phr{numel(blastReport.dbHashes.phr)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.phr'</span>));
0105     blastReport.dbHashes.pot{numel(blastReport.dbHashes.pot)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.pot'</span>));
0106     blastReport.dbHashes.psq{numel(blastReport.dbHashes.psq)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.psq'</span>));
0107     blastReport.dbHashes.pto{numel(blastReport.dbHashes.pto)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.pto'</span>));
0108 <span class="keyword">end</span>
0109 <span class="keyword">if</span> status~=0
0110     EM=[<span class="string">'makeblastdb did not run successfully, error: '</span>, num2str(status)];
0111     dispEM(EM,true);
0112 <span class="keyword">end</span>
0113 
0114 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0115     <span class="keyword">if</span> ~hideVerbose
0116         fprintf([<span class="string">'BLASTing &quot;'</span> modelIDs{i} <span class="string">'&quot; against &quot;'</span> organismID{1} <span class="string">'&quot;..\n'</span>]);
0117     <span class="keyword">end</span>
0118     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast+'</span>,[<span class="string">'blastp'</span> binEnd]) <span class="string">'&quot; -query '</span> refFastaFiles{i} <span class="string">' -out &quot;'</span> outFile <span class="string">'_'</span> num2str(i) <span class="string">'&quot; -db &quot;'</span> fullfile(tmpDB, <span class="string">'tmpDB'</span>) <span class="string">'&quot; -evalue 10e-5 -outfmt &quot;10 qseqid sseqid evalue pident length bitscore ppos&quot; -num_threads &quot;'</span> cores <span class="string">'&quot;'</span>]);
0119     <span class="keyword">if</span> developMode
0120         blastReport.blastTxtOutput{numel(blastReport.blastTxtOutput)+1}=importdata([outFile <span class="string">'_'</span> num2str(i)]);
0121     <span class="keyword">end</span>
0122     <span class="keyword">if</span> status~=0
0123         EM=[<span class="string">'blastp did not run successfully, error: '</span>, num2str(status)];
0124         dispEM(EM,true);
0125     <span class="keyword">end</span>
0126 <span class="keyword">end</span>
0127 delete([tmpDB filesep <span class="string">'tmpDB*'</span>]);
0128 
0129 <span class="comment">%Then create a database for each of the reference organisms and blast the</span>
0130 <span class="comment">%new organism against them</span>
0131 <span class="keyword">for</span> i=1:numel(refFastaFiles)
0132     <span class="keyword">if</span> ~hideVerbose
0133         fprintf([<span class="string">'BLASTing &quot;'</span> organismID{1} <span class="string">'&quot; against &quot;'</span> modelIDs{i} <span class="string">'&quot;..\n'</span>]);
0134     <span class="keyword">end</span>
0135     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast+'</span>,[<span class="string">'makeblastdb'</span> binEnd]) <span class="string">'&quot; -in '</span> refFastaFiles{i} <span class="string">' -out &quot;'</span> fullfile(tmpDB, <span class="string">'tmpDB'</span>) <span class="string">'&quot; -dbtype prot'</span>]);
0136     <span class="keyword">if</span> status~=0
0137         EM=[<span class="string">'makeblastdb did not run successfully, error: '</span>, num2str(status)];
0138         dispEM(EM,true);
0139     <span class="keyword">end</span>
0140     [status, ~]=system([<span class="string">'&quot;'</span> fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'blast+'</span>,[<span class="string">'blastp'</span> binEnd]) <span class="string">'&quot; -query '</span> fastaFile{1} <span class="string">' -out &quot;'</span> outFile <span class="string">'_r'</span> num2str(i) <span class="string">'&quot; -db &quot;'</span> fullfile(tmpDB, <span class="string">'tmpDB'</span>) <span class="string">'&quot; -evalue 10e-5 -outfmt &quot;10 qseqid sseqid evalue pident length bitscore ppos&quot; -num_threads &quot;'</span> cores <span class="string">'&quot;'</span>]);
0141     <span class="keyword">if</span> developMode
0142         blastReport.dbHashes.phr{numel(blastReport.dbHashes.phr)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.phr'</span>));
0143         blastReport.dbHashes.pot{numel(blastReport.dbHashes.pot)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.pot'</span>));
0144         blastReport.dbHashes.psq{numel(blastReport.dbHashes.psq)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.psq'</span>));
0145         blastReport.dbHashes.pto{numel(blastReport.dbHashes.pto)+1}=getMD5Hash(fullfile(tmpDB, <span class="string">'tmpDB.pto'</span>));
0146         blastReport.blastTxtOutput{numel(blastReport.blastTxtOutput)+1}=importdata([outFile <span class="string">'_r'</span> num2str(i)]);
0147     <span class="keyword">end</span>    
0148     <span class="keyword">if</span> status~=0
0149         EM=[<span class="string">'blastp did not run successfully, error: '</span>, num2str(status)];
0150         dispEM(EM,true);
0151     <span class="keyword">end</span>
0152     delete([tmpDB filesep <span class="string">'tmpDB*'</span>]);
0153 <span class="keyword">end</span>
0154 
0155 <span class="comment">%Done with the BLAST, do the parsing of the text files</span>
0156 <span class="keyword">for</span> i=1:numel(refFastaFiles)*2
0157     tempStruct=[];
0158     <span class="keyword">if</span> i&lt;=numel(refFastaFiles)
0159         tempStruct.fromId=modelIDs{i};
0160         tempStruct.toId=organismID{1};
0161         A=readtable([outFile <span class="string">'_'</span> num2str(i)],<span class="string">'Delimiter'</span>,<span class="string">','</span>,<span class="string">'Format'</span>,<span class="string">'%s%s%f%f%f%f%f'</span>);
0162     <span class="keyword">else</span>
0163         tempStruct.fromId=organismID{1};
0164         tempStruct.toId=modelIDs{i-numel(refFastaFiles)};
0165         A=readtable([outFile <span class="string">'_r'</span> num2str(i-numel(refFastaFiles))],<span class="string">'Delimiter'</span>,<span class="string">','</span>,<span class="string">'Format'</span>,<span class="string">'%s%s%f%f%f%f%f'</span>);
0166     <span class="keyword">end</span>
0167     tempStruct.fromGenes=A{:,1};
0168     tempStruct.toGenes=A{:,2};
0169     tempStruct.evalue=table2array(A(:,3));
0170     tempStruct.identity=table2array(A(:,4));
0171     tempStruct.aligLen=table2array(A(:,5));
0172     tempStruct.bitscore=table2array(A(:,6));
0173     tempStruct.ppos=table2array(A(:,7));
0174     blastStructure=[blastStructure tempStruct];
0175 <span class="keyword">end</span>
0176 
0177 <span class="comment">%Remove the old tempfiles</span>
0178 delete([outFile <span class="string">'*'</span>]);
0179 <span class="comment">%Remove the temp fasta files</span>
0180 delete(files{:})
0181 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>