<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getWoLFScores</title>
  <meta name="keywords" content="getWoLFScores">
  <meta name="description" content="getWoLFScores">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; getWoLFScores.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getWoLFScores
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getWoLFScores</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function geneScoreStruct=getWoLFScores(inputFile,kingdom) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getWoLFScores
   Calls WoLF PSort to predict the sub-cellular localization of proteins.
   The output can be used as input to predictLocalization. This function
   is currently only available for Linux and requires PERL to be
   installed. If you want to use another predictor, see parseScores.

   inputFile    a FASTA file with protein sequences
   kingdom     the kingdom of the organism, 'animal', 'fungi' or 'plant'.

   The function normalizes the scores so that the best score for each gene
   is 1.0.

   geneScoreStructure  a structure to be used in predictLocalization

   Usage: geneScoreStruct=getWoLFScores(inputFile,kingdom)

   Rasmus Agren, 2014-01-08</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function geneScoreStruct=getWoLFScores(inputFile,kingdom)</a>
0002 <span class="comment">% getWoLFScores</span>
0003 <span class="comment">%   Calls WoLF PSort to predict the sub-cellular localization of proteins.</span>
0004 <span class="comment">%   The output can be used as input to predictLocalization. This function</span>
0005 <span class="comment">%   is currently only available for Linux and requires PERL to be</span>
0006 <span class="comment">%   installed. If you want to use another predictor, see parseScores.</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   inputFile    a FASTA file with protein sequences</span>
0009 <span class="comment">%   kingdom     the kingdom of the organism, 'animal', 'fungi' or 'plant'.</span>
0010 <span class="comment">%</span>
0011 <span class="comment">%   The function normalizes the scores so that the best score for each gene</span>
0012 <span class="comment">%   is 1.0.</span>
0013 <span class="comment">%</span>
0014 <span class="comment">%   geneScoreStructure  a structure to be used in predictLocalization</span>
0015 <span class="comment">%</span>
0016 <span class="comment">%   Usage: geneScoreStruct=getWoLFScores(inputFile,kingdom)</span>
0017 <span class="comment">%</span>
0018 <span class="comment">%   Rasmus Agren, 2014-01-08</span>
0019 <span class="comment">%</span>
0020 
0021 <span class="keyword">if</span> ~strcmp(kingdom,<span class="string">'animal'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'fungi'</span>) &amp;&amp; ~strcmp(kingdom,<span class="string">'plant'</span>)
0022     EM=<span class="string">'Allowed kingdoms are &quot;animal&quot;, &quot;fungi&quot;, and &quot;plant&quot;'</span>;
0023     dispEM(EM);
0024 <span class="keyword">end</span>
0025 
0026 <span class="keyword">if</span> ispc==true
0027     EM=<span class="string">'This function currently runs only on Linux. Use parseScores if you want to use another predictor'</span>;
0028     dispEM(EM);
0029 <span class="keyword">end</span>
0030 
0031 <span class="comment">%Get the directory for RAVEN Toolbox. This may not be the easiest or best</span>
0032 <span class="comment">%way to do this</span>
0033 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0034 ravenPath=fileparts(fileparts(ST(I).file));
0035 
0036 <span class="comment">%Temporary output name</span>
0037 outFile=tempname;
0038 fid=fopen(outFile,<span class="string">'w'</span>);
0039 
0040 <span class="comment">%Do the prediction</span>
0041 [~, output]=unix([<span class="string">'perl &quot;'</span> ravenPath <span class="string">'/software/WoLFPSORT_package_v0.2/bin/runWolfPsortSummary&quot; '</span> kingdom <span class="string">' &lt; '</span> inputFile]);
0042 
0043 <span class="comment">%Save output and call the general parser</span>
0044 fprintf(fid,output);
0045 fclose(fid);
0046 geneScoreStruct=parseScores(outFile,<span class="string">'wolf'</span>);
0047 
0048 <span class="comment">%Clean up</span>
0049 delete(outFile);
0050 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Mon 14-May-2018 15:44:35 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>