<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of makeFakeBlastStructure</title>
  <meta name="keywords" content="makeFakeBlastStructure">
  <meta name="description" content="makeFakeBlastStructure">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">external</a> &gt; makeFakeBlastStructure.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>makeFakeBlastStructure
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>makeFakeBlastStructure</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function blastStructure=makeFakeBlastStructure(orthologList,sourceModelID,getModelFor) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> makeFakeBlastStructure
   Makes a fake blastStructure, that would normally be generated by
   getBlast. This allows to feed a predefined list of orthologs to
   getModelFromHomology while retaining the further use of that function.
   For this function to work, it is crucial that the orthologList is a
   cell array where the first column contains the genes from the source
   organism, and the second column contains the genes from the target
   organism
  
   orthologList        cell array of orthologous genes, where the first
                       column contains the genes from the source organism,
                       while the second column contains the genes from the
                       target organism
   sourceModelID       ID of the model that will be used as template, that
                       contains the genes in the first column of
                       orthologList
   getModelFor         the name of the organism to build a model for,
                       identical to the getModelFor parameter in the
                       getModelFromHomology function

   blastStructure      a fake blastStructure, where the evalue, identity
                       and aligLen are set at extreme values, such that
                       all orthologous pairs will pass the filter when
                       running getModelFromHomology

   Eduard Kerkhoven, 2017-10-22</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function blastStructure=makeFakeBlastStructure(orthologList,sourceModelID,getModelFor)</a>
0002 <span class="comment">% makeFakeBlastStructure</span>
0003 <span class="comment">%   Makes a fake blastStructure, that would normally be generated by</span>
0004 <span class="comment">%   getBlast. This allows to feed a predefined list of orthologs to</span>
0005 <span class="comment">%   getModelFromHomology while retaining the further use of that function.</span>
0006 <span class="comment">%   For this function to work, it is crucial that the orthologList is a</span>
0007 <span class="comment">%   cell array where the first column contains the genes from the source</span>
0008 <span class="comment">%   organism, and the second column contains the genes from the target</span>
0009 <span class="comment">%   organism</span>
0010 <span class="comment">%</span>
0011 <span class="comment">%   orthologList        cell array of orthologous genes, where the first</span>
0012 <span class="comment">%                       column contains the genes from the source organism,</span>
0013 <span class="comment">%                       while the second column contains the genes from the</span>
0014 <span class="comment">%                       target organism</span>
0015 <span class="comment">%   sourceModelID       ID of the model that will be used as template, that</span>
0016 <span class="comment">%                       contains the genes in the first column of</span>
0017 <span class="comment">%                       orthologList</span>
0018 <span class="comment">%   getModelFor         the name of the organism to build a model for,</span>
0019 <span class="comment">%                       identical to the getModelFor parameter in the</span>
0020 <span class="comment">%                       getModelFromHomology function</span>
0021 <span class="comment">%</span>
0022 <span class="comment">%   blastStructure      a fake blastStructure, where the evalue, identity</span>
0023 <span class="comment">%                       and aligLen are set at extreme values, such that</span>
0024 <span class="comment">%                       all orthologous pairs will pass the filter when</span>
0025 <span class="comment">%                       running getModelFromHomology</span>
0026 <span class="comment">%</span>
0027 <span class="comment">%   Eduard Kerkhoven, 2017-10-22</span>
0028 <span class="comment">%</span>
0029 
0030 <span class="keyword">if</span> nargin&lt;3
0031     throw(MException(<span class="string">''</span>,<span class="string">'All three parameters should be set'</span>));
0032 <span class="keyword">end</span>
0033 
0034 <span class="comment">% Check whether orthologList is a cell array, with two columns, and more</span>
0035 <span class="comment">% than 10 rows</span>
0036 <span class="keyword">if</span> (~iscell(orthologList) || ~(size(orthologList,2)==2) || ~(size(orthologList,1)&gt;=10))
0037     throw(MException(<span class="string">''</span>,<span class="string">'orthologList should be a cell array with two columns'</span>));
0038 <span class="keyword">end</span>
0039 
0040 blastStructure(1).fromId=sourceModelID;
0041 blastStructure(1).toId=getModelFor;
0042 blastStructure(1).fromGenes=orthologList(:,1);
0043 blastStructure(1).toGenes=orthologList(:,2);
0044 
0045 blastStructure(2).fromId=getModelFor;
0046 blastStructure(2).toId=sourceModelID;
0047 blastStructure(2).fromGenes=orthologList(:,2);
0048 blastStructure(2).toGenes=orthologList(:,1);
0049 
0050 blastStructure(1).evalue=zeros(size(orthologList,1),1);
0051 blastStructure(2).evalue=zeros(size(orthologList,1),1);
0052 blastStructure(1).identity(1:size(orthologList,1),1)=100;
0053 blastStructure(2).identity(1:size(orthologList,1),1)=100;
0054 blastStructure(1).aligLen(1:size(orthologList,1),1)=1000;
0055 blastStructure(2).aligLen(1:size(orthologList,1),1)=1000;
0056 blastStructure(1).bitscore(1:size(orthologList,1),1)=100;
0057 blastStructure(2).bitscore(1:size(orthologList,1),1)=100;
0058 blastStructure(1).ppos(1:size(orthologList,1),1)=100;
0059 blastStructure(2).ppos(1:size(orthologList,1),1)=100;
0060 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 06-Nov-2019 21:22:42 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>