<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getGenesFromKEGG</title>
  <meta name="keywords" content="getGenesFromKEGG">
  <meta name="description" content="getGenesFromKEGG">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getGenesFromKEGG.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getGenesFromKEGG
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getGenesFromKEGG</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function model=getGenesFromKEGG(keggPath,koList) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getGenesFromKEGG
   Retrieves information on all genes stored in KEGG database

   keggPath    if keggGenes.mat is not in the RAVEN\external\kegg directory,
               this function will attempt to read data from a local FTP dump
               of the KEGG database. keggPath is the path to the root of
               this database

   koList      the number of genes in KEGG is very large. koList can be a
               cell array with KO identifiers, in which case only genes
               belonging to one of those KEGG orthologies are retrieved (opt,
               default all KOs with associated reactions)

   model       a model structure generated from the database. The following
               fields are filled
               id:             'KEGG'
               description:    'Automatically generated from KEGG database'
               rxns:           KO ids
               rxnNames:       Name for each entry
               genes:          IDs for all the genes. Genes are saved as
                               organism abbreviation:id (same as in KEGG).
                               'HSA:124' for example is alcohol dehydrogenase
                               in Homo sapiens
               rxnGeneMat      A binary matrix that indicates whether a
                               specific gene is present in a KO id

   If the file keggGenes.mat is in the RAVEN\external\kegg directory it
   will be loaded instead of parsing of the KEGG files. If it does not exist
   it will be saved after parsing of the KEGG files. In general, you should
   remove the keggGenes.mat file if you want to rebuild the model structure
   from a newer version of KEGG.

   Usage: model=getGenesFromKEGG(keggPath,koList)

   Simonas Marcisauskas, 2017-05-02</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getModelFromKEGG.html" class="code" title="function [model, KOModel]=getModelFromKEGG(keggPath,keepUndefinedStoich,keepIncomplete,keepGeneral)">getModelFromKEGG</a>	getModelFromKEGG</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function allKOs=getAllKOs(keggPath)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=getGenesFromKEGG(keggPath,koList)</a>
0002 <span class="comment">% getGenesFromKEGG</span>
0003 <span class="comment">%   Retrieves information on all genes stored in KEGG database</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   keggPath    if keggGenes.mat is not in the RAVEN\external\kegg directory,</span>
0006 <span class="comment">%               this function will attempt to read data from a local FTP dump</span>
0007 <span class="comment">%               of the KEGG database. keggPath is the path to the root of</span>
0008 <span class="comment">%               this database</span>
0009 <span class="comment">%</span>
0010 <span class="comment">%   koList      the number of genes in KEGG is very large. koList can be a</span>
0011 <span class="comment">%               cell array with KO identifiers, in which case only genes</span>
0012 <span class="comment">%               belonging to one of those KEGG orthologies are retrieved (opt,</span>
0013 <span class="comment">%               default all KOs with associated reactions)</span>
0014 <span class="comment">%</span>
0015 <span class="comment">%   model       a model structure generated from the database. The following</span>
0016 <span class="comment">%               fields are filled</span>
0017 <span class="comment">%               id:             'KEGG'</span>
0018 <span class="comment">%               description:    'Automatically generated from KEGG database'</span>
0019 <span class="comment">%               rxns:           KO ids</span>
0020 <span class="comment">%               rxnNames:       Name for each entry</span>
0021 <span class="comment">%               genes:          IDs for all the genes. Genes are saved as</span>
0022 <span class="comment">%                               organism abbreviation:id (same as in KEGG).</span>
0023 <span class="comment">%                               'HSA:124' for example is alcohol dehydrogenase</span>
0024 <span class="comment">%                               in Homo sapiens</span>
0025 <span class="comment">%               rxnGeneMat      A binary matrix that indicates whether a</span>
0026 <span class="comment">%                               specific gene is present in a KO id</span>
0027 <span class="comment">%</span>
0028 <span class="comment">%   If the file keggGenes.mat is in the RAVEN\external\kegg directory it</span>
0029 <span class="comment">%   will be loaded instead of parsing of the KEGG files. If it does not exist</span>
0030 <span class="comment">%   it will be saved after parsing of the KEGG files. In general, you should</span>
0031 <span class="comment">%   remove the keggGenes.mat file if you want to rebuild the model structure</span>
0032 <span class="comment">%   from a newer version of KEGG.</span>
0033 <span class="comment">%</span>
0034 <span class="comment">%   Usage: model=getGenesFromKEGG(keggPath,koList)</span>
0035 <span class="comment">%</span>
0036 <span class="comment">%   Simonas Marcisauskas, 2017-05-02</span>
0037 <span class="comment">%</span>
0038 
0039 <span class="comment">%NOTE: This is how one entry looks in the file</span>
0040 
0041 <span class="comment">% ENTRY       K11440                      KO</span>
0042 <span class="comment">% NAME        gbsB</span>
0043 <span class="comment">% DEFINITION  choline dehydrogenase [EC:1.1.1.1]</span>
0044 <span class="comment">% CLASS       Metabolism; Amino Acid Metabolism; Glycine, serine and threonine</span>
0045 <span class="comment">%             metabolism [PATH:ko00260]</span>
0046 <span class="comment">% DBLINKS     RN: R08557 R08558</span>
0047 <span class="comment">%             COG: COG1454</span>
0048 <span class="comment">% GENES       BSU: BSU31050(gbsB)</span>
0049 <span class="comment">%             BCY: Bcer98_1477</span>
0050 <span class="comment">%             BLI: BL02563(gbsB)</span>
0051 <span class="comment">%             BLD: BLi03269(gbsB)</span>
0052 <span class="comment">%             BCL: ABC0979(gbsB)</span>
0053 <span class="comment">%             BAY: RBAM_028150(gbsB)</span>
0054 <span class="comment">%             BPU: BPUM_2734(gbsB)</span>
0055 <span class="comment">% ADDITIONAL  INFO...</span>
0056 <span class="comment">% ///</span>
0057 
0058 <span class="comment">%The file is not tab-delimited. Instead each label is 12 characters</span>
0059 <span class="comment">%(except for '///')</span>
0060 
0061 <span class="comment">%Check if the genes have been parsed before and saved. If so, load the</span>
0062 <span class="comment">%model.</span>
0063 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0064 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0065 genesFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggGenes.mat'</span>);
0066 <span class="keyword">if</span> exist(genesFile, <span class="string">'file'</span>)
0067     fprintf([<span class="string">'NOTE: Importing KEGG genes from '</span> strrep(genesFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0068     load(genesFile);
0069 <span class="keyword">else</span>
0070     fprintf([<span class="string">'Cannot locate '</span> strrep(genesFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">' and will try to generate it from the local KEGG database.\n'</span>]);
0071     <span class="keyword">if</span> ~exist(fullfile(keggPath,<span class="string">'ko'</span>),<span class="string">'file'</span>) || ~exist(fullfile(keggPath,<span class="string">'reaction'</span>),<span class="string">'file'</span>)
0072         EM=fprintf([<span class="string">'The files ''ko'' and ''reaction'' cannot be located at '</span> strrep(keggPath,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'/ and should be downloaded from the KEGG FTP.\n'</span>]);
0073         dispEM(EM);
0074     <span class="keyword">else</span>
0075         <span class="comment">%Get all KOs that are associated to reactions</span>
0076         allKOs=<a href="#_sub1" class="code" title="subfunction allKOs=getAllKOs(keggPath)">getAllKOs</a>(keggPath);
0077    
0078         <span class="comment">%Since the list of genes will be accessed many times I use a hash table</span>
0079         geneMap=containers.Map(<span class="string">'KeyType'</span>,<span class="string">'char'</span>,<span class="string">'ValueType'</span>,<span class="string">'int32'</span>);
0080    
0081         <span class="comment">%Add new functionality in the order specified in models</span>
0082         model.id=<span class="string">'KEGG'</span>;
0083         model.description=<span class="string">'Automatically generated from KEGG database'</span>;
0084    
0085         <span class="comment">%Preallocate memory</span>
0086         model.rxns=cell(numel(allKOs),1);
0087         model.rxnNames=cell(numel(allKOs),1);
0088    
0089         <span class="comment">%Reserve space for 500000 genes</span>
0090         model.genes=cell(500000,1);
0091    
0092         <span class="comment">%Load information on KO ID, name, and associated genes</span>
0093         fid = fopen(fullfile(keggPath,<span class="string">'ko'</span>), <span class="string">'r'</span>);
0094    
0095         <span class="comment">%Keeps track of how many KOs that have been added</span>
0096         koCounter=0;
0097         nGenes=0;
0098         addingGenes=false;
0099    
0100         <span class="comment">%These contain the mapping between KOs and genes and are used to</span>
0101         <span class="comment">%generate the model.rxnGeneMat in the end</span>
0102         koIndex=zeros(5000000,1);
0103         geneIndex=koIndex;
0104         nMappings=0;
0105    
0106         skipRead=false;
0107    
0108         <span class="comment">%Loop through the file</span>
0109         <span class="keyword">while</span> 1
0110           <span class="comment">%Get the next line</span>
0111           <span class="keyword">if</span> skipRead==false
0112             tline = fgetl(fid);
0113           <span class="keyword">else</span>
0114             skipRead=false;
0115           <span class="keyword">end</span>
0116    
0117           <span class="comment">%Abort at end of file</span>
0118           <span class="keyword">if</span> ~ischar(tline)
0119               <span class="keyword">break</span>;
0120           <span class="keyword">end</span>
0121    
0122           <span class="comment">%Skip '///'</span>
0123           <span class="keyword">if</span> numel(tline)&lt;12
0124               addingGenes=false;
0125               <span class="keyword">continue</span>;
0126           <span class="keyword">end</span>
0127    
0128           <span class="comment">%Check if it's a new reaction</span>
0129           <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'ENTRY       '</span>)
0130               <span class="comment">%Check if it should be added</span>
0131               koID=tline(13:18);
0132    
0133               <span class="keyword">if</span> ismember(koID,allKOs)
0134                  addMe=true;
0135                  koCounter=koCounter+1;
0136               <span class="keyword">else</span>
0137                  addMe=false;
0138                  <span class="keyword">continue</span>;
0139               <span class="keyword">end</span>
0140    
0141               <span class="comment">%Add reaction ID (always 6 characters)</span>
0142               model.rxns{koCounter}=koID;
0143               model.rxnNames{koCounter}=<span class="string">''</span>; <span class="comment">%Will be overwritten if it exists</span>
0144           <span class="keyword">end</span>
0145    
0146           <span class="comment">%To get here we must be in a KO that should be added</span>
0147           <span class="keyword">if</span> addMe==true
0148               <span class="comment">%Add name</span>
0149               <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'DEFINITION  '</span>)
0150                   model.rxnNames{koCounter}=tline(13:end);
0151               <span class="keyword">end</span>
0152    
0153               <span class="comment">%Check if it's time to start adding genes</span>
0154               <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'GENES       '</span>)
0155                   addingGenes=true;
0156               <span class="keyword">end</span>
0157    
0158               <span class="comment">%Add genes</span>
0159               <span class="keyword">if</span> addingGenes==true
0160                  <span class="comment">%We are now adding genes for the current KO. All gene rows start</span>
0161                  <span class="comment">%with 12 spaces. If this is not true it means that there is</span>
0162                  <span class="comment">%additional annotation such as AUTHORS. This is not common but it</span>
0163                  <span class="comment">%exists.</span>
0164                  <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'            '</span>) || strcmp(tline(1:12),<span class="string">'GENES       '</span>)
0165                      geneLine=tline(13:end);
0166    
0167                      <span class="comment">%Check if the line is from a new organism of from the same as</span>
0168                      <span class="comment">%before</span>
0169                      <span class="keyword">if</span> strcmp(geneLine(4),<span class="string">':'</span>)
0170                          <span class="comment">% if organism id contains 3 letters;</span>
0171                          currentOrganism=geneLine(1:3);
0172                          <span class="comment">%Parse the string</span>
0173                          genes=regexp(geneLine(6:end),<span class="string">' '</span>,<span class="string">'split'</span>);
0174                          genes=strcat(currentOrganism,<span class="string">':'</span>,genes(:));
0175                      <span class="keyword">else</span> <span class="keyword">if</span> strcmp(geneLine(5),<span class="string">':'</span>)
0176                          <span class="comment">% if organism id contains 4 letters;</span>
0177                          currentOrganism=geneLine(1:4);
0178                          <span class="comment">%Parse the string</span>
0179                          genes=regexp(geneLine(7:end),<span class="string">' '</span>,<span class="string">'split'</span>);
0180                          genes=strcat(currentOrganism,<span class="string">':'</span>,genes(:));
0181                          <span class="keyword">end</span>
0182                      <span class="keyword">end</span>
0183    
0184                      <span class="comment">%Add the genes to the gene list</span>
0185                      <span class="keyword">for</span> i=1:numel(genes)
0186                          geneString=genes{i};
0187                          <span class="keyword">if</span> geneMap.isKey(geneString);
0188                             index=geneMap(geneString);
0189                          <span class="keyword">else</span>
0190                             nGenes=nGenes+1;
0191                             model.genes{nGenes}=geneString;
0192                             index=nGenes;
0193                             geneMap(geneString)=index;
0194                          <span class="keyword">end</span>
0195    
0196                          <span class="comment">%Add the connection to the KO</span>
0197                          nMappings=nMappings+1;
0198                          koIndex(nMappings)=koCounter;
0199                          geneIndex(nMappings)=index;
0200                      <span class="keyword">end</span>
0201                  <span class="keyword">else</span>
0202                      <span class="comment">%Now we want to throw away everything until the next</span>
0203                      <span class="comment">%entry</span>
0204                      <span class="keyword">while</span> 1
0205                         tline = fgetl(fid);
0206                         <span class="keyword">if</span> strcmp(tline,<span class="string">'///'</span>)
0207                             <span class="comment">%When the new entry is found, skip reading one line to</span>
0208                             <span class="comment">%fit with the rest of the code</span>
0209                             skipRead=true;
0210                             <span class="keyword">break</span>;
0211                         <span class="keyword">end</span>
0212                      <span class="keyword">end</span>
0213                  <span class="keyword">end</span>
0214               <span class="keyword">end</span>
0215           <span class="keyword">end</span>
0216         <span class="keyword">end</span>
0217         <span class="comment">%Close the file</span>
0218         fclose(fid);
0219    
0220         <span class="comment">%If too much space was allocated, shrink the model</span>
0221         model.rxns=model.rxns(1:koCounter);
0222         model.rxnNames=model.rxnNames(1:koCounter);
0223         model.genes=model.genes(1:nGenes);
0224    
0225         <span class="comment">%Retrieve and add the gene associations</span>
0226         model.rxnGeneMat=sparse(koIndex(1:nMappings),geneIndex(1:nMappings),ones(nMappings,1));
0227    
0228         <span class="comment">%To make sure the size is correct if the last KOs don't have genes</span>
0229         <span class="keyword">if</span> size(model.rxnGeneMat,1)~=koCounter
0230             model.rxnGeneMat(koCounter,1)=0;
0231         <span class="keyword">end</span>
0232    
0233         <span class="comment">%Save the model structure</span>
0234         save(genesFile,<span class="string">'model'</span>);
0235     <span class="keyword">end</span>
0236 <span class="keyword">end</span>
0237 
0238 <span class="comment">%Only get the KOs in koList</span>
0239 I=~ismember(model.rxns,koList);
0240 model=removeReactions(model,I,true,true);
0241 <span class="keyword">end</span>
0242 
0243 <a name="_sub1" href="#_subfunctions" class="code">function allKOs=getAllKOs(keggPath)</a>
0244     <span class="comment">%Retrieves all KOs that are associated to reactions. This is because</span>
0245     <span class="comment">%the number of genes in KEGG is very large so without this parsing it</span>
0246     <span class="comment">%would take many hours</span>
0247 
0248     allKOs={};
0249 
0250     <span class="comment">%First check if the reactions have already been parsed</span>
0251     [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0252     ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0253     rxnsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggRxns.mat'</span>);
0254     <span class="keyword">if</span> exist(rxnsFile, <span class="string">'file'</span>)
0255         fprintf([<span class="string">'NOTE: Importing KEGG ORTHOLOGY list from '</span> strrep(rxnsFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0256         load(rxnsFile,<span class="string">'model'</span>);
0257 
0258         <span class="comment">%Loop through the reactions and add the corresponding genes</span>
0259         <span class="keyword">for</span> i=1:numel(model.rxns)
0260             <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0261                 <span class="comment">%Get all KOs</span>
0262                 allKOs=[allKOs;model.rxnMiriams{i}.value(strcmpi(model.rxnMiriams{i}.name,<span class="string">'kegg.orthology'</span>))];
0263             <span class="keyword">end</span>
0264         <span class="keyword">end</span>
0265     <span class="keyword">else</span>
0266         <span class="comment">%Parse the reaction file instead</span>
0267         <span class="comment">%First load information on reaction ID, reaction name, KO, pathway, and ec-number</span>
0268         fid = fopen(fullfile(keggPath,<span class="string">'reaction'</span>), <span class="string">'r'</span>);
0269         orthology=false;
0270         <span class="keyword">while</span> 1
0271           <span class="comment">%Get the next line</span>
0272           tline = fgetl(fid);
0273 
0274           <span class="comment">%Abort at end of file</span>
0275           <span class="keyword">if</span> ~ischar(tline)
0276               <span class="keyword">break</span>;
0277           <span class="keyword">end</span>
0278 
0279           <span class="comment">%Skip '///'</span>
0280           <span class="keyword">if</span> numel(tline)&lt;12
0281               <span class="keyword">continue</span>;
0282           <span class="keyword">end</span>
0283 
0284           <span class="comment">%Check if it's a new reaction</span>
0285           <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'ENTRY       '</span>)
0286               orthology=false;
0287           <span class="keyword">end</span>
0288 
0289           <span class="keyword">if</span> strcmp(tline(1:9),<span class="string">'REFERENCE'</span>)
0290               orthology=false;
0291           <span class="keyword">end</span>
0292 
0293           <span class="comment">%Add KO-ids</span>
0294           <span class="keyword">if</span> numel(tline)&gt;16
0295               <span class="keyword">if</span> strcmp(tline(1:16),<span class="string">'ORTHOLOGY   KO: '</span>) || strcmp(tline(1:16),<span class="string">'            KO: '</span>) || strcmp(tline(1:12),<span class="string">'ORTHOLOGY   '</span>) || orthology==true
0296                   <span class="keyword">if</span> strcmp(tline(13:16),<span class="string">'KO:'</span>) <span class="comment">%This is in the old version</span>
0297                     allKOs=[allKOs;tline(17:22)];
0298                   <span class="keyword">else</span>
0299                       <span class="keyword">if</span> strcmp(tline(1:12),<span class="string">'ORTHOLOGY   '</span>)
0300                         <span class="comment">%This means that it found one KO in the new format and that</span>
0301                         <span class="comment">%subsequent lines might be other KOs</span>
0302                         orthology=true;
0303                       <span class="keyword">end</span>
0304                       allKOs=[allKOs;tline(13:18)];
0305                   <span class="keyword">end</span>
0306               <span class="keyword">end</span>
0307           <span class="keyword">end</span>
0308         <span class="keyword">end</span>
0309 
0310         <span class="comment">%Close the file</span>
0311         fclose(fid);
0312     <span class="keyword">end</span>
0313     allKOs=unique(allKOs);
0314 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 28-Dec-2017 10:46:27 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>