<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromKEGG</title>
  <meta name="keywords" content="getModelFromKEGG">
  <meta name="description" content="getModelFromKEGG">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getModelFromKEGG.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getModelFromKEGG
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getModelFromKEGG</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,keepUndefinedStoich,keepIncomplete,keepGeneral) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromKEGG
   Retrieves information stored in KEGG database and generates a model

   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat or
                       keggRxns.mat are not in the RAVEN/external/kegg
                       directory, this function will attempt to read data
                       from a local FTP dump of the KEGG database.
                       keggPath is the path to the root of this database
                       (opt, default 'RAVEN/external/kegg'). If
                       keggModel.mat is present in the same directory, the
                       function reads the data from this file and ignores
                       keggGenes.mat, keggMets.mat and keggRxns.mat
   keepSpontaneous     include reactions labeled as &quot;spontaneous&quot; (opt,
                       default true)
   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These
                       will be dealt with as two separate metabolites
                       (opt, default true)
   keepIncomplete      include reactions which have been labelled as
                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,
                       default true)
   keepGeneral         include reactions which have been labelled as
                       &quot;general reaction&quot;. These are reactions on the form
                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore
                       unsuited for modelling purposes. Note that not all
                       reactions have this type of annotation, and the
                       script will therefore not be able to remove all
                       such reactions (opt, default false)

   model               a model structure generated from the database. All
                       reactions and the metabolites used in them will be
                       added
   KOModel             a model structure representing the KEGG Orthology
                       ids and their associated genes. The KO ids are
                       saved as reactions

   Usage: [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,...
    keepUndefinedStoich,keepIncomplete,keepGeneral)

   Note:               The model output from getModelFromKEGG can be used
                       as template for fillGaps. In that case, ensure that
                       the genes and rxnGeneMat fields are removed before
                       parsing: model=rmfield(model,'genes'), etc.

   Simonas Marcisauskas, 2018-09-06</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>	getGenesFromKEGG</li><li><a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>	getMetsFromKEGG</li><li><a href="getRxnsFromKEGG.html" class="code" title="function [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=getRxnsFromKEGG(keggPath)">getRxnsFromKEGG</a>	getRxnsFromKEGG</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepSpontaneous,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioKO,minScoreRatioG,maxPhylDist,nSequences,seqIdentity)">getKEGGModelForOrganism</a>	getKEGGModelForOrganism</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function output = isNewestFile(ravenPath)</a></li><li><a href="#_sub2" class="code">function modTime = getFileTime(fileName)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,</a><span class="keyword">...</span>
0002     keepUndefinedStoich,keepIncomplete,keepGeneral)
0003 <span class="comment">% getModelFromKEGG</span>
0004 <span class="comment">%   Retrieves information stored in KEGG database and generates a model</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat or</span>
0007 <span class="comment">%                       keggRxns.mat are not in the RAVEN/external/kegg</span>
0008 <span class="comment">%                       directory, this function will attempt to read data</span>
0009 <span class="comment">%                       from a local FTP dump of the KEGG database.</span>
0010 <span class="comment">%                       keggPath is the path to the root of this database</span>
0011 <span class="comment">%                       (opt, default 'RAVEN/external/kegg'). If</span>
0012 <span class="comment">%                       keggModel.mat is present in the same directory, the</span>
0013 <span class="comment">%                       function reads the data from this file and ignores</span>
0014 <span class="comment">%                       keggGenes.mat, keggMets.mat and keggRxns.mat</span>
0015 <span class="comment">%   keepSpontaneous     include reactions labeled as &quot;spontaneous&quot; (opt,</span>
0016 <span class="comment">%                       default true)</span>
0017 <span class="comment">%   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These</span>
0018 <span class="comment">%                       will be dealt with as two separate metabolites</span>
0019 <span class="comment">%                       (opt, default true)</span>
0020 <span class="comment">%   keepIncomplete      include reactions which have been labelled as</span>
0021 <span class="comment">%                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,</span>
0022 <span class="comment">%                       default true)</span>
0023 <span class="comment">%   keepGeneral         include reactions which have been labelled as</span>
0024 <span class="comment">%                       &quot;general reaction&quot;. These are reactions on the form</span>
0025 <span class="comment">%                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore</span>
0026 <span class="comment">%                       unsuited for modelling purposes. Note that not all</span>
0027 <span class="comment">%                       reactions have this type of annotation, and the</span>
0028 <span class="comment">%                       script will therefore not be able to remove all</span>
0029 <span class="comment">%                       such reactions (opt, default false)</span>
0030 <span class="comment">%</span>
0031 <span class="comment">%   model               a model structure generated from the database. All</span>
0032 <span class="comment">%                       reactions and the metabolites used in them will be</span>
0033 <span class="comment">%                       added</span>
0034 <span class="comment">%   KOModel             a model structure representing the KEGG Orthology</span>
0035 <span class="comment">%                       ids and their associated genes. The KO ids are</span>
0036 <span class="comment">%                       saved as reactions</span>
0037 <span class="comment">%</span>
0038 <span class="comment">%   Usage: [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,...</span>
0039 <span class="comment">%    keepUndefinedStoich,keepIncomplete,keepGeneral)</span>
0040 <span class="comment">%</span>
0041 <span class="comment">%   Note:               The model output from getModelFromKEGG can be used</span>
0042 <span class="comment">%                       as template for fillGaps. In that case, ensure that</span>
0043 <span class="comment">%                       the genes and rxnGeneMat fields are removed before</span>
0044 <span class="comment">%                       parsing: model=rmfield(model,'genes'), etc.</span>
0045 <span class="comment">%</span>
0046 <span class="comment">%   Simonas Marcisauskas, 2018-09-06</span>
0047 <span class="comment">%</span>
0048 
0049 <span class="keyword">if</span> nargin&lt;1
0050     keggPath=<span class="string">'RAVEN/external/kegg'</span>;
0051 <span class="keyword">end</span>
0052 <span class="keyword">if</span> nargin&lt;2
0053     keepSpontaneous=true;
0054 <span class="keyword">end</span>
0055 <span class="keyword">if</span> nargin&lt;3
0056     keepUndefinedStoich=true;
0057 <span class="keyword">end</span>
0058 <span class="keyword">if</span> nargin&lt;4
0059     keepIncomplete=true;
0060 <span class="keyword">end</span>
0061 <span class="keyword">if</span> nargin&lt;5
0062     keepGeneral=false;
0063 <span class="keyword">end</span>
0064 
0065 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0066 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0067 modelFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggModel.mat'</span>);
0068 <span class="keyword">if</span> exist(modelFile, <span class="string">'file'</span>) &amp;&amp; <a href="#_sub1" class="code" title="subfunction output = isNewestFile(ravenPath)">isNewestFile</a>(ravenPath)
0069     fprintf([<span class="string">'NOTE: Importing KEGG model from '</span> strrep(modelFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0070     load(modelFile);
0071 <span class="keyword">else</span>
0072     fprintf([<span class="string">'The file '</span> strrep(modelFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">' cannot be located or is older than kegg***.mat files. It will therefore be regenerated.\n'</span>]);
0073     <span class="comment">%First get all reactions</span>
0074     [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=<a href="getRxnsFromKEGG.html" class="code" title="function [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=getRxnsFromKEGG(keggPath)">getRxnsFromKEGG</a>(keggPath);
0075     fprintf(<span class="string">'KEGG reactions loaded\n'</span>);
0076     
0077     <span class="comment">%Get the KO ids that are associated with any of the reactions. They</span>
0078     <span class="comment">%will be used later on to create a rxn-gene matrix</span>
0079     KOs=cell(numel(model.rxns)*2,1);
0080     <span class="comment">%Make room for two KO ids per reaction</span>
0081     
0082     addToIndex=1;
0083     <span class="comment">%Add all KO ids that are used</span>
0084     <span class="keyword">for</span> i=1:numel(model.rxns)
0085         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0086             <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0087                 <span class="keyword">if</span> strcmpi(<span class="string">'kegg.orthology'</span>,model.rxnMiriams{i}.name{j})
0088                     <span class="comment">%Add the KO id</span>
0089                     KOs(addToIndex)=model.rxnMiriams{i}.value(j);
0090                     addToIndex=addToIndex+1;
0091                 <span class="keyword">end</span>
0092             <span class="keyword">end</span>
0093         <span class="keyword">end</span>
0094     <span class="keyword">end</span>
0095     
0096     KOs=KOs(1:addToIndex-1);
0097     KOs=unique(KOs);
0098     
0099     <span class="comment">%Get all genes from any organism in KEGG that is associated with any of</span>
0100     <span class="comment">%the KOs</span>
0101     KOModel=<a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>(keggPath,KOs);
0102     fprintf(<span class="string">'KEGG genes loaded\n'</span>);
0103     
0104     model.genes=KOModel.genes;
0105     
0106     <span class="comment">%It can be that there are KOs from the reactions that have no database</span>
0107     <span class="comment">%entry. These are (as far as I've seen) lumped versions of other KOs</span>
0108     <span class="comment">%and should be removed</span>
0109     KOsToRemove=setdiff(KOs, KOModel.rxns);
0110     
0111     <span class="comment">%Loop through all reactions and delete the KOs that weren't found</span>
0112     <span class="keyword">for</span> i=1:numel(model.rxns)
0113         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0114             <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0115                 toDel=[];
0116                 <span class="keyword">if</span> strcmp(model.rxnMiriams{i}.name{j},<span class="string">'kegg.orthology'</span>)
0117                     <span class="keyword">if</span> ismember(model.rxnMiriams{i}.value{j},KOsToRemove)
0118                         toDel=[toDel;j];
0119                     <span class="keyword">end</span>
0120                 <span class="keyword">end</span>
0121             <span class="keyword">end</span>
0122             <span class="comment">%Delete the KOs</span>
0123             <span class="keyword">if</span> any(toDel)
0124                 <span class="comment">%If all posts are deleted, then delete the whole structure</span>
0125                 <span class="keyword">if</span> numel(toDel)==j
0126                     model.rxnMiriams{i}=[];
0127                 <span class="keyword">else</span>
0128                     model.rxnMiriams{i}.name(toDel)=[];
0129                     model.rxnMiriams{i}.value(toDel)=[];
0130                 <span class="keyword">end</span>
0131             <span class="keyword">end</span>
0132         <span class="keyword">end</span>
0133     <span class="keyword">end</span>
0134     
0135     <span class="comment">%Create the rxnGeneMat for the reactions. This is simply done by</span>
0136     <span class="comment">%merging the gene associations for all the involved KOs</span>
0137     r=zeros(10000000,1);
0138     <span class="comment">%Store the positions since it's slow to write to a sparse array in a</span>
0139     <span class="comment">%loop</span>
0140     c=zeros(10000000,1);
0141     counter=1;
0142     <span class="keyword">for</span> i=1:numel(model.rxns)
0143         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0144             I=strncmp(<span class="string">'kegg.orthology'</span>,model.rxnMiriams{i}.name,18);
0145             [J, K]=ismember(model.rxnMiriams{i}.value(I),KOModel.rxns);
0146             <span class="comment">%Find all gene indexes that correspond to any of these KOs</span>
0147             [~, L]=find(KOModel.rxnGeneMat(K(J),:));
0148             <span class="keyword">if</span> any(L)
0149                 <span class="comment">%Allocate room for more elements if needed</span>
0150                 <span class="keyword">if</span> counter+numel(L)-1&gt;=numel(r)
0151                     r=[r;zeros(numel(r),1)];
0152                     c=[c;zeros(numel(c),1)];
0153                 <span class="keyword">end</span>
0154                 r(counter:counter+numel(L)-1)=ones(numel(L),1)*i;
0155                 c(counter:counter+numel(L)-1)=L(:);
0156                 counter=counter+numel(L);
0157             <span class="keyword">end</span>
0158         <span class="keyword">end</span>
0159     <span class="keyword">end</span>
0160     
0161     model.rxnGeneMat=sparse(r(1:counter-1),c(1:counter-1),ones(counter-1,1));
0162     <span class="keyword">if</span> size(model.rxnGeneMat,1)~=numel(model.rxns) || size(model.rxnGeneMat,2)~=numel(KOModel.genes)
0163         model.rxnGeneMat(numel(model.rxns),numel(KOModel.genes))=0;
0164     <span class="keyword">end</span>
0165     
0166     <span class="comment">%Then get all metabolites</span>
0167     metModel=<a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>(keggPath);
0168     fprintf(<span class="string">'KEGG metabolites loaded\n'</span>);
0169     
0170     <span class="comment">%Add information about all metabolites to the model</span>
0171     [a, b]=ismember(model.mets,metModel.mets);
0172     a=find(a);
0173     b=b(a);
0174     
0175     <span class="keyword">if</span> ~isfield(model,<span class="string">'metNames'</span>)
0176         model.metNames=cell(numel(model.mets),1);
0177         model.metNames(:)={<span class="string">''</span>};
0178     <span class="keyword">end</span>
0179     model.metNames(a)=metModel.metNames(b);
0180     
0181     <span class="keyword">if</span> ~isfield(model,<span class="string">'metFormulas'</span>)
0182         model.metFormulas=cell(numel(model.mets),1);
0183         model.metFormulas(:)={<span class="string">''</span>};
0184     <span class="keyword">end</span>
0185     model.metFormulas(a)=metModel.metFormulas(b);
0186     
0187     <span class="keyword">if</span> ~isfield(model,<span class="string">'inchis'</span>)
0188         model.inchis=cell(numel(model.mets),1);
0189         model.inchis(:)={<span class="string">''</span>};
0190     <span class="keyword">end</span>
0191     model.inchis(a)=metModel.inchis(b);
0192     
0193     <span class="keyword">if</span> ~isfield(model,<span class="string">'metMiriams'</span>)
0194         model.metMiriams=cell(numel(model.mets),1);
0195     <span class="keyword">end</span>
0196     model.metMiriams(a)=metModel.metMiriams(b);
0197     
0198     <span class="comment">%The composition should be loaded from InChIs when available</span>
0199     I=find(~cellfun(@isempty,model.inchis));
0200     <span class="keyword">for</span> i=1:numel(I)
0201         S=regexp(model.inchis(I(i)),<span class="string">'/'</span>,<span class="string">'split'</span>);
0202         S=S{1};
0203         <span class="keyword">if</span> numel(S)&gt;=2
0204             <span class="comment">%Don't copy if it doesn't look good</span>
0205             model.metFormulas(I(i))=S(2);
0206         <span class="keyword">end</span>
0207     <span class="keyword">end</span>
0208     
0209     <span class="comment">%Put all metabolites in one compartment called 's' (for system). This</span>
0210     <span class="comment">%is done just to be more compatible with the rest of the code</span>
0211     model.comps={<span class="string">'s'</span>};
0212     model.compNames={<span class="string">'System'</span>};
0213     model.metComps=ones(numel(model.mets),1);
0214     
0215     <span class="comment">%If reactions with undefined stoichiometry are kept, then the</span>
0216     <span class="comment">%corresponding metabolites will have ids such as &quot;(n+1) C00001&quot; and</span>
0217     <span class="comment">%their names will be empty. These ids are not valid SBML identifiers</span>
0218     <span class="comment">%and are therefore replaced with &quot;undefined1, undefined2...&quot;. The</span>
0219     <span class="comment">%former ids are stored as the new names</span>
0220     I=find(cellfun(@any,strfind(model.mets,<span class="string">'n'</span>)) | cellfun(@any,strfind(model.mets,<span class="string">'m'</span>)));
0221     model.metNames(I)=model.mets(I);
0222     repNums=1:numel(I);
0223     repIDs=strcat(<span class="string">'undefined_'</span>,cellfun(@num2str,num2cell(repNums(:)),<span class="string">'UniformOutput'</span>,false));
0224     model.mets(I)=repIDs;
0225     
0226     <span class="comment">%It could also be that the metabolite names are empty for some other</span>
0227     <span class="comment">%reason. In that case, use the ID instead</span>
0228     I=cellfun(@isempty,model.metNames);
0229     model.metNames(I)=model.mets(I);
0230     
0231     <span class="comment">%Save the model structure</span>
0232     save(modelFile,<span class="string">'model'</span>,<span class="string">'KOModel'</span>,<span class="string">'isGeneral'</span>,<span class="string">'isIncomplete'</span>,<span class="string">'isUndefinedStoich'</span>,<span class="string">'isSpontaneous'</span>);
0233 <span class="keyword">end</span>
0234 
0235 <span class="comment">%Delete reactions which are labeled as &quot;incomplete&quot;, &quot;erroneous&quot;,</span>
0236 <span class="comment">%&quot;unclear&quot;, &quot;general reaction&quot; or having undefined stoichiometry (depending</span>
0237 <span class="comment">%on settings)</span>
0238 <span class="keyword">if</span> keepSpontaneous==false
0239     model=removeReactions(model,intersect(isSpontaneous,model.rxns),true,true);
0240 <span class="keyword">end</span>
0241 <span class="keyword">if</span> keepUndefinedStoich==false
0242     model=removeReactions(model,intersect(isUndefinedStoich,model.rxns),true,true);
0243 <span class="keyword">end</span>
0244 <span class="keyword">if</span> keepIncomplete==false
0245     model=removeReactions(model,intersect(isIncomplete,model.rxns),true,true);
0246 <span class="keyword">end</span>
0247 <span class="keyword">if</span> keepGeneral==false
0248     model=removeReactions(model,intersect(isGeneral,model.rxns),true,true);
0249 <span class="keyword">end</span>
0250 
0251 <span class="keyword">end</span>
0252 
0253 <a name="_sub1" href="#_subfunctions" class="code">function output = isNewestFile(ravenPath)</a>
0254 <span class="comment">%The ad hoc function, which checks whether keggModel.mat is the more</span>
0255 <span class="comment">%recently modified than keggRxns.mat, keggGenes.mat and keggRxns.mat</span>
0256 modelFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggModel.mat'</span>);
0257 rxnsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggRxns.mat'</span>);
0258 genesFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggGenes.mat'</span>);
0259 metsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggMets.mat'</span>);
0260 <span class="keyword">if</span> (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(rxnsFile))&amp;&amp;<span class="keyword">...</span>
0261         (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(genesFile))&amp;&amp;<span class="keyword">...</span>
0262         (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(metsFile))
0263     output=1;
0264 <span class="keyword">else</span>
0265     output=0;
0266 <span class="keyword">end</span>
0267 <span class="keyword">end</span>
0268 
0269 <a name="_sub2" href="#_subfunctions" class="code">function modTime = getFileTime(fileName)</a>
0270 <span class="comment">%Gets a last modification time for a particular file in datenum format that</span>
0271 <span class="comment">%the numbers could be easily compared for different files</span>
0272 listing = dir(fileName);
0273 assert(numel(listing) == 1, <span class="string">'No such file: %s'</span>, fileName);
0274 modTime = listing.datenum;
0275 format long;
0276 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Thu 27-Sep-2018 11:46:32 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>