<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromKEGG</title>
  <meta name="keywords" content="getModelFromKEGG">
  <meta name="description" content="getModelFromKEGG">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../../index.html">Home</a> &gt;  <a href="../index.html">external</a> &gt; <a href="index.html">kegg</a> &gt; getModelFromKEGG.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../../index.html"><img alt="<" border="0" src="../../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for external\kegg&nbsp;<img alt=">" border="0" src="../../right.png"></a></td></tr></table>-->

<h1>getModelFromKEGG
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>getModelFromKEGG</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="box"><strong>function [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,keepUndefinedStoich,keepIncomplete,keepGeneral) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromKEGG
   Retrieves information stored in KEGG database and generates a model

   Input:
   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat or
                       keggRxns.mat are not in the RAVEN/external/kegg
                       directory, this function will attempt to read data
                       from a local FTP dump of the KEGG database.
                       keggPath is the path to the root of this database
                       (opt, default 'RAVEN/external/kegg'). If
                       keggModel.mat is present in the same directory, the
                       function reads the data from this file and ignores
                       keggGenes.mat, keggMets.mat and keggRxns.mat
   keepSpontaneous     include reactions labeled as &quot;spontaneous&quot; (opt,
                       default true)
   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These
                       will be dealt with as two separate metabolites
                       (opt, default true)
   keepIncomplete      include reactions which have been labelled as
                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,
                       default true)
   keepGeneral         include reactions which have been labelled as
                       &quot;general reaction&quot;. These are reactions on the form
                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore
                       unsuited for modelling purposes. Note that not all
                       reactions have this type of annotation, and the
                       script will therefore not be able to remove all
                       such reactions (opt, default false)

   Output:
   model               a model structure generated from the database. All
                       reactions and the metabolites used in them will be
                       added
   KOModel             a model structure representing the KEGG Orthology
                       ids and their associated genes. The KO ids are
                       saved as reactions

   NOTE: The model output from getModelFromKEGG can be used as template
   for fillGaps. In that case, ensure that the genes and rxnGeneMat fields
   are removed before parsing: model=rmfield(model,'genes'), etc.

   Usage: [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,...
    keepUndefinedStoich,keepIncomplete,keepGeneral)

   Simonas Marcisauskas, 2019-01-08</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>	getGenesFromKEGG</li><li><a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>	getMetsFromKEGG</li><li><a href="getRxnsFromKEGG.html" class="code" title="function [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=getRxnsFromKEGG(keggPath)">getRxnsFromKEGG</a>	getRxnsFromKEGG</li></ul>
This function is called by:
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="getKEGGModelForOrganism.html" class="code" title="function model=getKEGGModelForOrganism(organismID,fastaFile,dataDir,outDir,keepSpontaneous,keepUndefinedStoich,keepIncomplete,keepGeneral,cutOff,minScoreRatioKO,minScoreRatioG,maxPhylDist,nSequences,seqIdentity)">getKEGGModelForOrganism</a>	getKEGGModelForOrganism</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<ul style="list-style-image:url(../../matlabicon.gif)">
<li><a href="#_sub1" class="code">function output = isNewestFile(ravenPath)</a></li><li><a href="#_sub2" class="code">function modTime = getFileTime(fileName)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,</a><span class="keyword">...</span>
0002     keepUndefinedStoich,keepIncomplete,keepGeneral)
0003 <span class="comment">% getModelFromKEGG</span>
0004 <span class="comment">%   Retrieves information stored in KEGG database and generates a model</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   Input:</span>
0007 <span class="comment">%   keggPath            if keggGenes.mat, keggMets.mat, keggPhylDist.mat or</span>
0008 <span class="comment">%                       keggRxns.mat are not in the RAVEN/external/kegg</span>
0009 <span class="comment">%                       directory, this function will attempt to read data</span>
0010 <span class="comment">%                       from a local FTP dump of the KEGG database.</span>
0011 <span class="comment">%                       keggPath is the path to the root of this database</span>
0012 <span class="comment">%                       (opt, default 'RAVEN/external/kegg'). If</span>
0013 <span class="comment">%                       keggModel.mat is present in the same directory, the</span>
0014 <span class="comment">%                       function reads the data from this file and ignores</span>
0015 <span class="comment">%                       keggGenes.mat, keggMets.mat and keggRxns.mat</span>
0016 <span class="comment">%   keepSpontaneous     include reactions labeled as &quot;spontaneous&quot; (opt,</span>
0017 <span class="comment">%                       default true)</span>
0018 <span class="comment">%   keepUndefinedStoich include reactions in the form n A &lt;=&gt; n+1 A. These</span>
0019 <span class="comment">%                       will be dealt with as two separate metabolites</span>
0020 <span class="comment">%                       (opt, default true)</span>
0021 <span class="comment">%   keepIncomplete      include reactions which have been labelled as</span>
0022 <span class="comment">%                       &quot;incomplete&quot;, &quot;erroneous&quot; or &quot;unclear&quot; (opt,</span>
0023 <span class="comment">%                       default true)</span>
0024 <span class="comment">%   keepGeneral         include reactions which have been labelled as</span>
0025 <span class="comment">%                       &quot;general reaction&quot;. These are reactions on the form</span>
0026 <span class="comment">%                       &quot;an aldehyde &lt;=&gt; an alcohol&quot;, and are therefore</span>
0027 <span class="comment">%                       unsuited for modelling purposes. Note that not all</span>
0028 <span class="comment">%                       reactions have this type of annotation, and the</span>
0029 <span class="comment">%                       script will therefore not be able to remove all</span>
0030 <span class="comment">%                       such reactions (opt, default false)</span>
0031 <span class="comment">%</span>
0032 <span class="comment">%   Output:</span>
0033 <span class="comment">%   model               a model structure generated from the database. All</span>
0034 <span class="comment">%                       reactions and the metabolites used in them will be</span>
0035 <span class="comment">%                       added</span>
0036 <span class="comment">%   KOModel             a model structure representing the KEGG Orthology</span>
0037 <span class="comment">%                       ids and their associated genes. The KO ids are</span>
0038 <span class="comment">%                       saved as reactions</span>
0039 <span class="comment">%</span>
0040 <span class="comment">%   NOTE: The model output from getModelFromKEGG can be used as template</span>
0041 <span class="comment">%   for fillGaps. In that case, ensure that the genes and rxnGeneMat fields</span>
0042 <span class="comment">%   are removed before parsing: model=rmfield(model,'genes'), etc.</span>
0043 <span class="comment">%</span>
0044 <span class="comment">%   Usage: [model,KOModel]=getModelFromKEGG(keggPath,keepSpontaneous,...</span>
0045 <span class="comment">%    keepUndefinedStoich,keepIncomplete,keepGeneral)</span>
0046 <span class="comment">%</span>
0047 <span class="comment">%   Simonas Marcisauskas, 2019-01-08</span>
0048 <span class="comment">%</span>
0049 
0050 <span class="keyword">if</span> nargin&lt;1
0051     keggPath=<span class="string">'RAVEN/external/kegg'</span>;
0052 <span class="keyword">end</span>
0053 <span class="keyword">if</span> nargin&lt;2
0054     keepSpontaneous=true;
0055 <span class="keyword">end</span>
0056 <span class="keyword">if</span> nargin&lt;3
0057     keepUndefinedStoich=true;
0058 <span class="keyword">end</span>
0059 <span class="keyword">if</span> nargin&lt;4
0060     keepIncomplete=true;
0061 <span class="keyword">end</span>
0062 <span class="keyword">if</span> nargin&lt;5
0063     keepGeneral=false;
0064 <span class="keyword">end</span>
0065 
0066 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0067 ravenPath=fileparts(fileparts(fileparts(ST(I).file)));
0068 modelFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggModel.mat'</span>);
0069 <span class="keyword">if</span> exist(modelFile, <span class="string">'file'</span>) &amp;&amp; <a href="#_sub1" class="code" title="subfunction output = isNewestFile(ravenPath)">isNewestFile</a>(ravenPath)
0070     fprintf([<span class="string">'NOTE: Importing KEGG model from '</span> strrep(modelFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">'.\n'</span>]);
0071     load(modelFile);
0072 <span class="keyword">else</span>
0073     fprintf([<span class="string">'The file '</span> strrep(modelFile,<span class="string">'\'</span>,<span class="string">'/'</span>) <span class="string">' cannot be located or is older than kegg***.mat files. It will therefore be regenerated.\n'</span>]);
0074     <span class="comment">%First get all reactions</span>
0075     [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=<a href="getRxnsFromKEGG.html" class="code" title="function [model,isSpontaneous,isUndefinedStoich,isIncomplete,isGeneral]=getRxnsFromKEGG(keggPath)">getRxnsFromKEGG</a>(keggPath);
0076     fprintf(<span class="string">'KEGG reactions loaded\n'</span>);
0077     
0078     <span class="comment">%Get the KO ids that are associated with any of the reactions. They</span>
0079     <span class="comment">%will be used later on to create a rxn-gene matrix</span>
0080     KOs=cell(numel(model.rxns)*2,1);
0081     <span class="comment">%Make room for two KO ids per reaction</span>
0082     
0083     addToIndex=1;
0084     <span class="comment">%Add all KO ids that are used</span>
0085     <span class="keyword">for</span> i=1:numel(model.rxns)
0086         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0087             <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0088                 <span class="keyword">if</span> strcmpi(<span class="string">'kegg.orthology'</span>,model.rxnMiriams{i}.name{j})
0089                     <span class="comment">%Add the KO id</span>
0090                     KOs(addToIndex)=model.rxnMiriams{i}.value(j);
0091                     addToIndex=addToIndex+1;
0092                 <span class="keyword">end</span>
0093             <span class="keyword">end</span>
0094         <span class="keyword">end</span>
0095     <span class="keyword">end</span>
0096     
0097     KOs=KOs(1:addToIndex-1);
0098     KOs=unique(KOs);
0099     
0100     <span class="comment">%Get all genes from any organism in KEGG that is associated with any of</span>
0101     <span class="comment">%the KOs</span>
0102     KOModel=<a href="getGenesFromKEGG.html" class="code" title="function model=getGenesFromKEGG(keggPath,koList)">getGenesFromKEGG</a>(keggPath,KOs);
0103     fprintf(<span class="string">'KEGG genes loaded\n'</span>);
0104     
0105     model.genes=KOModel.genes;
0106     
0107     <span class="comment">%It can be that there are KOs from the reactions that have no database</span>
0108     <span class="comment">%entry. These are (as far as I've seen) lumped versions of other KOs</span>
0109     <span class="comment">%and should be removed</span>
0110     KOsToRemove=setdiff(KOs, KOModel.rxns);
0111     
0112     <span class="comment">%Loop through all reactions and delete the KOs that weren't found</span>
0113     <span class="keyword">for</span> i=1:numel(model.rxns)
0114         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0115             <span class="keyword">for</span> j=1:numel(model.rxnMiriams{i}.name)
0116                 toDel=[];
0117                 <span class="keyword">if</span> strcmp(model.rxnMiriams{i}.name{j},<span class="string">'kegg.orthology'</span>)
0118                     <span class="keyword">if</span> ismember(model.rxnMiriams{i}.value{j},KOsToRemove)
0119                         toDel=[toDel;j];
0120                     <span class="keyword">end</span>
0121                 <span class="keyword">end</span>
0122             <span class="keyword">end</span>
0123             <span class="comment">%Delete the KOs</span>
0124             <span class="keyword">if</span> any(toDel)
0125                 <span class="comment">%If all posts are deleted, then delete the whole structure</span>
0126                 <span class="keyword">if</span> numel(toDel)==j
0127                     model.rxnMiriams{i}=[];
0128                 <span class="keyword">else</span>
0129                     model.rxnMiriams{i}.name(toDel)=[];
0130                     model.rxnMiriams{i}.value(toDel)=[];
0131                 <span class="keyword">end</span>
0132             <span class="keyword">end</span>
0133         <span class="keyword">end</span>
0134     <span class="keyword">end</span>
0135     
0136     <span class="comment">%Create the rxnGeneMat for the reactions. This is simply done by</span>
0137     <span class="comment">%merging the gene associations for all the involved KOs</span>
0138     r=zeros(10000000,1);
0139     <span class="comment">%Store the positions since it's slow to write to a sparse array in a</span>
0140     <span class="comment">%loop</span>
0141     c=zeros(10000000,1);
0142     counter=1;
0143     <span class="keyword">for</span> i=1:numel(model.rxns)
0144         <span class="keyword">if</span> isstruct(model.rxnMiriams{i})
0145             I=strncmp(<span class="string">'kegg.orthology'</span>,model.rxnMiriams{i}.name,18);
0146             [J, K]=ismember(model.rxnMiriams{i}.value(I),KOModel.rxns);
0147             <span class="comment">%Find all gene indexes that correspond to any of these KOs</span>
0148             [~, L]=find(KOModel.rxnGeneMat(K(J),:));
0149             <span class="keyword">if</span> any(L)
0150                 <span class="comment">%Allocate room for more elements if needed</span>
0151                 <span class="keyword">if</span> counter+numel(L)-1&gt;=numel(r)
0152                     r=[r;zeros(numel(r),1)];
0153                     c=[c;zeros(numel(c),1)];
0154                 <span class="keyword">end</span>
0155                 r(counter:counter+numel(L)-1)=ones(numel(L),1)*i;
0156                 c(counter:counter+numel(L)-1)=L(:);
0157                 counter=counter+numel(L);
0158             <span class="keyword">end</span>
0159         <span class="keyword">end</span>
0160     <span class="keyword">end</span>
0161     
0162     model.rxnGeneMat=sparse(r(1:counter-1),c(1:counter-1),ones(counter-1,1));
0163     <span class="keyword">if</span> size(model.rxnGeneMat,1)~=numel(model.rxns) || size(model.rxnGeneMat,2)~=numel(KOModel.genes)
0164         model.rxnGeneMat(numel(model.rxns),numel(KOModel.genes))=0;
0165     <span class="keyword">end</span>
0166     
0167     <span class="comment">%Then get all metabolites</span>
0168     metModel=<a href="getMetsFromKEGG.html" class="code" title="function model=getMetsFromKEGG(keggPath)">getMetsFromKEGG</a>(keggPath);
0169     fprintf(<span class="string">'KEGG metabolites loaded\n'</span>);
0170     
0171     <span class="comment">%Add information about all metabolites to the model</span>
0172     [a, b]=ismember(model.mets,metModel.mets);
0173     a=find(a);
0174     b=b(a);
0175     
0176     <span class="keyword">if</span> ~isfield(model,<span class="string">'metNames'</span>)
0177         model.metNames=cell(numel(model.mets),1);
0178         model.metNames(:)={<span class="string">''</span>};
0179     <span class="keyword">end</span>
0180     model.metNames(a)=metModel.metNames(b);
0181     
0182     <span class="keyword">if</span> ~isfield(model,<span class="string">'metFormulas'</span>)
0183         model.metFormulas=cell(numel(model.mets),1);
0184         model.metFormulas(:)={<span class="string">''</span>};
0185     <span class="keyword">end</span>
0186     model.metFormulas(a)=metModel.metFormulas(b);
0187     
0188     <span class="keyword">if</span> ~isfield(model,<span class="string">'inchis'</span>)
0189         model.inchis=cell(numel(model.mets),1);
0190         model.inchis(:)={<span class="string">''</span>};
0191     <span class="keyword">end</span>
0192     model.inchis(a)=metModel.inchis(b);
0193     
0194     <span class="keyword">if</span> ~isfield(model,<span class="string">'metMiriams'</span>)
0195         model.metMiriams=cell(numel(model.mets),1);
0196     <span class="keyword">end</span>
0197     model.metMiriams(a)=metModel.metMiriams(b);
0198     
0199     <span class="comment">%The composition should be loaded from InChIs when available</span>
0200     I=find(~cellfun(@isempty,model.inchis));
0201     <span class="keyword">for</span> i=1:numel(I)
0202         S=regexp(model.inchis(I(i)),<span class="string">'/'</span>,<span class="string">'split'</span>);
0203         S=S{1};
0204         <span class="keyword">if</span> numel(S)&gt;=2
0205             <span class="comment">%Don't copy if it doesn't look good</span>
0206             model.metFormulas(I(i))=S(2);
0207         <span class="keyword">end</span>
0208     <span class="keyword">end</span>
0209     
0210     <span class="comment">%Put all metabolites in one compartment called 's' (for system). This</span>
0211     <span class="comment">%is done just to be more compatible with the rest of the code</span>
0212     model.comps={<span class="string">'s'</span>};
0213     model.compNames={<span class="string">'System'</span>};
0214     model.metComps=ones(numel(model.mets),1);
0215     
0216     <span class="comment">%If reactions with undefined stoichiometry are kept, then the</span>
0217     <span class="comment">%corresponding metabolites will have ids such as &quot;(n+1) C00001&quot; and</span>
0218     <span class="comment">%their names will be empty. These ids are not valid SBML identifiers</span>
0219     <span class="comment">%and are therefore replaced with &quot;undefined1, undefined2...&quot;. The</span>
0220     <span class="comment">%former ids are stored as the new names</span>
0221     I=find(cellfun(@any,strfind(model.mets,<span class="string">'n'</span>)) | cellfun(@any,strfind(model.mets,<span class="string">'m'</span>)));
0222     model.metNames(I)=model.mets(I);
0223     repNums=1:numel(I);
0224     repIDs=strcat(<span class="string">'undefined_'</span>,cellfun(@num2str,num2cell(repNums(:)),<span class="string">'UniformOutput'</span>,false));
0225     model.mets(I)=repIDs;
0226     
0227     <span class="comment">%It could also be that the metabolite names are empty for some other</span>
0228     <span class="comment">%reason. In that case, use the ID instead</span>
0229     I=cellfun(@isempty,model.metNames);
0230     model.metNames(I)=model.mets(I);
0231     
0232     <span class="comment">%Save the model structure</span>
0233     save(modelFile,<span class="string">'model'</span>,<span class="string">'KOModel'</span>,<span class="string">'isGeneral'</span>,<span class="string">'isIncomplete'</span>,<span class="string">'isUndefinedStoich'</span>,<span class="string">'isSpontaneous'</span>);
0234 <span class="keyword">end</span>
0235 
0236 <span class="comment">%Delete reactions which are labeled as &quot;incomplete&quot;, &quot;erroneous&quot;,</span>
0237 <span class="comment">%&quot;unclear&quot;, &quot;general reaction&quot; or having undefined stoichiometry (depending</span>
0238 <span class="comment">%on settings)</span>
0239 <span class="keyword">if</span> keepSpontaneous==false
0240     model=removeReactions(model,intersect(isSpontaneous,model.rxns),true,true);
0241 <span class="keyword">end</span>
0242 <span class="keyword">if</span> keepUndefinedStoich==false
0243     model=removeReactions(model,intersect(isUndefinedStoich,model.rxns),true,true);
0244 <span class="keyword">end</span>
0245 <span class="keyword">if</span> keepIncomplete==false
0246     model=removeReactions(model,intersect(isIncomplete,model.rxns),true,true);
0247 <span class="keyword">end</span>
0248 <span class="keyword">if</span> keepGeneral==false
0249     model=removeReactions(model,intersect(isGeneral,model.rxns),true,true);
0250 <span class="keyword">end</span>
0251 
0252 <span class="keyword">end</span>
0253 
0254 <a name="_sub1" href="#_subfunctions" class="code">function output = isNewestFile(ravenPath)</a>
0255 <span class="comment">%The ad hoc function, which checks whether keggModel.mat is the more</span>
0256 <span class="comment">%recently modified than keggRxns.mat, keggGenes.mat and keggRxns.mat</span>
0257 modelFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggModel.mat'</span>);
0258 rxnsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggRxns.mat'</span>);
0259 genesFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggGenes.mat'</span>);
0260 metsFile=fullfile(ravenPath,<span class="string">'external'</span>,<span class="string">'kegg'</span>,<span class="string">'keggMets.mat'</span>);
0261 <span class="keyword">if</span> (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(rxnsFile))&amp;&amp;<span class="keyword">...</span>
0262         (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(genesFile))&amp;&amp;<span class="keyword">...</span>
0263         (<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(modelFile)&gt;<a href="#_sub2" class="code" title="subfunction modTime = getFileTime(fileName)">getFileTime</a>(metsFile))
0264     output=1;
0265 <span class="keyword">else</span>
0266     output=0;
0267 <span class="keyword">end</span>
0268 <span class="keyword">end</span>
0269 
0270 <a name="_sub2" href="#_subfunctions" class="code">function modTime = getFileTime(fileName)</a>
0271 <span class="comment">%Gets a last modification time for a particular file in datenum format that</span>
0272 <span class="comment">%the numbers could be easily compared for different files</span>
0273 listing = dir(fileName);
0274 assert(numel(listing) == 1, <span class="string">'No such file: %s'</span>, fileName);
0275 modTime = listing.datenum;
0276 format long;
0277 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Sun 03-Feb-2019 21:32:08 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>