<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of tutorial6</title>
  <meta name="keywords" content="tutorial6">
  <meta name="description" content="tutorial6">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">tutorial</a> &gt; tutorial6.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for tutorial&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>tutorial6
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>tutorial6</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>This is a script file. </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> tutorial6
   This exercise demonstrates how to reconstruct a combined draft GEM by
   from KEGG and MetaCyc pathway databases. A combined model with the
   comprehensive coverage of metabolic pathways is generated from
   different de novo reconstruction approaches. The input is a FASTA
   format file with whole-proteome sequences. The combined model is
   subsequently used for refinement of existing high-quality model and
   generation of a new version of GEM, by utilizing the manual curation
   results. This tutorial is a showcase of the new features released in
   RAVEN 2.0 through demonstrating the utilization of the newly developed
   functions on GEM reconstruction and curation for Streptomyces
   coelicolor strain A3(2). Users may apply this script as the template in
   their own work for other organisms.
   This refers to Tutorial 6 from &quot;RAVEN tutorials.docx&quot;</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <span class="comment">% tutorial6</span>
0002 <span class="comment">%   This exercise demonstrates how to reconstruct a combined draft GEM by</span>
0003 <span class="comment">%   from KEGG and MetaCyc pathway databases. A combined model with the</span>
0004 <span class="comment">%   comprehensive coverage of metabolic pathways is generated from</span>
0005 <span class="comment">%   different de novo reconstruction approaches. The input is a FASTA</span>
0006 <span class="comment">%   format file with whole-proteome sequences. The combined model is</span>
0007 <span class="comment">%   subsequently used for refinement of existing high-quality model and</span>
0008 <span class="comment">%   generation of a new version of GEM, by utilizing the manual curation</span>
0009 <span class="comment">%   results. This tutorial is a showcase of the new features released in</span>
0010 <span class="comment">%   RAVEN 2.0 through demonstrating the utilization of the newly developed</span>
0011 <span class="comment">%   functions on GEM reconstruction and curation for Streptomyces</span>
0012 <span class="comment">%   coelicolor strain A3(2). Users may apply this script as the template in</span>
0013 <span class="comment">%   their own work for other organisms.</span>
0014 <span class="comment">%   This refers to Tutorial 6 from &quot;RAVEN tutorials.docx&quot;</span>
0015 
0016 <span class="comment">%Before reconstruction, a FASTA file with protein sequences of the target</span>
0017 <span class="comment">%organism needs to be prepared. In this tutorial, all protein sequences of</span>
0018 <span class="comment">%S. coelicolor A3(2) were downloaded from the NCBI genome database and</span>
0019 <span class="comment">%provided in current folder (Sco_all_protein.faa). The description lines</span>
0020 <span class="comment">%(starting with &quot;&gt;&quot; character) of this FASTA file were modified by keeping</span>
0021 <span class="comment">%only the locus tag (e.g. SCO6005), which is commonly used as the gene</span>
0022 <span class="comment">%names in GEMs.</span>
0023 
0024 <span class="comment">%Generate a draft GEM from MetaCyc database. The first parameter is</span>
0025 <span class="comment">%organismID and needs to be specified by user. The other parameters are set</span>
0026 <span class="comment">%to exclude unbalanced and undetermined reactions, but keep transport</span>
0027 <span class="comment">%reactions. The two parameters for homology search are set to the default</span>
0028 <span class="comment">%values that have been optimized to capture the protein hits with both</span>
0029 <span class="comment">%comprehensive coverage and the least false positives.</span>
0030 ScoMetaCycDraftModel=getMetaCycModelForOrganism(<span class="string">'ScoMetaCyc'</span>,<span class="string">'Sco_all_protein.faa'</span>,1);
0031 
0032 <span class="comment">%Generate the two draft models from KEGG database. The first one is</span>
0033 <span class="comment">%reconstructed from the genome information annotated by KEGG. Since KEGG</span>
0034 <span class="comment">%provides genome annotations for over 5000 organisms, their GEMs can thus</span>
0035 <span class="comment">%be generated without providing protein sequences while only using the</span>
0036 <span class="comment">%organismID. The organisms with KEGG annotation and their associated ids</span>
0037 <span class="comment">%(e.g. 'sco' for Streptomyces coelicolor) can be found from here:</span>
0038 <span class="comment">%https://www.kegg.jp/kegg/catalog/org_list.html.</span>
0039 <span class="comment">%Generate a draft model using KEGG annotation and exclude incomplete</span>
0040 <span class="comment">%reactions and reactions with undefined stoichiometry</span>
0041 ScoKEGGAnnotation=getKEGGModelForOrganism(<span class="string">'sco'</span>,<span class="string">''</span>,<span class="string">''</span>,<span class="string">''</span>,1,0,0);
0042 
0043 <span class="comment">%Generate the second KEGG-based draft model based on sequence homology to</span>
0044 <span class="comment">%KEGG Ortholog sequence clusters while excluding incomplete reactions and</span>
0045 <span class="comment">%the ones with undefined stoichiometry. Type &quot;help getKEGGModelForOrganism&quot;</span>
0046 <span class="comment">%to see the detailed instructions for the choice of different parameters.</span>
0047 <span class="comment">%The default values for homology search are used because they have been</span>
0048 <span class="comment">%optimized for the best performance.</span>
0049 ScoKEGGHomology=getKEGGModelForOrganism(<span class="string">'ScoKEGGHMMs'</span>,<span class="string">'Sco_all_protein.faa'</span>,<span class="string">'prok90_kegg105'</span>,<span class="string">''</span>,1,0,0);
0050 
0051 <span class="comment">%De novo reconstruction from MetaCyc should take about 10 minutes, while</span>
0052 <span class="comment">%both reconstructions from KEGG may take up to 50-60 minutes in overall</span>
0053 <span class="comment">%depending on the computer hardware and the system used. Given that KEGG</span>
0054 <span class="comment">%and MetaCyc databases are formulated in different ways; KEGG relies on</span>
0055 <span class="comment">%sequence-based annotation, while MetaCyc collects only experimentally</span>
0056 <span class="comment">%verified pathways. Therefore, integration of MetaCyc- and KEGG-based draft</span>
0057 <span class="comment">%models could have a better coverage of the metbolism for the target</span>
0058 <span class="comment">%organism.</span>
0059 
0060 <span class="comment">%At first, the two KEGG-based models can be directly merged</span>
0061 ScoKEGGDraftModel=mergeModels({ScoKEGGAnnotation ScoKEGGHomology});
0062 
0063 disp(numel(ScoKEGGHomology.rxns)+numel(ScoKEGGAnnotation.rxns));
0064 disp(numel(ScoKEGGDraftModel.rxns));
0065 <span class="comment">%By checking the reaction number, it can be seen that reaction number in</span>
0066 <span class="comment">%the merged model equals adding up the reaction numbers in homology and</span>
0067 <span class="comment">%annotation KEGG draft models. And there are duplicated reactions in this</span>
0068 <span class="comment">%merged model.</span>
0069 
0070 <span class="comment">%Merge the duplicated reactions into one and combine multiple iso-enzymes</span>
0071 <span class="comment">%into a single grRules. The expandModel and contractModel functions are</span>
0072 <span class="comment">%utilised, see their instructions for details.</span>
0073 ScoKEGGDraftModel=expandModel(ScoKEGGDraftModel);
0074 ScoKEGGDraftModel=contractModel(ScoKEGGDraftModel);
0075 
0076 <span class="comment">%In the end, KEGG- and MetaCyc-based draft models can be further combined</span>
0077 <span class="comment">%into an integrated GEM. This step is achieved by the function</span>
0078 <span class="comment">%combineMetaCycKEGGModels, which firstly converts metabolite and reaction</span>
0079 <span class="comment">%identifiers in the KEGG model into corresponding MetaCyc ids, and then</span>
0080 <span class="comment">%detects duplications and keeps only unique reactions and metabolites that</span>
0081 <span class="comment">%are mostly in MetaCyc identifiers.</span>
0082 ScoCombinedDraftModel=combineMetaCycKEGGModels(ScoMetaCycDraftModel, ScoKEGGDraftModel);
0083 
0084 <span class="comment">%With this combined model, the existing iMK1208 model is refined by</span>
0085 <span class="comment">%incorporating the new pathways/reactions that are found in the combined</span>
0086 <span class="comment">%model but absent from the previous iMK1208 model. At the time of</span>
0087 <span class="comment">%publication, a total of 398 reactions in the combined draft model were</span>
0088 <span class="comment">%determined as new pathways based on manual curation results, which have</span>
0089 <span class="comment">%been organized into TableS3. Load these manually selected reactions and</span>
0090 <span class="comment">%their subSystems as the pre-processed array structure selectedNewRxns.</span>
0091 load(<span class="string">'iMK1208+suppInfo.mat'</span>,<span class="string">'selectedNewRxns'</span>);
0092 disp(selectedNewRxns);
0093 
0094 <span class="comment">%Generate a sub-model that includes only these new reactions. This is</span>
0095 <span class="comment">%implemented by subtracting the other reactions from the combined model</span>
0096 <span class="comment">%using function removeReactions</span>
0097 rxnsToRemove=setdiff(ScoCombinedDraftModel.rxns,selectedNewRxns.rxns);
0098 newRxnSubModel=removeReactions(ScoCombinedDraftModel,rxnsToRemove,1,1);
0099 
0100 <span class="comment">%Now this newRxnSubModel contains only these new reactions. It needs to be</span>
0101 <span class="comment">%modified before merging with the iMK1208. Since these reactions are</span>
0102 <span class="comment">%metabolic ones and can all be assigned to the cytoplasm compartment</span>
0103 newRxnSubModel.comps{1}=<span class="string">'c'</span>;
0104 newRxnSubModel.compNames{1}=<span class="string">'Cytoplasm'</span>;
0105 
0106 <span class="comment">%Amend to the genes and rxnGeneMat fields</span>
0107 newRxnSubModel.genes={};
0108 <span class="keyword">for</span> k=1:numel(newRxnSubModel.rxns)
0109    newRxnSubModel.genes=[newRxnSubModel.genes;transpose(strsplit(newRxnSubModel.grRules{k},<span class="string">' or '</span>))];
0110 <span class="keyword">end</span>
0111 newRxnSubModel.genes=unique(newRxnSubModel.genes);
0112 [~, newRxnSubModel.rxnGeneMat, ~]=standardizeGrRules(newRxnSubModel);
0113 
0114 <span class="comment">%Since the newRxnSubModel is ready for incorportation, the iMK1208 model</span>
0115 <span class="comment">%can be loaded for integration</span>
0116 load(<span class="string">'iMK1208+suppInfo.mat'</span>,<span class="string">'iMK1208'</span>);
0117 
0118 <span class="comment">%It should be noted that the incompatible nomenclatures (especially the</span>
0119 <span class="comment">%metabolite identifiers) used in different GEMs and databases led to a</span>
0120 <span class="comment">%serious problem in model comparison, curation and integration. In order to</span>
0121 <span class="comment">%properly integrate the iMK1208 with this sub-model, both models should be</span>
0122 <span class="comment">%unified into the same name space, since the metabolites in the sub-model</span>
0123 <span class="comment">%use MetaCyc identifiers that are different from the ones used in iMK1208</span>
0124 <span class="comment">%model. TableS2 contains the results for database mining and intensive</span>
0125 <span class="comment">%manual curation while associating these metabolite identifiers. Load</span>
0126 <span class="comment">%these manually selected reactions and their subSystems as the</span>
0127 <span class="comment">%pre-processed array structure metaCycMetsIniMK.</span>
0128 load(<span class="string">'iMK1208+suppInfo.mat'</span>,<span class="string">'metaCycMetsIniMK'</span>);
0129 
0130 <span class="comment">%Replace the metabolites in the sub-model with the identifiers and names</span>
0131 <span class="comment">%used in iMK1208 according to the mapping information in TableS2</span>
0132 <span class="comment">%(metaCycMetsIniMK).</span>
0133 [a, b]=ismember(newRxnSubModel.mets,metaCycMetsIniMK);
0134 I=find(a);
0135 newRxnSubModel.mets=strcat(newRxnSubModel.mets,<span class="string">'_c'</span>);
0136 newRxnSubModel.metNames(I)=iMK1208.metNames(b(I));
0137 newRxnSubModel.mets(I)=iMK1208.mets(b(I));
0138 
0139 <span class="comment">%The necessary issues for merging iMK1208 with the new reactions determined</span>
0140 <span class="comment">%from de novo generated draft GEMs have been resolved. Those can be</span>
0141 <span class="comment">%directly merged into the Sco4 model, which refers to the 4th major update</span>
0142 <span class="comment">%of the GEM for Streptomyces coelicolor.</span>
0143 Sco4=mergeModels({iMK1208 newRxnSubModel});
0144 
0145 <span class="comment">%Check out if the newly generated Sco4 could grow</span>
0146 sol=solveLP(Sco4);
0147 disp(sol.f);
0148 <span class="comment">%The results indicate that this new model is functional</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>