<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of importExcelModel</title>
  <meta name="keywords" content="importExcelModel">
  <meta name="description" content="importExcelModel">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">io</a> &gt; importExcelModel.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for io&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>importExcelModel
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>importExcelModel</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function model=importExcelModel(fileName,removeExcMets,printWarnings,ignoreErrors) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> importExcelModel
   Imports a constraint-based model from a Excel file

   fileName      a Microsoft Excel file to import
   removeExcMets true if exchange metabolites should be removed. This is
                 needed to be able to run simulations, but it could also 
                 be done using simplifyModel at a later stage (opt,
                 default true)
   printWarnings true if warnings should be printed (opt, default true)
   ignoreErrors  true if errors should be ignored. See below for details
                 (opt, default false)

   model
       description      description of model contents
       id               model ID
       rxns             reaction ids
       mets             metabolite ids
       S                stoichiometric matrix
       lb               lower bounds
       ub               upper bounds
       rev              reversibility vector
       c                objective coefficients
       b                equality constraints for the metabolite equations
       comps            compartment ids
       compNames        compartment names
       compOutside      the id (as in comps) for the compartment
                        surrounding each of the compartments
       compMiriams      structure with MIRIAM information about the
                        compartments
       rxnNames         reaction description
       rxnComps         compartments for reactions
       grRules          reaction to gene rules in text form
       rxnGeneMat       reaction-to-gene mapping in sparse matrix form
       subSystems       subsystem name for each reaction
       eccodes          EC-codes for the reactions
       rxnMiriams       structure with MIRIAM information about the reactions
       genes            list of all genes
       geneComps        compartments for reactions
       geneMiriams      structure with MIRIAM information about the genes
       metNames         metabolite description
       metComps         compartments for metabolites
       inchis           InChI-codes for metabolites
       metFormulas      metabolite chemical formula
       metMiriams       structure with MIRIAM information about the metabolites
       unconstrained    true if the metabolite is an exchange metabolite

   Loads models in the RAVEN Toolbox Excel format. A number of consistency
   checks are performed in order to ensure that the model is valid. These
   can be ignored by putting ignoreErrors to true. However, this is highly
   advised against, as it can result in errors in simulations or other
   functionalities. The RAVEN Toolbox is made to function only on consistent
   models, and the only checks performed are when the model is imported.

   NOTE: Most errors are checked for by checkModelStruct, but some
   are checked for in this function as well. Those are ones which relate
   to missing model elements and so on, and which would make it impossible
   to construct the model structure. Those errors cannot be ignored by
   setting ignoreErrors to true.

   Usage: model=importExcelModel(fileName,removeExcMets,printWarnings,ignoreErrors)

   Rasmus Agren, 2014-01-06</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="../core/checkModelStruct.html" class="code" title="function checkModelStruct(model,throwErrors,trimWarnings)">checkModelStruct</a>	checkModelStruct</li><li><a href="../core/constructS.html" class="code" title="function [S mets badRxns reversible]=constructS(equations,mets,rxns)">constructS</a>	constructS</li><li><a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li><li><a href="../core/parseRxnEqu.html" class="code" title="function metabolites=parseRxnEqu(equations)">parseRxnEqu</a>	parseRxnEqu</li><li><a href="../core/simplifyModel.html" class="code" title="function [reducedModel, deletedReactions, deletedMetabolites]=simplifyModel(model,deleteUnconstrained, deleteDuplicates, deleteZeroInterval, deleteInaccessible, deleteMinMax, groupLinear, reservedRxns, suppressWarnings)">simplifyModel</a>	simplifyModel</li><li><a href="getFullPath.html" class="code" title="function File = getFullPath(File, Style)">getFullPath</a>	getFullPath - Get absolute canonical path of a file or folder</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="../installation/checkInstallation.html" class="code" title="function checkInstallation()">checkInstallation</a>	checkInstallation</li><li><a href="SBMLFromExcel.html" class="code" title="function SBMLFromExcel(fileName, outputFileName,toCOBRA,printWarnings)">SBMLFromExcel</a>	SBMLFromExcel</li><li><a href="../tutorial/tutorial0.html" class="code" title="">tutorial0</a>	****RAVEN TOOLBOX MINI TUTORIAL***</li><li><a href="../tutorial/tutorial1.html" class="code" title="">tutorial1</a>	This contains the code necessary for importing a model from Excel and into</li><li><a href="../tutorial/tutorial1_solutions.html" class="code" title="">tutorial1_solutions</a>	This is the solution to tutorial 1. The file small.xlsx contains all the</li><li><a href="../tutorial/tutorial2.html" class="code" title="">tutorial2</a>	This contains the code necessary for running Exercises 2. It is assumed</li><li><a href="../tutorial/tutorial2_solutions.html" class="code" title="">tutorial2_solutions</a>	This is the solution to tutorial 2. All parameters are set in this file,</li><li><a href="../tutorial/tutorial3.html" class="code" title="">tutorial3</a>	This contains the code necessary for running Exercises 3. It is assumed</li><li><a href="../tutorial/tutorial3_solutions.html" class="code" title="">tutorial3_solutions</a>	This contains the code necessary for running Exercises 3. It is assumed</li></ul>
<!-- crossreference -->

<h2><a name="_subfunctions"></a>SUBFUNCTIONS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="#_sub1" class="code">function raw=cleanImported(raw)</a></li><li><a href="#_sub2" class="code">function I=wrapperWS(A)</a></li><li><a href="#_sub3" class="code">function miriamStruct=parseMiriam(strings,miriamStruct)</a></li><li><a href="#_sub4" class="code">function [raw, flag]=loadSheet(workbook, sheet)</a></li><li><a href="#_sub5" class="code">function y=toStr(x)</a></li><li><a href="#_sub6" class="code">function y=toDouble(x,default)</a></li><li><a href="#_sub7" class="code">function y=boolToDouble(x)</a></li></ul>

<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function model=importExcelModel(fileName,removeExcMets,printWarnings,ignoreErrors)</a>
0002 <span class="comment">% importExcelModel</span>
0003 <span class="comment">%   Imports a constraint-based model from a Excel file</span>
0004 <span class="comment">%</span>
0005 <span class="comment">%   fileName      a Microsoft Excel file to import</span>
0006 <span class="comment">%   removeExcMets true if exchange metabolites should be removed. This is</span>
0007 <span class="comment">%                 needed to be able to run simulations, but it could also</span>
0008 <span class="comment">%                 be done using simplifyModel at a later stage (opt,</span>
0009 <span class="comment">%                 default true)</span>
0010 <span class="comment">%   printWarnings true if warnings should be printed (opt, default true)</span>
0011 <span class="comment">%   ignoreErrors  true if errors should be ignored. See below for details</span>
0012 <span class="comment">%                 (opt, default false)</span>
0013 <span class="comment">%</span>
0014 <span class="comment">%   model</span>
0015 <span class="comment">%       description      description of model contents</span>
0016 <span class="comment">%       id               model ID</span>
0017 <span class="comment">%       rxns             reaction ids</span>
0018 <span class="comment">%       mets             metabolite ids</span>
0019 <span class="comment">%       S                stoichiometric matrix</span>
0020 <span class="comment">%       lb               lower bounds</span>
0021 <span class="comment">%       ub               upper bounds</span>
0022 <span class="comment">%       rev              reversibility vector</span>
0023 <span class="comment">%       c                objective coefficients</span>
0024 <span class="comment">%       b                equality constraints for the metabolite equations</span>
0025 <span class="comment">%       comps            compartment ids</span>
0026 <span class="comment">%       compNames        compartment names</span>
0027 <span class="comment">%       compOutside      the id (as in comps) for the compartment</span>
0028 <span class="comment">%                        surrounding each of the compartments</span>
0029 <span class="comment">%       compMiriams      structure with MIRIAM information about the</span>
0030 <span class="comment">%                        compartments</span>
0031 <span class="comment">%       rxnNames         reaction description</span>
0032 <span class="comment">%       rxnComps         compartments for reactions</span>
0033 <span class="comment">%       grRules          reaction to gene rules in text form</span>
0034 <span class="comment">%       rxnGeneMat       reaction-to-gene mapping in sparse matrix form</span>
0035 <span class="comment">%       subSystems       subsystem name for each reaction</span>
0036 <span class="comment">%       eccodes          EC-codes for the reactions</span>
0037 <span class="comment">%       rxnMiriams       structure with MIRIAM information about the reactions</span>
0038 <span class="comment">%       genes            list of all genes</span>
0039 <span class="comment">%       geneComps        compartments for reactions</span>
0040 <span class="comment">%       geneMiriams      structure with MIRIAM information about the genes</span>
0041 <span class="comment">%       metNames         metabolite description</span>
0042 <span class="comment">%       metComps         compartments for metabolites</span>
0043 <span class="comment">%       inchis           InChI-codes for metabolites</span>
0044 <span class="comment">%       metFormulas      metabolite chemical formula</span>
0045 <span class="comment">%       metMiriams       structure with MIRIAM information about the metabolites</span>
0046 <span class="comment">%       unconstrained    true if the metabolite is an exchange metabolite</span>
0047 <span class="comment">%</span>
0048 <span class="comment">%   Loads models in the RAVEN Toolbox Excel format. A number of consistency</span>
0049 <span class="comment">%   checks are performed in order to ensure that the model is valid. These</span>
0050 <span class="comment">%   can be ignored by putting ignoreErrors to true. However, this is highly</span>
0051 <span class="comment">%   advised against, as it can result in errors in simulations or other</span>
0052 <span class="comment">%   functionalities. The RAVEN Toolbox is made to function only on consistent</span>
0053 <span class="comment">%   models, and the only checks performed are when the model is imported.</span>
0054 <span class="comment">%</span>
0055 <span class="comment">%   NOTE: Most errors are checked for by checkModelStruct, but some</span>
0056 <span class="comment">%   are checked for in this function as well. Those are ones which relate</span>
0057 <span class="comment">%   to missing model elements and so on, and which would make it impossible</span>
0058 <span class="comment">%   to construct the model structure. Those errors cannot be ignored by</span>
0059 <span class="comment">%   setting ignoreErrors to true.</span>
0060 <span class="comment">%</span>
0061 <span class="comment">%   Usage: model=importExcelModel(fileName,removeExcMets,printWarnings,ignoreErrors)</span>
0062 <span class="comment">%</span>
0063 <span class="comment">%   Rasmus Agren, 2014-01-06</span>
0064 <span class="comment">%</span>
0065 
0066 <span class="comment">%Adds the required classes to the Java path</span>
0067 [ST, I]=dbstack(<span class="string">'-completenames'</span>);
0068 ravenPath=fileparts(fileparts(ST(I).file));
0069 poiPATH=fullfile(ravenPath,<span class="string">'software'</span>,<span class="string">'apache-poi'</span>);
0070 javaaddpath(fullfile(poiPATH,<span class="string">'dom4j-1.6.1.jar'</span>));
0071 javaaddpath(fullfile(poiPATH,<span class="string">'poi-3.8-20120326.jar'</span>));
0072 javaaddpath(fullfile(poiPATH,<span class="string">'poi-ooxml-3.8-20120326.jar'</span>));
0073 javaaddpath(fullfile(poiPATH,<span class="string">'poi-ooxml-schemas-3.8-20120326.jar'</span>));
0074 javaaddpath(fullfile(poiPATH,<span class="string">'xmlbeans-2.3.0.jar'</span>));
0075 
0076 <span class="comment">%Import required classes from Apache POI</span>
0077 import org.apache.poi.ss.usermodel.*;
0078 import org.apache.poi.hssf.usermodel.*;
0079 import org.apache.poi.xssf.usermodel.*;
0080 import org.apache.poi.ss.usermodel.*;
0081 import org.apache.poi.ss.util.*;
0082 import java.io.FileInputStream;
0083 
0084 <span class="keyword">if</span> nargin&lt;2
0085     removeExcMets=true;
0086 <span class="keyword">end</span>
0087 <span class="keyword">if</span> nargin&lt;3
0088     printWarnings=true;
0089 <span class="keyword">end</span>
0090 <span class="keyword">if</span> nargin&lt;4
0091     ignoreErrors=false;
0092 <span class="keyword">end</span>
0093 
0094 <span class="comment">%This is to match the order of the fields to those you get from importing</span>
0095 <span class="comment">%from SBML</span>
0096 model=[];
0097 model.id=[];
0098 model.description=[];
0099 model.annotation=[];
0100 <span class="comment">%Default bounds if not defined</span>
0101 model.annotation.defaultLB=-1000;
0102 model.annotation.defaultUB=1000;
0103 model.rxns={};
0104 model.mets={};
0105 model.S=[];
0106 model.lb=[];
0107 model.ub=[];
0108 model.rev=[];
0109 model.c=[];
0110 model.b=[];
0111 model.comps={};
0112 model.compNames={};
0113 model.compOutside={};
0114 model.compMiriams={};
0115 model.rxnNames={};
0116 model.rxnComps={}; <span class="comment">%Will be double later</span>
0117 model.grRules={};
0118 model.rxnGeneMat=[];
0119 model.subSystems={};
0120 model.eccodes={};
0121 model.rxnMiriams={};
0122 model.genes={};
0123 model.geneComps={}; <span class="comment">%Will be double later</span>
0124 model.geneMiriams={};
0125 model.metNames={};
0126 model.metComps=[];
0127 model.inchis={};
0128 model.metFormulas={};
0129 model.metMiriams={};
0130 model.unconstrained=[];
0131     
0132 <span class="comment">%Check if the file exists</span>
0133 <span class="keyword">if</span> ~exist(fileName,<span class="string">'file'</span>)
0134     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The Excel file could not be found'</span>);
0135 <span class="keyword">end</span>
0136 
0137 <span class="comment">%Opens the workbook. The input stream is needed since it will otherwise</span>
0138 <span class="comment">%keep the file open</span>
0139 is=FileInputStream(<a href="getFullPath.html" class="code" title="function File = getFullPath(File, Style)">getFullPath</a>(fileName));
0140 workbook=WorkbookFactory.create(is);
0141 is.close();
0142 
0143 [raw, flag]=<a href="#_sub4" class="code" title="subfunction [raw, flag]=loadSheet(workbook, sheet)">loadSheet</a>(workbook,<span class="string">'MODEL'</span>);
0144 
0145 <span class="keyword">if</span> flag&lt;0
0146     <span class="keyword">if</span> printWarnings==true
0147         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'Could not load the MODEL sheet'</span>,false);
0148     <span class="keyword">end</span>
0149     model.id=<span class="string">'UNKNOWN'</span>;
0150     model.description=<span class="string">'No model details available'</span>;
0151 <span class="keyword">else</span>
0152     raw=<a href="#_sub1" class="code" title="subfunction raw=cleanImported(raw)">cleanImported</a>(raw);
0153 
0154     <span class="comment">%It is assumed that the first line is labels and that the second one is</span>
0155     <span class="comment">%info</span>
0156     allLabels={<span class="string">'ID'</span>;<span class="string">'DESCRIPTION'</span>;<span class="string">'DEFAULT LOWER'</span>;<span class="string">'DEFAULT UPPER'</span>;<span class="string">'CONTACT GIVEN NAME'</span>;<span class="string">'CONTACT FAMILY NAME'</span>;<span class="string">'CONTACT EMAIL'</span>;<span class="string">'ORGANIZATION'</span>;<span class="string">'TAXONOMY'</span>;<span class="string">'NOTES'</span>};
0157 
0158     <span class="comment">%Map to new captions</span>
0159     raw(1,:)=upper(raw(1,:));
0160     raw(1,:)=strrep(raw(1,:),<span class="string">'MODELID'</span>,<span class="string">'ID'</span>);
0161     raw(1,:)=strrep(raw(1,:),<span class="string">'MODELNAME'</span>,<span class="string">'DESCRIPTION'</span>);
0162 
0163     <span class="comment">%Loop through the labels</span>
0164     [I, J]=ismember(raw(1,:),allLabels);
0165     I=find(I);
0166     <span class="keyword">for</span> i=1:numel(I)
0167         <span class="keyword">switch</span> J(I(i))
0168             <span class="keyword">case</span> 1
0169                 <span class="keyword">if</span> any(raw{I(i),2})
0170                     model.id=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0171                 <span class="keyword">else</span>
0172                     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'No model ID supplied'</span>);
0173                 <span class="keyword">end</span>
0174             <span class="keyword">case</span> 2
0175                 <span class="keyword">if</span> any(raw{2,I(i)})
0176                     model.description=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0177                 <span class="keyword">else</span>
0178                     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'No model name supplied'</span>);
0179                 <span class="keyword">end</span>
0180             <span class="keyword">case</span> 3
0181                 <span class="keyword">if</span> ~isempty(raw{2,I(i)})
0182                     <span class="keyword">try</span>
0183                         model.annotation.defaultLB=<a href="#_sub6" class="code" title="subfunction y=toDouble(x,default)">toDouble</a>(raw{2,I(i)},NaN);
0184                     <span class="keyword">catch</span>
0185                         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'DEFAULT LOWER must be numeric'</span>);
0186                     <span class="keyword">end</span>
0187                 <span class="keyword">else</span>
0188                     <span class="keyword">if</span> printWarnings==true
0189                         fprintf(<span class="string">'NOTE: DEFAULT LOWER not supplied. Uses -1000\n'</span>);
0190                     <span class="keyword">end</span>
0191                     model.annotation.defaultLB=-1000;
0192                 <span class="keyword">end</span>
0193             <span class="keyword">case</span> 4
0194                 <span class="keyword">if</span> ~isempty(raw{2,I(i)})
0195                     <span class="keyword">try</span>
0196                         model.annotation.defaultUB=<a href="#_sub6" class="code" title="subfunction y=toDouble(x,default)">toDouble</a>(raw{2,I(i)},NaN);
0197                     <span class="keyword">catch</span>
0198                         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'DEFAULT UPPER must be numeric'</span>);
0199                     <span class="keyword">end</span>
0200                 <span class="keyword">else</span>
0201                     <span class="keyword">if</span> printWarnings==true
0202                         fprintf(<span class="string">'NOTE: DEFAULT UPPER not supplied. Uses 1000\n'</span>);
0203                     <span class="keyword">end</span>
0204                     model.annotation.defaultUB=1000;
0205                 <span class="keyword">end</span>
0206             <span class="keyword">case</span> 5
0207                 <span class="keyword">if</span> any(raw{2,I(i)})
0208                     model.annotation.givenName=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0209                 <span class="keyword">end</span>
0210             <span class="keyword">case</span> 6
0211                 <span class="keyword">if</span> any(raw{2,I(i)})
0212                     model.annotation.familyName=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0213                 <span class="keyword">end</span>
0214             <span class="keyword">case</span> 7
0215                 <span class="keyword">if</span> any(raw{2,I(i)})
0216                     model.annotation.email=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0217                 <span class="keyword">end</span>
0218             <span class="keyword">case</span> 8
0219                 <span class="keyword">if</span> any(raw{2,I(i)})
0220                     model.annotation.organization=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0221                 <span class="keyword">end</span>    
0222             <span class="keyword">case</span> 9
0223                 <span class="keyword">if</span> any(raw{2,I(i)})
0224                     model.annotation.taxonomy=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0225                 <span class="keyword">end</span> 
0226             <span class="keyword">case</span> 10
0227                 <span class="keyword">if</span> any(raw{2,I(i)})
0228                     model.annotation.note=<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>(raw{2,I(i)}); <span class="comment">%Should be string already</span>
0229                 <span class="keyword">end</span>     
0230         <span class="keyword">end</span>  
0231     <span class="keyword">end</span>
0232 <span class="keyword">end</span>
0233 
0234 <span class="comment">%Get compartment information</span>
0235 [raw, flag]=<a href="#_sub4" class="code" title="subfunction [raw, flag]=loadSheet(workbook, sheet)">loadSheet</a>(workbook,<span class="string">'COMPS'</span>);
0236 
0237 <span class="keyword">if</span> flag&lt;0
0238     <span class="keyword">if</span> printWarnings==true
0239         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'Could not load the COMPS sheet. All elements will be assigned to a compartment &quot;s&quot; for &quot;System&quot;'</span>,false);
0240     <span class="keyword">end</span>
0241     model.comps={<span class="string">'s'</span>};
0242     model.compNames={<span class="string">'System'</span>};
0243 <span class="keyword">else</span>
0244     raw=<a href="#_sub1" class="code" title="subfunction raw=cleanImported(raw)">cleanImported</a>(raw);
0245 
0246     <span class="comment">%Map to new captions</span>
0247     raw(1,:)=upper(raw(1,:));
0248     raw(1,:)=strrep(raw(1,:),<span class="string">'COMPABBREV'</span>,<span class="string">'ABBREVIATION'</span>);
0249     raw(1,:)=strrep(raw(1,:),<span class="string">'COMPNAME'</span>,<span class="string">'NAME'</span>);
0250 
0251     allLabels={<span class="string">'ABBREVIATION'</span>;<span class="string">'NAME'</span>;<span class="string">'INSIDE'</span>;<span class="string">'MIRIAM'</span>};
0252 
0253     <span class="comment">%Loop through the labels</span>
0254     [I, J]=ismember(raw(1,:),allLabels);
0255     I=find(I);
0256     <span class="keyword">for</span> i=1:numel(I)
0257         <span class="keyword">switch</span> J(I(i))
0258             <span class="keyword">case</span> 1
0259                 model.comps=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0260             <span class="keyword">case</span> 2
0261                 model.compNames=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0262             <span class="keyword">case</span> 3
0263                 model.compOutside=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0264             <span class="keyword">case</span> 4
0265                 model.compMiriams=raw(2:<span class="keyword">end</span>,I(i));
0266         <span class="keyword">end</span>
0267     <span class="keyword">end</span>
0268 
0269     <span class="comment">%Check that necessary fields are loaded</span>
0270     <span class="keyword">if</span> isempty(model.comps)
0271         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There must be a column named ABBREVIATION in the COMPS sheet'</span>);   
0272     <span class="keyword">end</span>
0273     <span class="keyword">if</span> isempty(model.compNames)
0274         model.compNames=model.comps; 
0275         <span class="keyword">if</span> printWarnings==true
0276             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named NAME in the COMPS sheet. ABBREVIATION will be used as name'</span>,false);
0277         <span class="keyword">end</span>
0278     <span class="keyword">end</span>
0279 
0280     model.compMiriams=<a href="#_sub3" class="code" title="subfunction miriamStruct=parseMiriam(strings,miriamStruct)">parseMiriam</a>(model.compMiriams);
0281 <span class="keyword">end</span>
0282 
0283 <span class="comment">%Get all the genes and info about them</span>
0284 [raw, flag]=<a href="#_sub4" class="code" title="subfunction [raw, flag]=loadSheet(workbook, sheet)">loadSheet</a>(workbook,<span class="string">'GENES'</span>);
0285 
0286 <span class="keyword">if</span> flag&lt;0
0287     <span class="keyword">if</span> printWarnings==true
0288         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no spreadsheet named GENES'</span>,false)
0289     <span class="keyword">end</span>
0290 <span class="keyword">else</span>
0291     raw=<a href="#_sub1" class="code" title="subfunction raw=cleanImported(raw)">cleanImported</a>(raw);
0292 
0293     <span class="comment">%Map to new captions</span>
0294     raw(1,:)=upper(raw(1,:));
0295     raw(1,:)=strrep(raw(1,:),<span class="string">'GENE NAME'</span>,<span class="string">'NAME'</span>);
0296 
0297     allLabels={<span class="string">'NAME'</span>;<span class="string">'MIRIAM'</span>;<span class="string">'COMPARTMENT'</span>};
0298     
0299     <span class="comment">%Loop through the labels</span>
0300     [I, J]=ismember(upper(raw(1,:)),allLabels);
0301     I=find(I);
0302     foundGenes=false;
0303     <span class="keyword">for</span> i=1:numel(I)
0304         <span class="keyword">switch</span> J(I(i))
0305             <span class="keyword">case</span> 1
0306                 model.genes=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0307                 foundGenes=true;
0308             <span class="keyword">case</span> 2
0309                 model.geneMiriams=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0310             <span class="keyword">case</span> 3
0311                 model.geneComps=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0312         <span class="keyword">end</span>
0313     <span class="keyword">end</span>
0314 
0315     <span class="keyword">if</span> foundGenes==false
0316         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There must be a column named NAME in the GENES sheet'</span>);
0317     <span class="keyword">end</span>
0318         
0319     <span class="comment">%Its ok if all of them are empty</span>
0320     <span class="keyword">if</span> all(cellfun(@isempty,model.geneComps))
0321         model.geneComps=[];
0322     <span class="keyword">end</span>
0323 
0324     <span class="comment">%Check that geneName contain only strings and no empty strings</span>
0325     <span class="keyword">if</span> ~iscellstr(model.genes)
0326         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'All gene names have to be strings'</span>);
0327     <span class="keyword">else</span>
0328         <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,model.genes))
0329             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There can be no empty strings in gene names'</span>);
0330         <span class="keyword">end</span>
0331     <span class="keyword">end</span>
0332 
0333     <span class="comment">%Check that geneComp contains only strings and no empty string</span>
0334     <span class="keyword">if</span> ~isempty(model.geneComps)
0335         <span class="keyword">if</span> ~iscellstr(model.geneComps)
0336             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'All gene compartments have to be strings'</span>);
0337         <span class="keyword">else</span>
0338             <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,model.geneComps))
0339                 <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There can be no empty strings in gene compartments'</span>);
0340             <span class="keyword">end</span>
0341         <span class="keyword">end</span>
0342         [I, model.geneComps]=ismember(model.geneComps,model.comps);
0343         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following genes have compartment abbreviations which could not be found:'</span>,true,model.genes(~I));
0344     <span class="keyword">end</span>
0345 <span class="keyword">end</span>
0346 
0347 model.geneMiriams=<a href="#_sub3" class="code" title="subfunction miriamStruct=parseMiriam(strings,miriamStruct)">parseMiriam</a>(model.geneMiriams);
0348 
0349 <span class="comment">%Loads the reaction data</span>
0350 [raw, flag]=<a href="#_sub4" class="code" title="subfunction [raw, flag]=loadSheet(workbook, sheet)">loadSheet</a>(workbook,<span class="string">'RXNS'</span>);
0351 
0352 <span class="keyword">if</span> flag&lt;0
0353     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'Could not load the RXNS sheet'</span>);
0354 <span class="keyword">end</span>
0355 
0356 raw=<a href="#_sub1" class="code" title="subfunction raw=cleanImported(raw)">cleanImported</a>(raw);
0357 
0358 <span class="comment">%Map to new captions</span>
0359 raw(1,:)=upper(raw(1,:));
0360 raw(1,:)=strrep(raw(1,:),<span class="string">'RXNID'</span>,<span class="string">'ID'</span>);
0361 
0362 allLabels={<span class="string">'ID'</span>;<span class="string">'NAME'</span>;<span class="string">'EQUATION'</span>;<span class="string">'EC-NUMBER'</span>;<span class="string">'GENE ASSOCIATION'</span>;<span class="string">'LOWER BOUND'</span>;<span class="string">'UPPER BOUND'</span>;<span class="string">'OBJECTIVE'</span>;<span class="string">'COMPARTMENT'</span>;<span class="string">'SUBSYSTEM'</span>;<span class="string">'REPLACEMENT ID'</span>;<span class="string">'MIRIAM'</span>};
0363 equations={};
0364 reactionReplacement={};
0365 
0366 <span class="comment">%Loop through the labels</span>
0367 [I, J]=ismember(raw(1,:),allLabels);
0368 I=find(I);
0369 <span class="keyword">for</span> i=1:numel(I)
0370     <span class="keyword">switch</span> J(I(i))
0371         <span class="keyword">case</span> 1
0372            model.rxns=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0373         <span class="keyword">case</span> 2
0374            model.rxnNames=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0375         <span class="keyword">case</span> 3
0376            equations=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0377         <span class="keyword">case</span> 4
0378            model.eccodes=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0379         <span class="keyword">case</span> 5
0380            model.grRules=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0381         <span class="keyword">case</span> 6
0382            <span class="keyword">try</span>
0383                model.lb=cellfun(@(x) <a href="#_sub6" class="code" title="subfunction y=toDouble(x,default)">toDouble</a>(x,NaN),raw(2:<span class="keyword">end</span>,I(i)));
0384            <span class="keyword">catch</span>
0385                <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The lower bounds must be numerical values'</span>); 
0386            <span class="keyword">end</span>
0387         <span class="keyword">case</span> 7
0388            <span class="keyword">try</span>
0389                model.ub=cellfun(@(x) <a href="#_sub6" class="code" title="subfunction y=toDouble(x,default)">toDouble</a>(x,NaN),raw(2:<span class="keyword">end</span>,I(i)));
0390            <span class="keyword">catch</span>
0391                <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The upper bounds must be numerical values'</span>); 
0392            <span class="keyword">end</span>
0393         <span class="keyword">case</span> 8
0394            <span class="keyword">try</span>
0395                model.c=cellfun(@(x) <a href="#_sub6" class="code" title="subfunction y=toDouble(x,default)">toDouble</a>(x,0),raw(2:<span class="keyword">end</span>,I(i)));
0396            <span class="keyword">catch</span>
0397                <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The objective coefficients must be numerical values'</span>); 
0398            <span class="keyword">end</span>
0399         <span class="keyword">case</span> 9
0400             model.rxnComps=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0401         <span class="keyword">case</span> 10
0402             model.subSystems=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0403         <span class="keyword">case</span> 11
0404             reactionReplacement=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);    
0405         <span class="keyword">case</span> 12
0406             model.rxnMiriams=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0407     <span class="keyword">end</span>
0408 <span class="keyword">end</span>
0409 
0410 <span class="comment">%Check that all necessary reaction info has been loaded</span>
0411 <span class="keyword">if</span> isempty(equations)
0412      <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There must be a column named EQUATION in the RXNS sheet'</span>);   
0413 <span class="keyword">end</span>
0414 <span class="keyword">if</span> isempty(model.rxns)
0415      <span class="keyword">if</span> printWarnings==true
0416         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named ID in the RXNS sheet. The reactions will be named as &quot;r1&quot;, &quot;r2&quot;...'</span>,false);
0417      <span class="keyword">end</span>
0418      I=num2cell((1:numel(equations))');
0419      model.rxns=strcat(<span class="string">'r'</span>,cellfun(@num2str,I,<span class="string">'UniformOutput'</span>,false));
0420 <span class="keyword">end</span>
0421 
0422 <span class="comment">%Check if some other stuff is loaded and populate with default values</span>
0423 <span class="comment">%otherwise</span>
0424 <span class="keyword">if</span> isempty(model.rxnNames)
0425    model.rxnNames=cell(numel(model.rxns),1);
0426    model.rxnNames(:)={<span class="string">''</span>};
0427    <span class="keyword">if</span> printWarnings==true
0428        <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named NAME in the RXNS sheet. Empty strings will be used as reaction names'</span>,false);
0429    <span class="keyword">end</span>
0430 <span class="keyword">end</span>
0431 <span class="keyword">if</span> isempty(model.lb)
0432    <span class="comment">%This is not set here since the reversibility isn't known yet</span>
0433    model.lb=nan(numel(model.rxns),1);
0434    <span class="keyword">if</span> printWarnings==true
0435         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named LOWER BOUND in the RXNS sheet. Default bounds will be used'</span>,false);
0436    <span class="keyword">end</span>
0437 <span class="keyword">end</span>
0438 <span class="keyword">if</span> isempty(model.ub)
0439    model.ub=nan(numel(model.rxns),1);
0440    <span class="keyword">if</span> printWarnings==true
0441         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named UPPER BOUND in the RXNS sheet. Default bounds will be used'</span>,false);
0442    <span class="keyword">end</span>
0443 <span class="keyword">end</span>
0444 <span class="keyword">if</span> isempty(model.c)
0445    model.c=zeros(numel(model.rxns),1);
0446    <span class="keyword">if</span> printWarnings==true
0447         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named OBJECTIVE in the RXNS sheet'</span>,false);
0448    <span class="keyword">end</span>
0449 <span class="keyword">end</span>
0450 
0451 <span class="comment">%Either all reactions must have a compartment string or none of them.</span>
0452 <span class="comment">%Check if it's only empty and if so return it to []</span>
0453 <span class="keyword">if</span> ~isempty(model.rxnComps)
0454     <span class="keyword">if</span> all(cellfun(@isempty,model.rxnComps))
0455         model.rxnComps=[];
0456     <span class="keyword">end</span>
0457 <span class="keyword">end</span>
0458 
0459 <span class="comment">%Construct the rxnMiriams structure</span>
0460 model.rxnMiriams=<a href="#_sub3" class="code" title="subfunction miriamStruct=parseMiriam(strings,miriamStruct)">parseMiriam</a>(model.rxnMiriams);
0461 
0462 <span class="comment">%Replace the reaction IDs for those IDs that have a corresponding</span>
0463 <span class="comment">%replacement name.</span>
0464 I=cellfun(@any,reactionReplacement);
0465 model.rxns(I)=reactionReplacement(I);
0466 
0467 <span class="comment">%Check that there are no empty strings in reactionIDs or equations</span>
0468 <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,model.rxns))
0469     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There are empty reaction IDs'</span>); 
0470 <span class="keyword">end</span>
0471 
0472 <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,equations))
0473     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There are empty equations'</span>); 
0474 <span class="keyword">end</span>
0475 
0476 <span class="keyword">if</span> ~isempty(model.rxnComps)
0477     <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,model.rxnComps))
0478         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'Either all reactions must have an associated compartment string or none of them'</span>); 
0479     <span class="keyword">end</span>
0480 <span class="keyword">end</span>
0481     
0482 <span class="comment">%Check gene association for each reaction and populate rxnGeneMat</span>
0483 <span class="keyword">if</span> ~isempty(model.genes)
0484     model.rxnGeneMat=zeros(numel(model.rxns),numel(model.genes));
0485 <span class="keyword">end</span>
0486 <span class="keyword">if</span> ~isempty(model.grRules)
0487     <span class="keyword">for</span> i=1:length(model.rxns)    
0488        <span class="comment">%Check that all gene associations have a match in the gene list</span>
0489        <span class="keyword">if</span> ~isempty(model.grRules{i})
0490            indexes=strfind(model.grRules{i},<span class="string">':'</span>); <span class="comment">%Genes are separated by &quot;:&quot; for AND and &quot;;&quot; for OR</span>
0491            indexes=unique([indexes strfind(model.grRules{i},<span class="string">';'</span>)]);
0492            <span class="keyword">if</span> isempty(indexes)
0493                <span class="comment">%See if you have a match</span>
0494                I=find(strcmp(model.grRules{i},model.genes));
0495                <span class="keyword">if</span> isempty(I)
0496                    <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>([<span class="string">'The gene association in reaction '</span> model.rxns{i} <span class="string">' ('</span> model.grRules{i} <span class="string">') is not present in the gene list'</span>]);
0497                <span class="keyword">end</span>
0498                model.rxnGeneMat(i,I)=1;
0499            <span class="keyword">else</span>
0500                temp=[0 indexes numel(model.grRules{i})+1];
0501                <span class="keyword">for</span> j=1:numel(indexes)+1;
0502                    <span class="comment">%The reaction has several associated genes</span>
0503                    geneName=model.grRules{i}(temp(j)+1:temp(j+1)-1);
0504                    I=find(strcmp(geneName,model.genes));
0505                    <span class="keyword">if</span> isempty(I)
0506                        <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>([<span class="string">'The gene association in reaction '</span> model.rxns{i} <span class="string">' ('</span> geneName <span class="string">') is not present in the gene list'</span>]);
0507                    <span class="keyword">end</span>
0508                    model.rxnGeneMat(i,I)=1;
0509                <span class="keyword">end</span>
0510            <span class="keyword">end</span>
0511             <span class="comment">%In order to adhere to the COBRA standards it should be like</span>
0512             <span class="comment">%this:</span>
0513             <span class="comment">%-If only one gene then no parentheses</span>
0514             <span class="comment">%-If only &quot;and&quot; or only &quot;or&quot; there should only be one set of</span>
0515             <span class="comment">%parentheses</span>
0516             <span class="comment">%-If both &quot;and&quot; and &quot;or&quot;, then split on &quot;or&quot;. This is not</span>
0517             <span class="comment">%complete, but it's the type of relationship supported by the</span>
0518             <span class="comment">%Excel formulation</span>
0519             aSign=strfind(model.grRules{i},<span class="string">':'</span>);
0520             oSign=strfind(model.grRules{i},<span class="string">';'</span>);
0521             <span class="keyword">if</span> isempty(aSign) &amp;&amp; isempty(oSign)
0522                 model.grRules{i}=model.grRules{i};
0523             <span class="keyword">else</span>
0524                 <span class="keyword">if</span> isempty(aSign)
0525                     model.grRules{i}=[<span class="string">'('</span> strrep(model.grRules{i},<span class="string">';'</span>,<span class="string">' or '</span>) <span class="string">')'</span>];
0526                 <span class="keyword">else</span>
0527                     <span class="keyword">if</span> isempty(oSign)
0528                         model.grRules{i}=[<span class="string">'('</span> strrep(model.grRules{i},<span class="string">':'</span>,<span class="string">' and '</span>) <span class="string">')'</span>];
0529                     <span class="keyword">else</span>
0530                         model.grRules{i}=[<span class="string">'(('</span> strrep(model.grRules{i},<span class="string">';'</span>,<span class="string">') or ('</span>) <span class="string">'))'</span>];
0531                         model.grRules{i}=strrep(model.grRules{i},<span class="string">':'</span>,<span class="string">' and '</span>);
0532                     <span class="keyword">end</span>
0533                 <span class="keyword">end</span>
0534             <span class="keyword">end</span>
0535        <span class="keyword">end</span>
0536     <span class="keyword">end</span>
0537 <span class="keyword">end</span>
0538 model.rxnGeneMat=sparse(model.rxnGeneMat);
0539 
0540 <span class="comment">%Check that the compartment for each reaction can be found</span>
0541 <span class="keyword">if</span> ~isempty(model.rxnComps)
0542     [I, model.rxnComps]=ismember(model.rxnComps,model.comps);
0543     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following reactions have compartment abbreviations which could not be found:'</span>,true,model.rxns(~I));
0544 <span class="keyword">end</span>
0545 
0546 <span class="comment">%Get all the metabolites and info about them</span>
0547 [raw, flag]=<a href="#_sub4" class="code" title="subfunction [raw, flag]=loadSheet(workbook, sheet)">loadSheet</a>(workbook,<span class="string">'METS'</span>);
0548 
0549 <span class="keyword">if</span> flag&lt;0
0550     <span class="keyword">if</span> printWarnings==true
0551         <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no spreadsheet named METS. The metabolites will be named &quot;m1&quot;, &quot;m2&quot;... and assigned to the first compartment'</span>,false);
0552     <span class="keyword">end</span>
0553     <span class="comment">%Parse the equations to find out how many metabolites there are</span>
0554     metsForParsing=<a href="../core/parseRxnEqu.html" class="code" title="function metabolites=parseRxnEqu(equations)">parseRxnEqu</a>(equations);
0555     I=num2cell((1:numel(metsForParsing))');
0556     model.mets=strcat(<span class="string">'m'</span>,cellfun(@num2str,I,<span class="string">'UniformOutput'</span>,false));
0557     model.metComps=ones(numel(model.mets),1);
0558     model.unconstrained=zeros(numel(model.mets),1);
0559     model.metNames=metsForParsing;
0560 <span class="keyword">else</span>
0561     raw=<a href="#_sub1" class="code" title="subfunction raw=cleanImported(raw)">cleanImported</a>(raw);            
0562 
0563     <span class="comment">%Map to new captions</span>
0564     raw(1,:)=upper(raw(1,:));
0565     raw(1,:)=strrep(raw(1,:),<span class="string">'METID'</span>,<span class="string">'ID'</span>);
0566     raw(1,:)=strrep(raw(1,:),<span class="string">'METNAME'</span>,<span class="string">'NAME'</span>);
0567 
0568     allLabels={<span class="string">'ID'</span>;<span class="string">'NAME'</span>;<span class="string">'UNCONSTRAINED'</span>;<span class="string">'MIRIAM'</span>;<span class="string">'COMPOSITION'</span>;<span class="string">'INCHI'</span>;<span class="string">'COMPARTMENT'</span>;<span class="string">'REPLACEMENT ID'</span>};
0569 
0570     <span class="comment">%Load the metabolite information</span>
0571     metReplacement={};
0572 
0573     <span class="comment">%Loop through the labels</span>
0574     [I, J]=ismember(raw(1,:),allLabels);
0575     I=find(I);
0576     <span class="keyword">for</span> i=1:numel(I)
0577         <span class="keyword">switch</span> J(I(i))
0578             <span class="keyword">case</span> 1
0579                model.mets=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0580             <span class="keyword">case</span> 2
0581                model.metNames=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0582             <span class="keyword">case</span> 3
0583                model.unconstrained=cellfun(@<a href="#_sub7" class="code" title="subfunction y=boolToDouble(x)">boolToDouble</a>,raw(2:<span class="keyword">end</span>,I(i)));
0584                
0585                <span class="comment">%NaN is returned if the values couldn't be parsed</span>
0586                <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The UNCONSTRAINED property for the following metabolites must be &quot;true&quot;/&quot;false&quot;, 1/0, TRUE/FALSE or not set:'</span>,true,model.mets(isnan(model.unconstrained)));
0587             <span class="keyword">case</span> 4
0588                model.metMiriams=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0589             <span class="keyword">case</span> 5
0590                model.metFormulas=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0591             <span class="keyword">case</span> 6
0592                model.inchis=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0593             <span class="keyword">case</span> 7
0594                 model.metComps=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0595 
0596                 <span class="comment">%Check that all metabolites have compartments defined</span>
0597                 <span class="keyword">if</span> any(strcmp(<span class="string">''</span>,model.metComps))
0598                     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'All metabolites must have an associated compartment string'</span>); 
0599                 <span class="keyword">end</span>
0600             <span class="keyword">case</span> 8
0601                metReplacement=cellfun(@<a href="#_sub5" class="code" title="subfunction y=toStr(x)">toStr</a>,raw(2:<span class="keyword">end</span>,I(i)),<span class="string">'UniformOutput'</span>,false);
0602         <span class="keyword">end</span>
0603     <span class="keyword">end</span>
0604 
0605     <span class="comment">%Check that necessary fields are loaded (METID)</span>
0606     <span class="keyword">if</span> isempty(model.mets)
0607          <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There must be a column named ID in the METS sheet'</span>);   
0608     <span class="keyword">end</span>
0609 
0610     <span class="comment">%Check that some other stuff is loaded and use default values otherwise</span>
0611     <span class="keyword">if</span> isempty(model.metNames)
0612        model.metNames=cell(numel(model.mets),1);
0613        <span class="keyword">if</span> printWarnings==true
0614             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named NAME in the METS sheet. ID will be used as name'</span>,false);
0615        <span class="keyword">end</span>
0616     <span class="keyword">end</span>
0617     <span class="keyword">if</span> isempty(model.unconstrained)
0618        model.unconstrained=zeros(numel(model.mets),1);
0619        <span class="keyword">if</span> printWarnings==true
0620             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named UNCONSTRAINED in the METS sheet. All metabolites will be constrained'</span>,false);
0621        <span class="keyword">end</span>
0622     <span class="keyword">end</span>
0623 
0624     <span class="keyword">if</span> isempty(model.metComps)
0625        model.metComps=cell(numel(model.mets),1);
0626        model.metComps(:)=model.comps(1);
0627        <span class="keyword">if</span> printWarnings==true
0628             <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'There is no column named COMPARTMENT in the METS sheet. All metabolites will be assigned to the first compartment in COMPS. Note that RAVEN makes extensive use of metabolite names and compartments. Some features will therefore not function correctly if metabolite compartments are not correctly assigned'</span>,false);
0629        <span class="keyword">end</span>
0630     <span class="keyword">end</span>
0631 
0632     <span class="comment">%The composition should be loaded from InChIs when available</span>
0633     I=find(~cellfun(@isempty,model.inchis));
0634     <span class="keyword">for</span> i=1:numel(I)
0635         S=regexp(model.inchis(I(i)),<span class="string">'/'</span>,<span class="string">'split'</span>);
0636         S=S{1};
0637         <span class="keyword">if</span> numel(S)&gt;=2
0638             <span class="comment">%Don't copy if it doesn't look good</span>
0639             model.metFormulas(I(i))=S(2);
0640         <span class="keyword">end</span>
0641     <span class="keyword">end</span>
0642 
0643     <span class="comment">%Check that the compartment for each metabolite can be found. Also convert</span>
0644     <span class="comment">%from id to index</span>
0645     [I, model.metComps]=ismember(model.metComps,model.comps);
0646     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following metabolites have compartment abbreviations which could not be found:'</span>,true,model.mets(~I));
0647     
0648     <span class="comment">%Check that the model.mets vector is unique. The problem is that</span>
0649     <span class="comment">%the checkModelStruct cannot check for that since only the metReplacements</span>
0650     <span class="comment">%(if used) end up in the model structure</span>
0651     I=false(numel(model.mets),1);
0652     [J, K]=unique(model.mets);
0653     <span class="keyword">if</span> numel(J)~=numel(model.mets)
0654         L=1:numel(model.mets);
0655         L(K)=[];
0656         I(L)=true;
0657     <span class="keyword">end</span>
0658     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following metabolites are duplicates:'</span>,~ignoreErrors,model.mets(I));
0659 
0660     <span class="comment">%Check that there are no metabolite IDs which are numbers. This would</span>
0661     <span class="comment">%give errors when parsing the equations</span>
0662     I=cellfun(@str2double,model.mets);
0663     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following metabolites have names which cannot be distinguished from numbers:'</span>,~ignoreErrors,model.mets(~isnan(I)));
0664     I=cellfun(@str2double,metReplacement);
0665     <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following metabolites have names which cannot be distinguished from numbers:'</span>,~ignoreErrors,metReplacement(~isnan(I)));
0666     
0667     <span class="comment">%Replace the metabolite IDs for those IDs that have a corresponding</span>
0668     <span class="comment">%replacement metabolite. This is not used for matching, but will be checked</span>
0669     <span class="comment">%for consistency with SBML naming conventions</span>
0670     metsForParsing=model.mets; <span class="comment">%This is because the equations are written with this</span>
0671     I=cellfun(@any,metReplacement);
0672     model.mets(I)=metReplacement(I);
0673     
0674     <span class="comment">%If the metabolite name isn't set, replace it with the metabolite id</span>
0675     I=~cellfun(@any,model.metNames);
0676     model.metNames(I)=model.mets(I);
0677 
0678     <span class="comment">%Construct the metMiriams structure</span>
0679     model.metMiriams=<a href="#_sub3" class="code" title="subfunction miriamStruct=parseMiriam(strings,miriamStruct)">parseMiriam</a>(model.metMiriams);
0680 <span class="keyword">end</span>
0681 
0682 <span class="comment">%Everything seems fine with the metabolite IDs, compartments, genes, and</span>
0683 <span class="comment">%reactions</span>
0684 
0685 <span class="comment">%Parse the equations</span>
0686 [model.S, mets, badRxns, model.rev]=<a href="../core/constructS.html" class="code" title="function [S mets badRxns reversible]=constructS(equations,mets,rxns)">constructS</a>(equations,metsForParsing,model.rxns);
0687 model.rev=model.rev*1; <span class="comment">%Typecast to double</span>
0688 
0689 <span class="comment">%Add default constraints</span>
0690 model.lb(isnan(model.lb))=model.annotation.defaultLB.*model.rev(isnan(model.lb));
0691 model.ub(isnan(model.ub))=model.annotation.defaultUB;
0692 
0693 <span class="comment">%Reorder the S matrix so that it fits with the metabolite list in the</span>
0694 <span class="comment">%structure</span>
0695 [crap, I]=ismember(mets,metsForParsing);
0696 model.S=model.S(I,:);
0697 
0698 <span class="comment">%Print warnings about the reactions which contain the same metabolite as</span>
0699 <span class="comment">%both reactants and products</span>
0700 <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(<span class="string">'The following reactions have metabolites which are present more than once. Only the net reactions will be exported:'</span>,false,model.rxns(badRxns));
0701 
0702 model.b=zeros(numel(model.mets),1);
0703 
0704 <span class="comment">%Remove unused fields</span>
0705 <span class="keyword">if</span> isempty(model.compOutside)
0706     model=rmfield(model,<span class="string">'compOutside'</span>);
0707 <span class="keyword">end</span>
0708 <span class="keyword">if</span> isempty(model.compMiriams)
0709     model=rmfield(model,<span class="string">'compMiriams'</span>);
0710 <span class="keyword">end</span>
0711 <span class="keyword">if</span> isempty(model.rxnComps)
0712     model=rmfield(model,<span class="string">'rxnComps'</span>);
0713 <span class="keyword">end</span>
0714 <span class="keyword">if</span> isempty(model.grRules)
0715     model=rmfield(model,<span class="string">'grRules'</span>);
0716 <span class="keyword">end</span>
0717 <span class="keyword">if</span> isempty(model.rxnGeneMat)
0718     model=rmfield(model,<span class="string">'rxnGeneMat'</span>);
0719 <span class="keyword">end</span>
0720 <span class="keyword">if</span> isempty(model.subSystems)
0721     model=rmfield(model,<span class="string">'subSystems'</span>);
0722 <span class="keyword">end</span>
0723 <span class="keyword">if</span> isempty(model.eccodes)
0724     model=rmfield(model,<span class="string">'eccodes'</span>);
0725 <span class="keyword">end</span>
0726 <span class="keyword">if</span> isempty(model.rxnMiriams)
0727     model=rmfield(model,<span class="string">'rxnMiriams'</span>);
0728 <span class="keyword">end</span>
0729 <span class="keyword">if</span> isempty(model.genes)
0730     model=rmfield(model,<span class="string">'genes'</span>);
0731 <span class="keyword">end</span>
0732 <span class="keyword">if</span> isempty(model.geneComps)
0733     model=rmfield(model,<span class="string">'geneComps'</span>);
0734 <span class="keyword">end</span>
0735 <span class="keyword">if</span> isempty(model.geneMiriams)
0736     model=rmfield(model,<span class="string">'geneMiriams'</span>);
0737 <span class="keyword">end</span>
0738 <span class="keyword">if</span> isempty(model.inchis)
0739     model=rmfield(model,<span class="string">'inchis'</span>);
0740 <span class="keyword">end</span>
0741 <span class="keyword">if</span> isempty(model.metFormulas)
0742     model=rmfield(model,<span class="string">'metFormulas'</span>);
0743 <span class="keyword">end</span>
0744 <span class="keyword">if</span> isempty(model.metMiriams)
0745     model=rmfield(model,<span class="string">'metMiriams'</span>);
0746 <span class="keyword">end</span>
0747 
0748 <span class="comment">%The model structure has now been reconstructed but it can still contain</span>
0749 <span class="comment">%many types of errors. The checkModelConsistency function is used to make</span>
0750 <span class="comment">%sure that naming and mapping of stuff looks good</span>
0751 <a href="../core/checkModelStruct.html" class="code" title="function checkModelStruct(model,throwErrors,trimWarnings)">checkModelStruct</a>(model,~ignoreErrors);
0752 
0753 <span class="keyword">if</span> removeExcMets==true
0754     model=<a href="../core/simplifyModel.html" class="code" title="function [reducedModel, deletedReactions, deletedMetabolites]=simplifyModel(model,deleteUnconstrained, deleteDuplicates, deleteZeroInterval, deleteInaccessible, deleteMinMax, groupLinear, reservedRxns, suppressWarnings)">simplifyModel</a>(model);
0755 <span class="keyword">end</span>
0756 <span class="keyword">end</span>
0757 
0758 <span class="comment">%Cleans up the structure that is imported from using xlsread</span>
0759 <a name="_sub1" href="#_subfunctions" class="code">function raw=cleanImported(raw)</a>
0760     <span class="comment">%First check that it's a cell array. If a sheet is completely empty,</span>
0761     <span class="comment">%then raw=NaN</span>
0762     <span class="keyword">if</span> iscell(raw)
0763         <span class="comment">%Clear cells which contain only white spaces or NaN. This could</span>
0764         <span class="comment">%happen if you accidentally inserted a space for example. I don't</span>
0765         <span class="comment">%know how NaN could occur after switching to Apache POI, but I</span>
0766         <span class="comment">%clear it to be sure</span>
0767         whites=cellfun(@<a href="#_sub2" class="code" title="subfunction I=wrapperWS(A)">wrapperWS</a>,raw);
0768         raw(whites)={[]};
0769         
0770         <span class="comment">%Remove columns that don't have string headers. If you cut and paste</span>
0771         <span class="comment">%a lot in the sheet there tends to be columns that are empty</span>
0772         I=cellfun(@isstr,raw(1,:));
0773         raw=raw(:,I);
0774         
0775         <span class="comment">%Find the rows that are not commented. This corresponds to the</span>
0776         <span class="comment">%first row and the ones which are empty in the first column</span>
0777         
0778         keepers=cellfun(@isempty,raw(:,1));
0779         keepers(1)=true;
0780         raw=raw(keepers,:);
0781 
0782         <span class="comment">%Check if there are any rows that are all empty. This could happen if</span>
0783         <span class="comment">%it reads too far or if the user has inserted them for layout reasons.</span>
0784         <span class="comment">%Remove any such rows.</span>
0785         I=~all(cellfun(@isempty,raw),2);
0786         raw=raw(I,:);
0787     <span class="keyword">else</span>
0788         raw={[]};
0789     <span class="keyword">end</span>
0790     
0791     <span class="comment">%Checks if something is all white spaces or NaN</span>
0792     <a name="_sub2" href="#_subfunctions" class="code">function I=wrapperWS(A)</a>
0793         <span class="keyword">if</span> isnan(A)
0794             I=true;
0795         <span class="keyword">else</span>
0796             <span class="comment">%isstrprop gives an error if boolean</span>
0797             <span class="keyword">if</span> islogical(A)
0798                 I=false;
0799             <span class="keyword">else</span>
0800                 <span class="comment">%I have to make this check, since otherwise it will pick up</span>
0801                 <span class="comment">%on doubles for which the ASCII representation is a white</span>
0802                 <span class="comment">%space character</span>
0803                 <span class="keyword">if</span> isnumeric(A)
0804                     I=false;
0805                 <span class="keyword">else</span>
0806                     I=all(isstrprop(A,<span class="string">'wspace'</span>));
0807                 <span class="keyword">end</span>
0808             <span class="keyword">end</span>
0809         <span class="keyword">end</span>
0810     <span class="keyword">end</span>
0811 <span class="keyword">end</span>
0812 <a name="_sub3" href="#_subfunctions" class="code">function miriamStruct=parseMiriam(strings,miriamStruct)</a>
0813 <span class="comment">%Gets the names and values of Miriam-string. Nothing fancy at all, just to</span>
0814 <span class="comment">%prevent using the same code for metabolites, genes, and reactions. The</span>
0815 <span class="comment">%function also allows for supplying a miriamStruct and the info will then</span>
0816 <span class="comment">%be added</span>
0817 
0818 <span class="keyword">if</span> nargin&lt;2
0819     miriamStruct=cell(numel(strings),1);
0820 <span class="keyword">end</span>
0821 <span class="keyword">for</span> i=1:numel(strings)
0822     <span class="keyword">if</span> any(strings{i})
0823         <span class="comment">%A Miriam string can be several ids separated by &quot;;&quot;. Each id is</span>
0824         <span class="comment">%&quot;name(..:..):value&quot;</span>
0825         I=regexp(strings{i},<span class="string">';'</span>,<span class="string">'split'</span>);
0826         <span class="keyword">if</span> isfield(miriamStruct{i},<span class="string">'name'</span>)
0827             startIndex=numel(miriamStruct{i}.name);
0828             miriamStruct{i}.name=[miriamStruct{i}.name;cell(numel(I),1)];
0829             miriamStruct{i}.value=[miriamStruct{i}.value;cell(numel(I),1)];
0830         <span class="keyword">else</span>
0831             startIndex=0;
0832             miriamStruct{i}.name=cell(numel(I),1);
0833             miriamStruct{i}.value=cell(numel(I),1);
0834         <span class="keyword">end</span>
0835         
0836         <span class="keyword">for</span> j=1:numel(I)
0837             index=max(strfind(I{j},<span class="string">':'</span>));
0838             <span class="keyword">if</span> any(index)
0839                 miriamStruct{i}.name{startIndex+j}=I{j}(1:index-1);
0840                 miriamStruct{i}.value{startIndex+j}=I{j}(index+1:end);
0841             <span class="keyword">else</span>
0842                 <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>([<span class="string">'&quot;'</span> I{j} <span class="string">'&quot; is not a valid MIRIAM string. The format must be &quot;identifier:value&quot;'</span>]);
0843             <span class="keyword">end</span>
0844         <span class="keyword">end</span>
0845     <span class="keyword">end</span>
0846 <span class="keyword">end</span>
0847 <span class="keyword">end</span>
0848 
0849 <span class="comment">%Loads a sheet into a cell matrix using the Java library Apache POI.</span>
0850 
0851 <span class="comment">%flag       -1 if the sheet couldn't be found</span>
0852 <span class="comment">%           0 if everything worked well</span>
0853 <a name="_sub4" href="#_subfunctions" class="code">function [raw, flag]=loadSheet(workbook, sheet)      </a>
0854     flag=0;
0855     raw={};
0856     
0857     sh=workbook.getSheet(sheet);
0858     <span class="keyword">if</span> isempty(sh)
0859         flag=-1;
0860         <span class="keyword">return</span>;
0861     <span class="keyword">end</span>
0862     
0863     lastRow=sh.getLastRowNum();
0864     wasEmpty=false(lastRow+1,1);
0865     raw=cell(lastRow+1,0); <span class="comment">%Allocate space for the cell array. The number of columns isn't know yet, as it's saved row-by-row</span>
0866     <span class="keyword">for</span> i=0:lastRow
0867         row=sh.getRow(i);
0868         <span class="comment">%Sometimes the last rows only contain formatting (or some other</span>
0869         <span class="comment">%weird Excel thing). Ignore such empty rows. Note the +1 to deal with that Matlab indexing</span>
0870         <span class="comment">%starts at 1</span>
0871         <span class="keyword">if</span> isempty(row)
0872             wasEmpty(i+1)=true;
0873             <span class="keyword">continue</span>;
0874         <span class="keyword">end</span>
0875         lastCol=row.getLastCellNum();
0876    
0877         <span class="comment">%Adjust the size of the cell array if needed</span>
0878         <span class="keyword">if</span> (lastCol+1)&gt;size(raw,2)
0879             raw=[raw cell(lastRow+1,lastCol+1-size(raw,2))];
0880         <span class="keyword">end</span>
0881         
0882         <span class="comment">%Loop over the columns</span>
0883         <span class="keyword">for</span> j=0:lastCol
0884            c=row.getCell(j,row.RETURN_BLANK_AS_NULL);
0885            <span class="keyword">if</span> ~isempty(c)
0886                <span class="comment">%Then decide how to save it depending on the type</span>
0887                <span class="keyword">switch</span> c.getCellType()
0888                    <span class="keyword">case</span> c.CELL_TYPE_STRING
0889                         raw{i+1,j+1}=char(c.getRichStringCellValue().getString());
0890                    <span class="keyword">case</span> c.CELL_TYPE_NUMERIC
0891                         raw{i+1,j+1}=c.getNumericCellValue();
0892                    <span class="keyword">case</span> c.CELL_TYPE_BOOLEAN
0893                         raw{i+1,j+1}=c.getBooleanCellValue();
0894                <span class="keyword">end</span>
0895            <span class="keyword">end</span>
0896         <span class="keyword">end</span>
0897     <span class="keyword">end</span>
0898     
0899     <span class="comment">%Remove empty rows</span>
0900     raw(wasEmpty,:)=[];
0901 <span class="keyword">end</span>
0902 
0903 <span class="comment">%For converting a value to string. This is used instead of num2str because</span>
0904 <span class="comment">%I want to convert empty cells to {''}.</span>
0905 <a name="_sub5" href="#_subfunctions" class="code">function y=toStr(x)</a>
0906     <span class="comment">%x can be empty, numerical, string or boolean. It cannot be NaN.</span>
0907     <span class="comment">%Boolean values will be converted to '1'/'0'</span>
0908     <span class="keyword">if</span> isempty(x)
0909         y=<span class="string">''</span>;
0910     <span class="keyword">else</span>
0911         y=num2str(x);
0912     <span class="keyword">end</span>
0913 <span class="keyword">end</span>
0914 
0915 <span class="comment">%For converting to numeric. This is used instead of str2num because I want</span>
0916 <span class="comment">%to be able to choose what empty values should be mapped to.</span>
0917 <span class="comment">%</span>
0918 <span class="comment">% default the value to use for empty input</span>
0919 <a name="_sub6" href="#_subfunctions" class="code">function y=toDouble(x,default)</a>
0920     <span class="keyword">if</span> isempty(x) <span class="comment">%Note that this catches '' as well</span>
0921         y=default;
0922     <span class="keyword">else</span>
0923         <span class="keyword">if</span> isnumeric(x)
0924             y=x;
0925         <span class="keyword">else</span>
0926             y=str2double(x);
0927             
0928             <span class="comment">%This happens if the input couldn't be converted.</span>
0929             <span class="comment">%Note that the input itself cannot be NaN since it was fixed in</span>
0930             <span class="comment">%clean imported</span>
0931             <span class="keyword">if</span> isnan(y)
0932                 <a href="../core/dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>([<span class="string">'Cannot convert the string &quot;'</span> x <span class="string">'&quot; to double'</span>]);
0933             <span class="keyword">end</span>
0934         <span class="keyword">end</span>
0935     <span class="keyword">end</span>
0936 <span class="keyword">end</span>
0937 
0938 <span class="comment">%For converting boolean (the UNCONSTRAINED field) to double (the</span>
0939 <span class="comment">%model.unconstrained field)</span>
0940 <a name="_sub7" href="#_subfunctions" class="code">function y=boolToDouble(x)</a>
0941     <span class="keyword">if</span> isempty(x)
0942         y=0;
0943         <span class="keyword">return</span>;
0944     <span class="keyword">end</span>
0945     <span class="keyword">if</span> islogical(x)
0946         y=x*1; <span class="comment">%Typecast to double</span>
0947         <span class="keyword">return</span>;
0948     <span class="keyword">end</span>
0949     <span class="keyword">if</span> isnumeric(x)
0950        <span class="keyword">if</span> x~=0
0951            y=1;
0952            <span class="keyword">return</span>;
0953        <span class="keyword">else</span>
0954            y=0;
0955            <span class="keyword">return</span>;
0956        <span class="keyword">end</span>
0957     <span class="keyword">end</span>
0958     <span class="keyword">if</span> ischar(x)
0959        <span class="keyword">if</span> strcmpi(x,<span class="string">'TRUE'</span>)
0960            y=1;
0961            <span class="keyword">return</span>;
0962        <span class="keyword">end</span>
0963        <span class="keyword">if</span> strcmpi(x,<span class="string">'FALSE'</span>)
0964            y=0;
0965            <span class="keyword">return</span>;
0966        <span class="keyword">end</span>
0967     <span class="keyword">end</span>
0968     y=NaN; <span class="comment">%This means that the input couldn't be parsed</span>
0969 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Fri 14-Oct-2016 12:02:10 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>