<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of setColorToMapRxns</title>
  <meta name="keywords" content="setColorToMapRxns">
  <meta name="description" content="USAGE:">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">plotting</a> &gt; setColorToMapRxns.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for plotting&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>setColorToMapRxns
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>USAGE:</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function [modifiedMap, xmlMap, model2] = setColorToMapRxns (mapFileName, model, v1, rxnsFluxTask, rxnLineWidth, rxnLineColor, molFillColor) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> USAGE:
 [modifiedMap, xmlMap, model2] =
   setColorToMapRxns (mapFileName, model, v1, rxnsFluxTask, rxnLineWidth, rxnLineColor, molFillColor)

 INPUTS:
 mapFileName           relative path to XML map (e.g. 'yeastMap1.0.xml')
 model                 model structure
 v1                    vector of fluxes
 rxnsFluxTask          1 = Flux
                       2 = Flux essentiality; map reactions causing any
                           effect on growth (i.e. obj function)
                       3 = Flux essentiality; map reactions causing no
                           growth(i.e. obj function) when deleted, i.e.
                           essential reactions
                       4 = Flux essentiality; map reactions with not any
                           effect on growth (i.e. obj function) when
                           deleted
                       5 = Omics
 ADDITIONAL INPUTS:
 rxnLineWidth          reaction line width; default = 5
 rxnLineColor          reaction line color; default = Pathways; write
                       'Subsystems' to color five subsystems only (central,
                       AA, lipids, cofactors, nuclotides and energy).
                       'Subsystems' option is the deault option at task 5
 molFillColor          molecule fill color; default = true (i.e. white)
                       molecule fill color; true (i.e. like CellDesigner default)
 OUTPUTS:
 modifiedMap           map with MATLAB structure, and modified reactions' line color and width
 xmlMap                xml structure of the original map obtained from mapFileName
 model2                model with modified subsystems</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>	Modified from colorSubsystemCD as distributed through COBRA Toolbox</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [modifiedMap, xmlMap, model2] = setColorToMapRxns (mapFileName, model, v1, rxnsFluxTask, rxnLineWidth, rxnLineColor, molFillColor)</a>
0002 <span class="comment">% USAGE:</span>
0003 <span class="comment">% [modifiedMap, xmlMap, model2] =</span>
0004 <span class="comment">%   setColorToMapRxns (mapFileName, model, v1, rxnsFluxTask, rxnLineWidth, rxnLineColor, molFillColor)</span>
0005 <span class="comment">%</span>
0006 <span class="comment">% INPUTS:</span>
0007 <span class="comment">% mapFileName           relative path to XML map (e.g. 'yeastMap1.0.xml')</span>
0008 <span class="comment">% model                 model structure</span>
0009 <span class="comment">% v1                    vector of fluxes</span>
0010 <span class="comment">% rxnsFluxTask          1 = Flux</span>
0011 <span class="comment">%                       2 = Flux essentiality; map reactions causing any</span>
0012 <span class="comment">%                           effect on growth (i.e. obj function)</span>
0013 <span class="comment">%                       3 = Flux essentiality; map reactions causing no</span>
0014 <span class="comment">%                           growth(i.e. obj function) when deleted, i.e.</span>
0015 <span class="comment">%                           essential reactions</span>
0016 <span class="comment">%                       4 = Flux essentiality; map reactions with not any</span>
0017 <span class="comment">%                           effect on growth (i.e. obj function) when</span>
0018 <span class="comment">%                           deleted</span>
0019 <span class="comment">%                       5 = Omics</span>
0020 <span class="comment">% ADDITIONAL INPUTS:</span>
0021 <span class="comment">% rxnLineWidth          reaction line width; default = 5</span>
0022 <span class="comment">% rxnLineColor          reaction line color; default = Pathways; write</span>
0023 <span class="comment">%                       'Subsystems' to color five subsystems only (central,</span>
0024 <span class="comment">%                       AA, lipids, cofactors, nuclotides and energy).</span>
0025 <span class="comment">%                       'Subsystems' option is the deault option at task 5</span>
0026 <span class="comment">% molFillColor          molecule fill color; default = true (i.e. white)</span>
0027 <span class="comment">%                       molecule fill color; true (i.e. like CellDesigner default)</span>
0028 <span class="comment">% OUTPUTS:</span>
0029 <span class="comment">% modifiedMap           map with MATLAB structure, and modified reactions' line color and width</span>
0030 <span class="comment">% xmlMap                xml structure of the original map obtained from mapFileName</span>
0031 <span class="comment">% model2                model with modified subsystems</span>
0032 
0033 <span class="keyword">if</span> ~(exist(<span class="string">'transformXML2Map'</span>)==2)
0034     error(<span class="string">'COBRA Toolbox is required, go to https://opencobra.github.io/cobratoolbox/'</span>)
0035 <span class="keyword">end</span>
0036 
0037 <span class="keyword">if</span> nargin &lt; 5 || rxnLineWidth == 0
0038     rxnLineWidth = 0;
0039     rxnLineColor = <span class="string">''</span>;
0040     molFillColor = true;
0041 <span class="keyword">elseif</span> nargin &lt; 6
0042     rxnLineColor = <span class="string">''</span>;
0043     molFillColor = true;
0044 <span class="keyword">elseif</span> nargin &lt; 7
0045     molFillColor = true;
0046 <span class="keyword">end</span>
0047 <span class="keyword">if</span> rxnsFluxTask == 5
0048     rxnLineColor = <span class="string">'Subsystems'</span>;
0049 <span class="keyword">end</span>
0050 
0051 [xmlMap, mapMap] = transformXML2Map(mapFileName);
0052 <span class="keyword">if</span> molFillColor == true
0053     modifiedMap = unifyMetabolicMapCD(mapMap); <span class="comment">% molColor whrite and rxnColor = ligthgray</span>
0054 <span class="keyword">elseif</span> molFillColor == false
0055     modifiedMap = mapMap; <span class="comment">% like CellDesigner default</span>
0056 <span class="keyword">end</span>
0057 
0058 <span class="comment">%%%%%%%% Gettin reactions line width with normalized fluxes %%%%%%%%%%%%%</span>
0059 model2 = model; <span class="comment">% to not overwrite model</span>
0060 flxs(:, 1) = v1; <span class="comment">% to not overwrite v1</span>
0061 absFlxs = abs(flxs);
0062 rxnLineWidthInMap = absFlxs / max(absFlxs); <span class="comment">%normalize fluxes line width</span>
0063 <span class="keyword">if</span> rxnsFluxTask == 1 &amp;&amp; rxnLineWidth == 0
0064     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.1 &amp; rxnLineWidthInMap &lt;=  1) = 50;
0065     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.01 &amp; rxnLineWidthInMap &lt;= 0.1) = 25;
0066     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.001 &amp; rxnLineWidthInMap &lt;= 0.01) = 16;
0067     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.0001 &amp; rxnLineWidthInMap &lt;= 0.001) = 10;
0068     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.00001 &amp; rxnLineWidthInMap &lt;= 0.0001) = 6;
0069 <span class="keyword">elseif</span> rxnsFluxTask == 2 &amp;&amp; rxnLineWidth == 0
0070     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.90 &amp; rxnLineWidthInMap &lt;=  0.95) =  6;
0071     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.80 &amp; rxnLineWidthInMap &lt;= 0.90) =  8;
0072     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.50 &amp; rxnLineWidthInMap &lt;= 0.80) = 11;
0073     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.25 &amp; rxnLineWidthInMap &lt;= 0.50) = 15;
0074     rxnLineWidthInMap(rxnLineWidthInMap &gt;= 0.00 &amp; rxnLineWidthInMap &lt;= 0.25) = 20;
0075 <span class="keyword">elseif</span> rxnsFluxTask == 1 &amp;&amp; rxnLineWidth ~= 0.0
0076     rxnLineWidthInMap(rxnLineWidthInMap == 1.0) = rxnLineWidth;
0077 <span class="keyword">elseif</span> rxnsFluxTask == 2 &amp;&amp; rxnLineWidth ~= 0.0
0078     rxnLineWidthInMap(rxnLineWidthInMap &lt; 0.95) = rxnLineWidth;
0079 <span class="keyword">elseif</span> rxnsFluxTask == 3 &amp;&amp; rxnLineWidth == 0.0
0080     rxnLineWidthInMap(rxnLineWidthInMap == 0.0) = 5;
0081 <span class="keyword">elseif</span> rxnsFluxTask == 3 &amp;&amp; rxnLineWidth ~= 0.0
0082     rxnLineWidthInMap(rxnLineWidthInMap == 0.0) = rxnLineWidth;
0083 <span class="keyword">elseif</span> rxnsFluxTask == 4 &amp;&amp; rxnLineWidth == 0.0
0084     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.95 &amp; rxnLineWidthInMap &lt;= 1.0) = 5.0;
0085 <span class="keyword">elseif</span> rxnsFluxTask == 4 &amp;&amp; rxnLineWidth ~= 0.0
0086     rxnLineWidthInMap(rxnLineWidthInMap &gt; 0.95 &amp; rxnLineWidthInMap &lt;= 1.0) = rxnLineWidth;
0087 <span class="keyword">elseif</span> rxnsFluxTask == 5 &amp;&amp; rxnLineWidth == 0.0
0088     cutValue = 2.0; <span class="comment">% transcriptomics minimum cut value of log change</span>
0089     rxnLineWidthInMap = absFlxs;
0090     rxnLineWidthInMap(rxnLineWidthInMap &gt; 5.0) = 30;
0091     rxnLineWidthInMap(rxnLineWidthInMap &gt; 4.0 &amp; rxnLineWidthInMap &lt;= 5.0) = 25;
0092     rxnLineWidthInMap(rxnLineWidthInMap &gt; 3.0 &amp; rxnLineWidthInMap &lt;= 4.0) = 20;
0093     rxnLineWidthInMap(rxnLineWidthInMap &gt; 2.0 &amp; rxnLineWidthInMap &lt;= 3.0) = 15;
0094     rxnLineWidthInMap(rxnLineWidthInMap &gt; 1.0 &amp; rxnLineWidthInMap &lt;= 2.0) = 10;
0095     rxnLineWidthInMap(rxnLineWidthInMap &gt;= 0.5 &amp; rxnLineWidthInMap &lt;= 1.0) =  5;
0096 <span class="keyword">end</span>
0097 <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0098     a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0099     <span class="keyword">if</span> isempty(a)
0100         modifiedMap.rxnWidth{i} = 1;
0101     <span class="keyword">else</span>
0102         modifiedMap.rxnWidth{i} = rxnLineWidthInMap(a);
0103     <span class="keyword">end</span>
0104 <span class="keyword">end</span>
0105 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0106 
0107 <span class="comment">%%%%%%%%% Cofactors not included in the analyses %%%%%%%%%%%%%%%%%%%%%%%%</span>
0108 <span class="comment">%%%%%%%%% Compartments included in the analyses %%%%%%%%%%%%%%%%%%%%%%%%%</span>
0109 
0110 <span class="comment">% propossed cofactors list for PART A</span>
0111 cofactorsListA ={<span class="string">'h'</span>; <span class="string">'h2o'</span>; <span class="string">'nad'</span>; <span class="string">'nadh'</span>; <span class="string">'nadp'</span>; <span class="string">'nadph'</span>;<span class="keyword">...</span>
0112     <span class="string">'fad'</span>; <span class="string">'fadh'</span>; <span class="string">'ppi'</span>; <span class="string">'adp'</span>; <span class="string">'gtp'</span>; <span class="string">'gdp'</span>; <span class="string">'gmp'</span>; <span class="keyword">...</span>
0113     <span class="string">'utp'</span>; <span class="string">'udp'</span>; <span class="string">'ump'</span>; <span class="string">'ctp'</span>; <span class="string">'cdp'</span>; <span class="string">'cmp'</span>};
0114 
0115 <span class="comment">% propossed cofactors list for PART B</span>
0116 cofactorsListB ={<span class="string">'h'</span>; <span class="string">'h2o'</span>; <span class="string">'co2'</span>; <span class="string">'coa'</span>; <span class="string">'accoa'</span>; <span class="string">'nad'</span>; <span class="string">'nadh'</span>; <span class="string">'nadp'</span>; <span class="string">'nadph'</span>;<span class="keyword">...</span>
0117     <span class="string">'fad'</span>; <span class="string">'fadh'</span>; <span class="string">'pi'</span>; <span class="string">'ppi'</span>; <span class="string">'atp'</span>; <span class="string">'adp'</span>; <span class="string">'amp'</span>; <span class="string">'gtp'</span>; <span class="string">'gdp'</span>; <span class="string">'gmp'</span>;<span class="keyword">...</span>
0118     <span class="string">'utp'</span>; <span class="string">'udp'</span>; <span class="string">'ump'</span>; <span class="string">'ctp'</span>; <span class="string">'cdp'</span>; <span class="string">'cmp'</span>; <span class="string">'gthox'</span>; <span class="string">'gthrd_c'</span>; <span class="string">'amet_c'</span>;<span class="keyword">...</span>
0119     <span class="string">'ahcys_c'</span>; <span class="string">'glu__L'</span>; <span class="string">'gln__L'</span>; <span class="string">'akg'</span>; <span class="string">'nh4'</span>};
0120 
0121 <span class="comment">% compartments list</span>
0122 compartmentsList = {<span class="string">'_e'</span>; <span class="string">'_c'</span>; <span class="string">'_m'</span>; <span class="string">'_n'</span>; <span class="string">'_x'</span>; <span class="string">'_v'</span>; <span class="string">'_r'</span>; <span class="string">'_g'</span>};
0123 
0124 <span class="comment">% include compartment to cofactor lists A and B</span>
0125 cofactorsListWcompA = cell(length(cofactorsListA), 1);
0126 c = 1; <span class="comment">% general counter</span>
0127 <span class="keyword">for</span> i = 1:length(cofactorsListA)
0128     <span class="keyword">for</span> j = 1:length(compartmentsList)
0129         cofactorsListWcompA{c, 1} = horzcat(cofactorsListA{i},compartmentsList{j});
0130         c = c + 1;
0131     <span class="keyword">end</span>
0132 <span class="keyword">end</span>
0133 cofactorsListWcompB = cell(length(cofactorsListB), 1);
0134 c = 1;
0135 <span class="keyword">for</span> i = 1:length(cofactorsListB)
0136     <span class="keyword">for</span> j = 1:length(compartmentsList)
0137         cofactorsListWcompB{c, 1}= horzcat(cofactorsListB{i},compartmentsList{j});
0138         c = c + 1;
0139     <span class="keyword">end</span>
0140 <span class="keyword">end</span>
0141 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0142 
0143 
0144 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%%%%%% PART A %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0145 
0146 <span class="comment">% getting transport reactions</span>
0147 [nrxns,~] = size(model2.rxns);
0148 transRxns = cell(nrxns, 1);
0149 <span class="keyword">for</span> j = 1:size(model2.rxns)
0150     <span class="keyword">if</span> true(contains(model2.subSystems {j}, <span class="string">'S_Transport'</span>))
0151         transRxns{j, 1} = model2.rxns{j};
0152     <span class="keyword">end</span>
0153 <span class="keyword">end</span>
0154 transRxns = transRxns(~cellfun(<span class="string">'isempty'</span>,transRxns));
0155 
0156 <span class="comment">% getting subStystems without transport</span>
0157 SubSystems = (unique(getModelSubSystems(model2)));
0158 [nSubSystems, ~] = size(SubSystems);
0159 subSystemsNoTrans = cell(nSubSystems, 1);
0160 <span class="keyword">for</span> i = 1:nSubSystems
0161     <span class="keyword">if</span> ~contains(SubSystems{i,1},<span class="string">'S_Transport'</span>)&amp;&amp; <span class="keyword">...</span>
0162             ~contains(SubSystems{i, 1},<span class="string">'Extracellular exchange'</span>) &amp;&amp;<span class="keyword">...</span>
0163             ~contains(SubSystems{i, 1},<span class="string">'S_tRNA_charging'</span>)
0164         subSystemsNoTrans{i, 1} = SubSystems{i, 1};
0165     <span class="keyword">else</span>
0166         subSystemsNoTrans{i, 1} = {};
0167     <span class="keyword">end</span>
0168 <span class="keyword">end</span>
0169 subSystemsNoTrans = subSystemsNoTrans(~cellfun(<span class="string">'isempty'</span>,subSystemsNoTrans));
0170 
0171 <span class="comment">% defining subSystems of trasnport reactions to be colored, these</span>
0172 <span class="comment">% transport reactions complete reaction conections in subSystems which</span>
0173 <span class="comment">% require conexions among diferents compartments</span>
0174 [nTransRxns,~] = size(transRxns);
0175 [nSubSystemsNoTrans, ~] = size (subSystemsNoTrans);
0176 totalRxnsToIncludeA = cell(nTransRxns,1);
0177 <span class="keyword">for</span> i = 1:nSubSystemsNoTrans
0178     subSystemRxns = findRxnsFromSubSystem(model2, subSystemsNoTrans(i));
0179     metsSubSystemRxns = findMetsFromRxns(model2, subSystemRxns);
0180     [nMetsSubSystemRxns, ~] = size (metsSubSystemRxns);
0181     <span class="comment">% getting unique compartments</span>
0182     metsCompartment = cell(nMetsSubSystemRxns,1);
0183     <span class="keyword">for</span> j = 1:nMetsSubSystemRxns
0184         str = metsSubSystemRxns{j};
0185         [~, s] = size(str);
0186         metsCompartment{j,1} = str(s-1:s);
0187     <span class="keyword">end</span>
0188     metsCompartment = (unique(metsCompartment));
0189     [nUniqueMetsComp, ~] = size(metsCompartment);
0190     
0191     <span class="keyword">if</span> nUniqueMetsComp &gt; 1
0192         <span class="comment">% getting duplicated mets like 'pyr_c' and 'pyr_m' which are 'pyr'</span>
0193         <span class="comment">% and pyr after erasing the compartment signature</span>
0194         metsFromSubSystemRxnsNoComp = cell(nMetsSubSystemRxns,1);
0195         <span class="keyword">for</span> j = 1:nMetsSubSystemRxns
0196             str = metsSubSystemRxns{j};
0197             [z, s] = size(str);
0198             metsFromSubSystemRxnsNoComp{j, 1} = str(1:s-2); <span class="comment">% erasing comp signature</span>
0199         <span class="keyword">end</span>
0200         [~, o, ~] = unique(metsFromSubSystemRxnsNoComp,<span class="string">'first'</span>);
0201         indexToDupes = find(not(ismember(1:numel(metsFromSubSystemRxnsNoComp),o)));
0202         [~, nIndexToDupes] = size(indexToDupes);
0203         duplicatedMets = cell(nIndexToDupes,1);
0204         <span class="keyword">for</span> j = 1:nIndexToDupes
0205             duplicatedMets(j, 1) = metsFromSubSystemRxnsNoComp (indexToDupes(1, j));
0206         <span class="keyword">end</span>
0207         
0208         <span class="comment">% getting reactions from duplicated mets</span>
0209         dulicatedRxnMets = cell(nIndexToDupes * nUniqueMetsComp,1);
0210         c = 1; <span class="comment">% duplicatedRxnMets counter</span>
0211         <span class="keyword">for</span> j = 1:nIndexToDupes
0212             str = duplicatedMets{j, 1};
0213             <span class="keyword">if</span> ~ismember(str, cofactorsListA)
0214                 <span class="keyword">for</span> k = 1:nUniqueMetsComp
0215                     str1 = metsCompartment{k};
0216                     <span class="keyword">if</span> ~ismember(str1, compartmentsList(2)) <span class="comment">% compartment '_c'</span>
0217                         dulicatedRxnMets{c, 1} = {horzcat(str, <span class="string">'_c'</span>), horzcat(str, str1)};
0218                         c = c + 1;
0219                     <span class="keyword">end</span>
0220                 <span class="keyword">end</span>
0221             <span class="keyword">end</span>
0222         <span class="keyword">end</span>
0223         dulicatedRxnMets = dulicatedRxnMets(~cellfun(<span class="string">'isempty'</span>,dulicatedRxnMets));
0224         [ndulicatedRxnMets, ~] = size(dulicatedRxnMets);
0225         <span class="comment">% from duplicated reactions, get reactions to subscribe a reactions</span>
0226         <span class="comment">% subsystem from non transport subsystems</span>
0227         rxnsToInclude = cell(ndulicatedRxnMets, 1);
0228         <span class="keyword">for</span> j = 1:ndulicatedRxnMets
0229             metsToFindInTrasnportRxns = dulicatedRxnMets{j, 1};
0230             <span class="keyword">for</span> k = 1:nTransRxns
0231                 rxnFormula = printRxnFormula(model2, transRxns{k, 1}, false);
0232                 <span class="keyword">if</span> true(contains(string(rxnFormula), metsToFindInTrasnportRxns (1, 1)))
0233                     <span class="keyword">if</span> true(contains(rxnFormula, metsToFindInTrasnportRxns (1, 2)))
0234                         rxnsToInclude{j, 1} = transRxns{k, 1};
0235                     <span class="keyword">end</span>
0236                 <span class="keyword">end</span>
0237             <span class="keyword">end</span>
0238         <span class="keyword">end</span>
0239         rxnsToInclude = rxnsToInclude(~cellfun(<span class="string">'isempty'</span>,rxnsToInclude));
0240         rxnsToInclude = unique(rxnsToInclude);
0241         [s, ~] = size(rxnsToInclude);
0242         totalRxnsToIncludeA = [totalRxnsToIncludeA; rxnsToInclude];
0243         
0244         <span class="keyword">for</span> j = 1:s
0245             rxnID = findRxnIDs(model2, rxnsToInclude{j});
0246             model2.subSystems(rxnID) = {subSystemsNoTrans(i)};
0247         <span class="keyword">end</span>
0248         
0249     <span class="keyword">end</span>
0250 <span class="keyword">end</span>
0251 totalRxnsToIncludeA = totalRxnsToIncludeA(~cellfun(<span class="string">'isempty'</span>,totalRxnsToIncludeA));
0252 totalRxnsToIncludeA = unique(totalRxnsToIncludeA);
0253 
0254 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%%%%%% PART B %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0255 
0256 <span class="comment">% Get subsystems category. This is based on the number of conexions that a</span>
0257 <span class="comment">% subsystem includes in the genome scale metabolic model2</span>
0258 subsLength ={};
0259 <span class="keyword">for</span> i = 1:length(subSystemsNoTrans)
0260     rxnsSubs = findRxnsFromSubSystem(model2,subSystemsNoTrans(i));
0261     metsSubsRxns = findMetsFromRxns(model2, rxnsSubs);
0262     metsSubsRxnsNoCof = {};
0263     <span class="keyword">for</span> j = 1:length(metsSubsRxns)
0264         <span class="keyword">if</span> ~ismember(metsSubsRxns(j),cofactorsListWcompB)
0265             metsSubsRxnsNoCof{j,1} = char(metsSubsRxns(j));
0266         <span class="keyword">end</span>
0267     <span class="keyword">end</span>
0268     metsSubsRxnsNoCof = metsSubsRxnsNoCof(~cellfun(<span class="string">'isempty'</span>,metsSubsRxnsNoCof));
0269     
0270     metsRxns = findRxnsFromMets(model2, metsSubsRxnsNoCof);
0271     metsRxnsSubs = {};
0272     <span class="keyword">for</span> j = 1:length(metsRxns)
0273         rxnID = findRxnIDs(model2,metsRxns(j));
0274         metsRxnsSubs{j,1} = string(model2.subSystems(rxnID));
0275     <span class="keyword">end</span>
0276     metsRxnsSubs = unique(string(metsRxnsSubs));
0277     subsLength{i,1} = length(metsRxnsSubs);
0278     subsLength{i,2} = subSystemsNoTrans{i};
0279 <span class="keyword">end</span>
0280 
0281 <span class="comment">% Get  remaining transport reacctions, namely, reactions not included in</span>
0282 <span class="comment">% PART A</span>
0283 
0284 [nTransRxns, ~] = size(transRxns);
0285 remainingTransRxns = cell(nTransRxns, 1);
0286 <span class="keyword">for</span> i = 1:nTransRxns
0287     <span class="keyword">if</span> ~ismember(transRxns(i),totalRxnsToIncludeA)
0288         remainingTransRxns{i,1} = transRxns{i, 1};
0289     <span class="keyword">end</span>
0290 <span class="keyword">end</span>
0291 remainingTransRxns = remainingTransRxns(~cellfun(<span class="string">'isempty'</span>,remainingTransRxns));
0292 
0293 <span class="comment">% Assign remaining transport reactions</span>
0294 nonIncludSubs = {<span class="string">'S_tRNA_charging'</span>, <span class="string">'S_Other'</span>,<span class="keyword">...</span>
0295     <span class="string">'Biomass and maintenance functions'</span>, };
0296 totalRxnsToIncludeAB = totalRxnsToIncludeA;
0297 remTransRxnsWnewSubs = {};
0298 remTransRxnsSubs = {};
0299 rxnsToInclude = cell(length(remainingTransRxns), 1);
0300 subSToInclude = cell(length(remainingTransRxns), 1);
0301 <span class="keyword">for</span> i = 1:length(remainingTransRxns)
0302     metsRemTransRxns = findMetsFromRxns(model2, remainingTransRxns(i));
0303     metsRxns = {};
0304     <span class="keyword">for</span> j = 1:length(metsRemTransRxns)
0305         <span class="keyword">if</span> true(contains(metsRemTransRxns{j},<span class="string">'_c'</span>) &amp;&amp;<span class="keyword">...</span>
0306                 ~ismember(metsRemTransRxns(j), cofactorsListWcompB))
0307             metsRxns = findRxnsFromMets(model2,metsRemTransRxns(j));
0308         <span class="keyword">end</span>
0309     <span class="keyword">end</span>
0310     metsRxns = metsRxns(~cellfun(<span class="string">'isempty'</span>,metsRxns));
0311     
0312     rxnSubs = {};
0313     <span class="keyword">for</span> j = 1:length(metsRxns)
0314         rxnID = findRxnIDs(model2, metsRxns(j));
0315         rxnSubs{j,1} = model2.subSystems{rxnID};
0316     <span class="keyword">end</span>
0317     rxnSubs = rxnSubs(~cellfun(<span class="string">'isempty'</span>,rxnSubs));
0318     
0319     r = [];
0320     rankMax = 0;
0321     subSToUseStr = cell(1,1);
0322     <span class="keyword">for</span> j = 1:length(rxnSubs)
0323         <span class="keyword">if</span> true(~ismember(string(metsRxns(j)),totalRxnsToIncludeA))
0324             <span class="keyword">for</span> k = 1:length(subsLength)
0325                 <span class="keyword">if</span> true(and(ismember(rxnSubs{j}, subsLength{k,2}),<span class="keyword">...</span>
0326                         ~ismember(rxnSubs{j}, nonIncludSubs)))
0327                     r(k, 1) = subsLength{k, 1};
0328                     subSToUseStr{1, 1} = string(subsLength(k,2));
0329                     <span class="keyword">if</span> k &gt; 1
0330                         <span class="keyword">if</span> true(rankMax &lt; r(k, 1))
0331                             rankMax = r(k, 1);
0332                             subSToUseStr{1, 1} = string(subsLength(k,2));
0333                         <span class="keyword">end</span>
0334                     <span class="keyword">end</span>
0335                 <span class="keyword">end</span>
0336             <span class="keyword">end</span>
0337         <span class="keyword">end</span>
0338     <span class="keyword">end</span>
0339     <span class="keyword">if</span> ~isempty(subSToUseStr(1,1))
0340         rxnsToInclude{i} = remainingTransRxns{i};
0341         subSToInclude{i} = char(subSToUseStr{1, 1});
0342         remTransRxnsWnewSubs{i,1} = remainingTransRxns(i);
0343         remTransRxnsSubs{i,1} = subSToUseStr(1, 1);
0344         totalRxnsToIncludeAB = [totalRxnsToIncludeAB; rxnsToInclude];
0345     <span class="keyword">end</span>
0346 <span class="keyword">end</span>
0347 totalRxnsToIncludeAB = totalRxnsToIncludeAB(~cellfun(<span class="string">'isempty'</span>,totalRxnsToIncludeAB));
0348 totalRxnsToIncludeAB = unique(totalRxnsToIncludeAB);
0349 
0350 <span class="keyword">for</span> j = 1:length(rxnsToInclude)
0351     rxnID = findRxnIDs(model2, rxnsToInclude{j});
0352     model2.subSystems(rxnID) = {subSToInclude(j)};
0353 <span class="keyword">end</span>
0354 
0355 <span class="comment">%%%%%%%%%%%%%%%%%%%%%%%% ASIGN SUBSYSTEMS' COLOR %%%%%%%%%%%%%%%%%%%%%%%%%%%</span>
0356 
0357 <span class="comment">% Selection of reactions to include in the map</span>
0358 <span class="keyword">if</span> rxnsFluxTask == 1
0359     <span class="keyword">for</span> i = 1:length(model2.rxns)
0360         <span class="keyword">if</span> absFlxs(i)==0 <span class="comment">% zero flux reactions are not included</span>
0361             model2.subSystems(i) = {{<span class="string">'NonIncluded'</span>}};  <span class="comment">% included at the end</span>
0362         <span class="keyword">end</span>
0363     <span class="keyword">end</span>
0364 <span class="keyword">elseif</span> rxnsFluxTask == 2
0365     <span class="keyword">for</span> i = 1:length(model2.rxns)
0366         <span class="keyword">if</span> absFlxs(i) &gt; 0.95 * max(absFlxs)<span class="comment">% reactions causing growth (i.e. obj function) defect when deleting</span>
0367             model2.subSystems(i) = {{<span class="string">'NonIncluded'</span>}};   <span class="comment">% included at the end</span>
0368         <span class="keyword">end</span>
0369     <span class="keyword">end</span>
0370 <span class="keyword">elseif</span> rxnsFluxTask == 3
0371     <span class="keyword">for</span> i = 1:length(model2.rxns)
0372         <span class="keyword">if</span> absFlxs(i) &gt; 0.0 <span class="comment">% zero flux reactions only, i.e. essential reations</span>
0373             model2.subSystems(i) = {{<span class="string">'NonIncluded'</span>}};   <span class="comment">% included at the end</span>
0374         <span class="keyword">end</span>
0375     <span class="keyword">end</span>
0376 <span class="keyword">elseif</span> rxnsFluxTask == 4
0377     <span class="keyword">for</span> i = 1:length(model2.rxns)
0378         <span class="keyword">if</span> absFlxs(i) &lt; 0.95 * max(absFlxs)<span class="comment">% reactions fluxes causing no growth (i.e. obj function) defect when deleting</span>
0379             model2.subSystems(i) = {{<span class="string">'NonIncluded'</span>}};   <span class="comment">% included at the end</span>
0380         <span class="keyword">end</span>
0381     <span class="keyword">end</span>
0382 <span class="keyword">elseif</span> rxnsFluxTask == 5
0383     <span class="keyword">for</span> i = 1:length(model2.rxns)
0384         <span class="keyword">if</span> absFlxs(i) &lt; cutValue <span class="comment">% zero flux reactions are not included</span>
0385             model2.subSystems(i) = {{<span class="string">'NonIncluded'</span>}};  <span class="comment">% included at the end</span>
0386         <span class="keyword">end</span>
0387     <span class="keyword">end</span>
0388 <span class="keyword">end</span>
0389 
0390 <span class="comment">%See colors letter-code at:</span>
0391 <span class="comment">%la.mathworks.com/matlabcentral/mlc-downloads/downloads/submissions/24497/versions/2/screenshot.PNG</span>
0392 <span class="comment">%[xmlMap, mapMap] = transformXML2Map('YeastMapUnified1.0b.xml');</span>
0393 centralPathsList = {<span class="string">'S_GlycolysisGluconeogenesis'</span>, <span class="string">'S_Pentose_Phosphate_Pathway'</span>,<span class="keyword">...</span>
0394     <span class="string">'S_Pyruvate_Metabolism'</span>, <span class="string">'S_Citric_Acid_Cycle'</span>, <span class="string">'S_Taurine_Metabolism'</span>,<span class="keyword">...</span>
0395     <span class="string">'S_Starch_and_Sucrose_Metabolism'</span>, <span class="string">'S_Xylose_Metabolism'</span>,<span class="keyword">...</span>
0396     <span class="string">'S_Fructose_and_Mannose_Metabolism'</span>, <span class="string">'S_Galactose_metabolism'</span>,<span class="keyword">...</span>
0397     <span class="string">'S_Arabinose_Metabolism'</span>, <span class="string">'S_Alternate_Carbon_Metabolism'</span>, <span class="string">'S_Methane_Metabolism'</span>};
0398 aminoacPathsList = {<span class="string">'S_Glutamate_metabolism'</span>, <span class="string">'S_Glutamine_Metabolism'</span>,<span class="keyword">...</span>
0399     <span class="string">'S_Alanine_and_Aspartate_Metabolism'</span>, <span class="string">'S_Histidine_Metabolism'</span>,<span class="keyword">...</span>
0400     <span class="string">'S_Threonine_and_Lysine_Metabolism'</span>, <span class="string">'S_Glycine_and_Serine_Metabolism'</span>,<span class="keyword">...</span>
0401     <span class="string">'S_Tyrosine__Tryptophan__and_Phenylalanine_Metabolism'</span>,<span class="keyword">...</span>
0402     <span class="string">'S_Arginine_and_Proline_Metabolism'</span>, <span class="string">'S_Cysteine_Metabolism'</span>,<span class="keyword">...</span>
0403     <span class="string">'S_Methionine_Metabolism'</span>, <span class="string">'S_Valine__Leucine__and_Isoleucine_Metabolism'</span>,<span class="keyword">...</span>
0404     <span class="string">'S_Other_Amino_Acid_Metabolism'</span>, <span class="string">'S_Asparagine_metabolism'</span>, <span class="string">'S_Nitrogen_Metabolism'</span>};
0405 cofactorsPathsList = {<span class="string">'S_NAD_Biosynthesis'</span>, <span class="string">'S_Pantothenate_and_CoA_Biosynthesis'</span>,<span class="keyword">...</span>
0406     <span class="string">'S_Folate_Metabolism'</span>, <span class="string">'S_Riboflavin_Metabolism'</span>, <span class="string">'S_Quinone_Biosynthesis'</span>,<span class="keyword">...</span>
0407     <span class="string">'S_Porphyrin_and_Chlorophyll_Metabolism'</span>};
0408 lipidsPathsList = {<span class="string">'S_Fatty_Acid__Biosynthesis'</span>, <span class="string">'S_Fatty_Acid_Degradation'</span>,<span class="keyword">...</span>
0409     <span class="string">'S_Fatty_Acid_Metabolism'</span>, <span class="string">'S_Glycerolipid_Metabolism'</span>, <span class="string">'S_Glycoprotein_Metabolism'</span>,<span class="keyword">...</span>
0410     <span class="string">'S_Phospholipid_Biosynthesis'</span>, <span class="string">'S_Phospholipid_Metabolism'</span>,<span class="keyword">...</span>
0411     <span class="string">'S_Sphingolipid_Metabolism'</span>, <span class="string">'S_Sterol_Metabolism'</span>};
0412 nucleotidsPathsList = {<span class="string">'S_Purine_and_Pyrimidine_Biosynthesis'</span>,<span class="keyword">...</span>
0413     <span class="string">'S_Nucleotide_Salvage_Pathway'</span>, <span class="string">'S_Thiamine_Metabolism'</span>, <span class="string">'S_Pyridoxine_Metabolism'</span>};
0414 enrgyPathsList = {<span class="string">'Biomass and maintenance functions'</span>, <span class="string">'S_Anaplerotic_reactions'</span>,<span class="keyword">...</span>
0415     <span class="string">'S_Oxidative_Phosphorylation'</span>};
0416 colors = createColorsMap;
0417 
0418 <span class="keyword">if</span> isempty(rxnLineColor)||rxnLineColor == &quot;Pathways&quot;
0419     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_GlycolysisGluconeogenesis'</span>, <span class="string">'DARKBLUE'</span>);
0420     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Pentose_Phosphate_Pathway'</span>, <span class="string">'STEELBLUE'</span>);
0421     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Pyruvate_Metabolism'</span>, <span class="string">'DODGERBLUE'</span>);
0422     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Citric_Acid_Cycle'</span>, <span class="string">'MEDIUMBLUE'</span>);
0423     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Taurine_Metabolism'</span>, <span class="string">'MEDIUMBLUE'</span>);
0424     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Starch_and_Sucrose_Metabolism'</span>, <span class="string">'BLUE'</span>);
0425     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Xylose_Metabolism'</span>, <span class="string">'STEELBLUE'</span>);
0426     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Fructose_and_Mannose_Metabolism'</span>, <span class="string">'DODGERBLUE'</span>);
0427     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Galactose_metabolism'</span>, <span class="string">'MEDIUMBLUE'</span>);
0428     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Arabinose_Metabolism'</span>, <span class="string">'BLUE'</span>);
0429     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Alternate_Carbon_Metabolism'</span>, <span class="string">'DARKBLUE'</span>);
0430     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Methane_Metabolism'</span>, <span class="string">'DARKBLUE'</span>);
0431     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Taurine_Metabolism'</span>, <span class="string">'DARKBLUE'</span>);
0432     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Anaplerotic_reactions'</span>, <span class="string">'CRIMSON'</span>);
0433     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Complex_Alcohol_Metabolism'</span>, <span class="string">'MAROON'</span>);
0434     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Oxidative_Phosphorylation'</span>, <span class="string">'RED'</span>);
0435     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Purine_and_Pyrimidine_Biosynthesis'</span>, <span class="string">'PINK'</span>);
0436     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Nucleotide_Salvage_Pathway'</span>, <span class="string">'HOTPINK'</span>);
0437     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Thiamine_Metabolism'</span>, <span class="string">'MAGENTA'</span>);
0438     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Pyridoxine_Metabolism'</span>, <span class="string">'PINK'</span>);
0439     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_NAD_Biosynthesis'</span>, <span class="string">'PLUM'</span>);
0440     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Pantothenate_and_CoA_Biosynthesis'</span>, <span class="string">'VIOLET'</span>);
0441     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Folate_Metabolism'</span>, <span class="string">'DARKORCHID'</span>);
0442     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Riboflavin_Metabolism'</span>, <span class="string">'PURPLE'</span>);
0443     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Quinone_Biosynthesis'</span>, <span class="string">'MEDIUMORCHID'</span>);
0444     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Porphyrin_and_Chlorophyll_Metabolism'</span>, <span class="string">'PLUM'</span>);
0445     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Glutamate_metabolism'</span>, <span class="string">'ORANGE'</span>);
0446     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Glutamine_Metabolism'</span>, <span class="string">'DARKORANGE'</span>);
0447     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Alanine_and_Aspartate_Metabolism'</span>, <span class="string">'CORAL'</span>);
0448     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Histidine_Metabolism'</span>, <span class="string">'GOLD'</span>);
0449     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Threonine_and_Lysine_Metabolism'</span>, <span class="string">'ORANGE'</span>);
0450     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Glycine_and_Serine_Metabolism'</span>, <span class="string">'DARKORANGE'</span>);
0451     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Tyrosine__Tryptophan__and_Phenylalanine_Metabolism'</span>, <span class="string">'CORAL'</span>);
0452     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Arginine_and_Proline_Metabolism'</span>, <span class="string">'GOLD'</span>);
0453     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Cysteine_Metabolism'</span>, <span class="string">'ORANGE'</span>);
0454     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Methionine_Metabolism'</span>, <span class="string">'DARKORANGE'</span>);
0455     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Valine__Leucine__and_Isoleucine_Metabolism'</span>, <span class="string">'CORAL'</span>);
0456     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Other_Amino_Acid_Metabolism'</span>, <span class="string">'GOLD'</span>);
0457     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Asparagine_metabolism'</span>, <span class="string">'ORANGE'</span>);
0458     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Nitrogen_Metabolism'</span>, <span class="string">'DARKORANGE'</span>);
0459     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Fatty_Acid__Biosynthesis'</span>, <span class="string">'FORESTGREEN'</span>);
0460     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Fatty_Acid_Degradation'</span>, <span class="string">'MEDIUMSEAGREEN'</span>);
0461     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Fatty_Acid_Metabolism'</span>, <span class="string">'DARKGREEN'</span>);
0462     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Glycerolipid_Metabolism'</span>, <span class="string">'SEAGREEN'</span>);
0463     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Glycoprotein_Metabolism'</span>, <span class="string">'FORESTGREEN'</span>);
0464     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Phospholipid_Biosynthesis'</span>, <span class="string">'MEDIUMSEAGREEN'</span>);
0465     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Phospholipid_Metabolism'</span>, <span class="string">'DARKGREEN'</span>);
0466     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Sphingolipid_Metabolism'</span>, <span class="string">'SEAGREEN'</span>);
0467     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Sterol_Metabolism'</span>, <span class="string">'FORESTGREEN'</span>);
0468     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Other'</span>, <span class="string">'BLACK'</span>);
0469     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'Biomass and maintenance functions'</span>, <span class="string">'RED'</span>);
0470 <span class="keyword">elseif</span> rxnLineColor == &quot;Subsystems&quot;
0471     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, centralPathsList, <span class="string">'BLUE'</span>); <span class="comment">%mapMap to fluxMap</span>
0472     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, aminoacPathsList, <span class="string">'ORANGE'</span>);
0473     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, cofactorsPathsList, <span class="string">'ORCHID'</span>);
0474     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, lipidsPathsList, <span class="string">'LIME'</span>);
0475     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, nucleotidsPathsList, <span class="string">'MAGENTA'</span>);
0476     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, enrgyPathsList, <span class="string">'RED'</span>);
0477     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Complex_Alcohol_Metabolism'</span>, <span class="string">'GOLDENROD'</span>);
0478     modifiedMap = <a href="colorSubsystem.html" class="code" title="function newMap = colorSubsystem(map, model, subsystem, color)">colorSubsystem</a>(modifiedMap, model2, <span class="string">'S_Other'</span>, <span class="string">'GOLDENROD'</span>);
0479 <span class="keyword">elseif</span> ~(rxnLineColor == &quot;Subsystems&quot; || rxnLineColor == &quot;Payhways&quot; || isempty(rxnLineColor))
0480     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0481         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0482         <span class="keyword">if</span> ~isempty(a) &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0483             modifiedMap.rxnColor(i, 1) = {colors(rxnLineColor)};
0484         <span class="keyword">end</span>
0485     <span class="keyword">end</span>
0486 <span class="keyword">end</span>
0487 <span class="keyword">if</span> rxnLineColor == &quot;Subsystems&quot; || rxnLineColor == &quot;Payhways&quot; || isempty(rxnLineColor)
0488     h2o = {<span class="string">'H2Ot'</span>, <span class="string">'H2Otm'</span>,    <span class="string">'H2Otn'</span>, <span class="string">'H2Otv'</span>, <span class="string">'H2Otp'</span>, <span class="string">'H2Oter'</span>};
0489     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0490         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0491         <span class="keyword">for</span> j = 1:length(h2o)
0492             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(h2o(j))<span class="keyword">...</span>
0493                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0494                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'CYAN'</span>)};
0495             <span class="keyword">end</span>
0496         <span class="keyword">end</span>
0497     <span class="keyword">end</span>
0498     o2 = {<span class="string">'O2t'</span>, <span class="string">'O2tm'</span>, <span class="string">'O2ter'</span>, <span class="string">'O2tp'</span>};
0499     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0500         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0501         <span class="keyword">for</span> j = 1:length(o2)
0502             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(o2(j))<span class="keyword">...</span>
0503                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0504                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'CYAN'</span>)};
0505             <span class="keyword">end</span>
0506         <span class="keyword">end</span>
0507     <span class="keyword">end</span>
0508     Pi = {<span class="string">'PIt2r'</span>, <span class="string">'PIt2m'</span>, <span class="string">'PIt5m'</span>, <span class="string">'PIt2n'</span>, <span class="string">'PIt2v'</span>, <span class="string">'PIt2p'</span>, <span class="string">'ATPtm_H'</span>, <span class="string">'ATPtp_H'</span>, <span class="keyword">...</span>
0509         <span class="string">'ATP2tp_H'</span>, <span class="string">'ATPS3g'</span>, <span class="string">'ATPS3v'</span>, <span class="string">'UTPtm'</span>, <span class="string">'CTPtm'</span>, <span class="string">'GTPt2m'</span>, <span class="string">'ATPS'</span>};
0510     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0511         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0512         <span class="keyword">for</span> j = 1:length(Pi)
0513             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(Pi(j))<span class="keyword">...</span>
0514                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0515                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'RED'</span>)};
0516             <span class="keyword">end</span>
0517         <span class="keyword">end</span>
0518     <span class="keyword">end</span>
0519     co2 = {<span class="string">'CO2t'</span>, <span class="string">'CO2tg'</span>, <span class="string">'CO2tm'</span>, <span class="string">'CO2tn'</span>, <span class="string">'HCO3tn'</span>, <span class="string">'CO2tv'</span>, <span class="string">'CO2tp'</span>};
0520     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0521         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0522         <span class="keyword">for</span> j = 1:length(co2)
0523             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(co2(j))<span class="keyword">...</span>
0524                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0525                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'SLATEGRAY'</span>)};
0526             <span class="keyword">end</span>
0527         <span class="keyword">end</span>
0528     <span class="keyword">end</span>
0529     aa = {<span class="string">'MMETt2'</span>, <span class="string">'GLUt2r'</span>, <span class="string">'GLNt2r'</span>, <span class="string">'GLNt7'</span>, <span class="string">'GLUt7'</span>, <span class="string">'LCYSTintv'</span>, <span class="string">'NH4t'</span>, <span class="string">'TCHOLAabcv'</span>};
0530     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0531         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0532         <span class="keyword">for</span> j = 1:length(aa)
0533             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(aa(j))<span class="keyword">...</span>
0534                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0535                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'ORANGE'</span>)};
0536             <span class="keyword">end</span>
0537         <span class="keyword">end</span>
0538     <span class="keyword">end</span>
0539     cenMet = {<span class="string">'COAtim'</span>, <span class="string">'NAt3_1'</span>, <span class="string">'MALS'</span>};
0540     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0541         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0542         <span class="keyword">for</span> j = 1:length(cenMet)
0543             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(cenMet(j))<span class="keyword">...</span>
0544                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0545                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'BLUE'</span>)};
0546             <span class="keyword">end</span>
0547         <span class="keyword">end</span>
0548     <span class="keyword">end</span>
0549     lip = {<span class="string">'G3PIt'</span>};
0550     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0551         a = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0552         <span class="keyword">for</span> j = 1:length(lip)
0553             <span class="keyword">if</span> ~isempty(a) &amp; string(modifiedMap.rxnName{i}) == string(lip(j))<span class="keyword">...</span>
0554                     &amp; string(model2.subSystems(a)) ~= &quot;NonIncluded&quot;
0555                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'LIME'</span>)};
0556             <span class="keyword">end</span>
0557         <span class="keyword">end</span>
0558     <span class="keyword">end</span>
0559 <span class="keyword">end</span>
0560 
0561 <span class="keyword">if</span> rxnsFluxTask == 5 <span class="comment">% change color to dark-color in negative values</span>
0562     <span class="keyword">for</span> i = 1:length(modifiedMap.rxnName)
0563         valueskeys = values(colors)';
0564         valueskeys (:,2) = keys(colors)';
0565         a = find(ismember(valueskeys, modifiedMap.rxnColor{i, 1}));
0566         b = find(ismember(model2.rxns, modifiedMap.rxnName{i}));
0567         <span class="keyword">if</span> ~isempty(b) &amp; v1(b) &lt; 0 &amp; string(model2.subSystems(b)) ~= &quot;NonIncluded&quot;
0568             <span class="keyword">if</span> string(valueskeys (a,2)) == <span class="string">'LIME'</span>
0569                 modifiedMap.rxnColor(i, 1) = {colors(<span class="string">'GREEN'</span>)};
0570             <span class="keyword">else</span>
0571                 rxnColor =strcat(<span class="string">'DARK'</span>, char(valueskeys (a(1,1),2)));
0572                 <span class="keyword">if</span> string(rxnColor) == <span class="string">'DARKAQUA'</span>
0573                     rxnColor = <span class="string">'DARKCYAN'</span>;
0574                 <span class="keyword">elseif</span> string(rxnColor) == <span class="string">'DARKFUCHSIA'</span>
0575                     rxnColor = <span class="string">'DARKMAGENTA'</span>;
0576                 <span class="keyword">end</span>
0577                 modifiedMap.rxnColor(i, 1) = {colors(rxnColor)};
0578             <span class="keyword">end</span>
0579         <span class="keyword">end</span>
0580     <span class="keyword">end</span>
0581 <span class="keyword">end</span>
0582 
0583 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>