<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getExpressionStructure</title>
  <meta name="keywords" content="getExpressionStructure">
  <meta name="description" content="getExpressionStructure">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">core</a> &gt; getExpressionStructure.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for core&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getExpressionStructure
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getExpressionStructure</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function experiment=getExpressionStructure(fileName) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getExpressionStructure
   Loads a representation of an experiment from an Excel file (see
   comments further down)

   fileName            an Excel representation on an experiment

   experiment          an experiment structure
       data            matrix with expression values
       orfs            the corresponding ORFs
       experiments     the titles of the experiments
       boundNames      reaction names for the bounds
       upperBoundaries matrix with the upper bound values
       fitNames        reaction names for the measured fluxes
       fitTo           matrix with the measured fluxes

   A very common data set when working with genome-scale metabolic models
   is that you have measured fermentation data, gene expression data,
   and some different 'bounds' (for example different carbon sources
   or genes that are knocked out) in a number of conditions. This function
   reads an Excel representation of such an experiment.
   The Excel file must contain three sheets, 'EXPRESSION', 'BOUNDS',
   'FITTING'. Below are some examples to show how they should be
   formatted:

   -EXPRESSION
       ORF    dsm_paa    wisc_paa
       Pc00e00030    79.80942723    78.14755338
   Shows the expression of the gene Pc00e00030 under two different
   conditions (in this case a DSM strain and a Wisconsin strain of P.
   chrysogenum with PSS in the media)

   -BOUNDS
       Fixed Upper    dsm_paa    wisc_paa
       paaIN    0.1    0.2
   The upper bound for the reaction paaIN should be 0.1 for the first
   condition and 0.2 for the second

   -FITTING
       Fit to    dsm_paa    wisc_paa
       co2OUT    2.85    3.05
       glcIN   1.2     0.9
   The measured fluxes for CO2 production and glucose uptake for the two
   conditions. The model(s) can later be fitted to match these values as
   good as possible.

   Usage: experiment=getExpressionStructure(fileName)

   Rasmus Agren, 2014-01-08</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function experiment=getExpressionStructure(fileName)</a>
0002 <span class="comment">% getExpressionStructure</span>
0003 <span class="comment">%   Loads a representation of an experiment from an Excel file (see</span>
0004 <span class="comment">%   comments further down)</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   fileName            an Excel representation on an experiment</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   experiment          an experiment structure</span>
0009 <span class="comment">%       data            matrix with expression values</span>
0010 <span class="comment">%       orfs            the corresponding ORFs</span>
0011 <span class="comment">%       experiments     the titles of the experiments</span>
0012 <span class="comment">%       boundNames      reaction names for the bounds</span>
0013 <span class="comment">%       upperBoundaries matrix with the upper bound values</span>
0014 <span class="comment">%       fitNames        reaction names for the measured fluxes</span>
0015 <span class="comment">%       fitTo           matrix with the measured fluxes</span>
0016 <span class="comment">%</span>
0017 <span class="comment">%   A very common data set when working with genome-scale metabolic models</span>
0018 <span class="comment">%   is that you have measured fermentation data, gene expression data,</span>
0019 <span class="comment">%   and some different 'bounds' (for example different carbon sources</span>
0020 <span class="comment">%   or genes that are knocked out) in a number of conditions. This function</span>
0021 <span class="comment">%   reads an Excel representation of such an experiment.</span>
0022 <span class="comment">%   The Excel file must contain three sheets, 'EXPRESSION', 'BOUNDS',</span>
0023 <span class="comment">%   'FITTING'. Below are some examples to show how they should be</span>
0024 <span class="comment">%   formatted:</span>
0025 <span class="comment">%</span>
0026 <span class="comment">%   -EXPRESSION</span>
0027 <span class="comment">%       ORF    dsm_paa    wisc_paa</span>
0028 <span class="comment">%       Pc00e00030    79.80942723    78.14755338</span>
0029 <span class="comment">%   Shows the expression of the gene Pc00e00030 under two different</span>
0030 <span class="comment">%   conditions (in this case a DSM strain and a Wisconsin strain of P.</span>
0031 <span class="comment">%   chrysogenum with PSS in the media)</span>
0032 <span class="comment">%</span>
0033 <span class="comment">%   -BOUNDS</span>
0034 <span class="comment">%       Fixed Upper    dsm_paa    wisc_paa</span>
0035 <span class="comment">%       paaIN    0.1    0.2</span>
0036 <span class="comment">%   The upper bound for the reaction paaIN should be 0.1 for the first</span>
0037 <span class="comment">%   condition and 0.2 for the second</span>
0038 <span class="comment">%</span>
0039 <span class="comment">%   -FITTING</span>
0040 <span class="comment">%       Fit to    dsm_paa    wisc_paa</span>
0041 <span class="comment">%       co2OUT    2.85    3.05</span>
0042 <span class="comment">%       glcIN   1.2     0.9</span>
0043 <span class="comment">%   The measured fluxes for CO2 production and glucose uptake for the two</span>
0044 <span class="comment">%   conditions. The model(s) can later be fitted to match these values as</span>
0045 <span class="comment">%   good as possible.</span>
0046 <span class="comment">%</span>
0047 <span class="comment">%   Usage: experiment=getExpressionStructure(fileName)</span>
0048 <span class="comment">%</span>
0049 <span class="comment">%   Rasmus Agren, 2014-01-08</span>
0050 <span class="comment">%</span>
0051 
0052 [type, sheets]=xlsfinfo(fileName);
0053 
0054 <span class="comment">%Check if the file is a Microsoft Excel Spreadsheet</span>
0055 <span class="keyword">if</span> ~strcmp(type,<span class="string">'Microsoft Excel Spreadsheet'</span>)
0056     EM=<span class="string">'The file is not a Microsoft Excel Spreadsheet'</span>;
0057     <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0058 <span class="keyword">end</span>
0059 
0060 <span class="comment">%Check that all sheets are present and saves the index of each</span>
0061 exprSheet=find(strcmp(<span class="string">'EXPRESSION'</span>, sheets));
0062 boundSheet=find(strcmp(<span class="string">'BOUNDS'</span>, sheets));
0063 fitSheet=find(strcmp(<span class="string">'FITTING'</span>, sheets));
0064 
0065 <span class="keyword">if</span> length(exprSheet)~=1 || length(boundSheet)~=1 || length(fitSheet)~=1
0066     EM=<span class="string">'Not all required spreadsheets are present in the file'</span>;
0067     <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0068 <span class="keyword">end</span>
0069 
0070 <span class="comment">%Load the expression data</span>
0071 [discard,dataSheet]=xlsread(fileName,exprSheet);
0072 experiment.data=discard;
0073 experiment.orfs=dataSheet(2:size(dataSheet,1),1);
0074 experiment.experiments=dataSheet(1,2:size(dataSheet,2));
0075 
0076 <span class="comment">%Loads the maximal boundaries</span>
0077 [discard,dataSheet]=xlsread(fileName,boundSheet);
0078 experiment.boundNames=dataSheet(2:size(dataSheet,1),1);
0079 experiment.upperBoundaries=discard;
0080 
0081 <span class="comment">%Loads the experimental data to fit to</span>
0082 [discard,dataSheet]=xlsread(fileName,fitSheet);
0083 experiment.fitNames=dataSheet(2:size(dataSheet,1),1);
0084 experiment.fitTo=discard;
0085 
0086 <span class="comment">%Check to see that the dimensions are correct</span>
0087 <span class="keyword">if</span> length(experiment.orfs)~=size(experiment.data,1) || (length(experiment.experiments)~=size(experiment.data,2) &amp;&amp; ~isempty(experiment.data))
0088     EM=<span class="string">'The expression data does not seem to be formated in the expected manner'</span>;
0089     <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0090 <span class="keyword">end</span>
0091 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Wed 14-Mar-2018 21:08:28 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>