<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromHomology</title>
  <meta name="keywords" content="getModelFromHomology">
  <meta name="description" content="getModelFromHomology">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">core</a> &gt; getModelFromHomology.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for core&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getModelFromHomology
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getModelFromHomology</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function draftModel=getModelFromHomology(models,blastStructure,getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,minLen,minSim,mapNewGenesToOld) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromHomology
   Constructs a new model from a set of existing models and gene
   homology information.

   models            a cell array of model structures to build the model
                     from
   blastStructure    a blastStructure as produced by getBlast or
                     getBlastFromExcel
   getModelFor       the name of the organism to build a model for. Must
                     have hits in both directions to the organisms in models
   preferredOrder    the order in which reactions should be added from the
                     models. If not supplied, reactions will be included from
                     all models, otherwise one gene will only result in reactions
                     from one model (opt, default {})
   strictness        integer that specifies which reactions that should be included
                     1: Include only 1-1 orthologs (only include genes that
                     map back to the original gene in the blast in the
                     opposite direction)
                     2: Include the reactions for all genes below the cutoff
                     3: Include only best 1-1 orthologs (opt, default 1)
   onlyGenesInModels blast only against genes that exists in the models.
                     This tends to import a larger fraction from the existing
                     models but may give less reliable results. Has effect only
                     if strictness=3 (opt, default false)
   maxE              only look at genes with E-values &lt;= this value (opt, default 10^-30)
   minLen            only look at genes with overlap &gt;= this value (opt,
                     default 200)
   minSim            only look at genes with similarity &gt;= this value (opt,
                     default 40 (%))
   mapNewGenesToOld  determines how to match genes if not looking at only
                     1-1 orthologs. Either map the new genes to the old or
                     old genes to new. The default is to map the new genes
                     (opt, default true)

   draftModel        a model structure for the new organism

   The models in the models structure should have named the metabolites in
   the same manner, have their reversible reactions in the
   same direction (run sortModel), and use the same compartment names.
   To avoid keeping unneccesary old genes, the models should not have
   'or'-relations in their grRules (use expandModel).

   Returns a model structure for the new organism.

   draftModel        a new model structure

   Simonas Marcisauskas, 2018-03-17</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li><li><a href="mergeModels.html" class="code" title="function model=mergeModels(models,supressWarnings)">mergeModels</a>	mergeModels</li><li><a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>	removeGenes</li><li><a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>	removeReactions</li><li><a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>	standardizeGrRules</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function draftModel=getModelFromHomology(models,blastStructure,getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,minLen,minSim,mapNewGenesToOld)</a>
0002 <span class="comment">% getModelFromHomology</span>
0003 <span class="comment">%   Constructs a new model from a set of existing models and gene</span>
0004 <span class="comment">%   homology information.</span>
0005 <span class="comment">%</span>
0006 <span class="comment">%   models            a cell array of model structures to build the model</span>
0007 <span class="comment">%                     from</span>
0008 <span class="comment">%   blastStructure    a blastStructure as produced by getBlast or</span>
0009 <span class="comment">%                     getBlastFromExcel</span>
0010 <span class="comment">%   getModelFor       the name of the organism to build a model for. Must</span>
0011 <span class="comment">%                     have hits in both directions to the organisms in models</span>
0012 <span class="comment">%   preferredOrder    the order in which reactions should be added from the</span>
0013 <span class="comment">%                     models. If not supplied, reactions will be included from</span>
0014 <span class="comment">%                     all models, otherwise one gene will only result in reactions</span>
0015 <span class="comment">%                     from one model (opt, default {})</span>
0016 <span class="comment">%   strictness        integer that specifies which reactions that should be included</span>
0017 <span class="comment">%                     1: Include only 1-1 orthologs (only include genes that</span>
0018 <span class="comment">%                     map back to the original gene in the blast in the</span>
0019 <span class="comment">%                     opposite direction)</span>
0020 <span class="comment">%                     2: Include the reactions for all genes below the cutoff</span>
0021 <span class="comment">%                     3: Include only best 1-1 orthologs (opt, default 1)</span>
0022 <span class="comment">%   onlyGenesInModels blast only against genes that exists in the models.</span>
0023 <span class="comment">%                     This tends to import a larger fraction from the existing</span>
0024 <span class="comment">%                     models but may give less reliable results. Has effect only</span>
0025 <span class="comment">%                     if strictness=3 (opt, default false)</span>
0026 <span class="comment">%   maxE              only look at genes with E-values &lt;= this value (opt, default 10^-30)</span>
0027 <span class="comment">%   minLen            only look at genes with overlap &gt;= this value (opt,</span>
0028 <span class="comment">%                     default 200)</span>
0029 <span class="comment">%   minSim            only look at genes with similarity &gt;= this value (opt,</span>
0030 <span class="comment">%                     default 40 (%))</span>
0031 <span class="comment">%   mapNewGenesToOld  determines how to match genes if not looking at only</span>
0032 <span class="comment">%                     1-1 orthologs. Either map the new genes to the old or</span>
0033 <span class="comment">%                     old genes to new. The default is to map the new genes</span>
0034 <span class="comment">%                     (opt, default true)</span>
0035 <span class="comment">%</span>
0036 <span class="comment">%   draftModel        a model structure for the new organism</span>
0037 <span class="comment">%</span>
0038 <span class="comment">%   The models in the models structure should have named the metabolites in</span>
0039 <span class="comment">%   the same manner, have their reversible reactions in the</span>
0040 <span class="comment">%   same direction (run sortModel), and use the same compartment names.</span>
0041 <span class="comment">%   To avoid keeping unneccesary old genes, the models should not have</span>
0042 <span class="comment">%   'or'-relations in their grRules (use expandModel).</span>
0043 <span class="comment">%</span>
0044 <span class="comment">%   Returns a model structure for the new organism.</span>
0045 <span class="comment">%</span>
0046 <span class="comment">%   draftModel        a new model structure</span>
0047 <span class="comment">%</span>
0048 <span class="comment">%   Simonas Marcisauskas, 2018-03-17</span>
0049 <span class="comment">%</span>
0050 
0051 <span class="comment">%NOTE: &quot;to&quot; and &quot;from&quot; means relative the new organism</span>
0052 
0053 <span class="keyword">if</span> nargin&lt;4
0054     preferredOrder=[];
0055 <span class="keyword">end</span>
0056 <span class="keyword">if</span> nargin&lt;5
0057     strictness=1;
0058 <span class="keyword">end</span>
0059 <span class="keyword">if</span> nargin&lt;6
0060     onlyGenesInModels=false;
0061 <span class="keyword">end</span>
0062 <span class="keyword">if</span> nargin&lt;7
0063     maxE=10^-30;
0064 <span class="keyword">end</span>
0065 <span class="keyword">if</span> nargin&lt;8
0066     minLen=200;
0067 <span class="keyword">end</span>
0068 <span class="keyword">if</span> nargin&lt;9
0069     minSim=40;
0070 <span class="keyword">end</span>
0071 <span class="keyword">if</span> nargin&lt;10
0072     mapNewGenesToOld=true;
0073 <span class="keyword">end</span>
0074 
0075 preferredOrder=preferredOrder(:);
0076 
0077 <span class="comment">%Check that all the information is in the blast structure</span>
0078 modelNames=cell(numel(models),1);
0079 <span class="keyword">for</span> i=1:numel(models)
0080     modelNames{i}=models{i}.id;
0081 <span class="keyword">end</span>
0082 
0083 <span class="comment">%Assume for now that all information is there and that it's correct This is</span>
0084 <span class="comment">%important to fix since no further checks are being made!</span>
0085 
0086 <span class="comment">%Remove all gene matches that are below the cutoffs</span>
0087 <span class="keyword">for</span> i=1:numel(blastStructure)
0088     indexes=blastStructure(i).evalue&lt;maxE &amp; blastStructure(i).aligLen&gt;=minLen &amp; blastStructure(i).identity&gt;=minSim; <span class="comment">%Do it in this direction to lose NaNs</span>
0089     blastStructure(i).fromGenes(~indexes)=[];
0090     blastStructure(i).toGenes(~indexes)=[];
0091     blastStructure(i).evalue(~indexes)=[];
0092     blastStructure(i).identity(~indexes)=[];
0093     blastStructure(i).aligLen(~indexes)=[];
0094     blastStructure(i).bitscore(~indexes)=[];
0095     blastStructure(i).ppos(~indexes)=[];
0096 <span class="keyword">end</span>
0097 
0098 <span class="comment">%Remove all reactions from the models that have no genes encoding for them.</span>
0099 <span class="comment">%Also remove all genes that encode for no reactions. There shouldn't be any</span>
0100 <span class="comment">%but there might be mistakes</span>
0101 <span class="keyword">for</span> i=1:numel(models)
0102     [hasGenes, ~]=find(models{i}.rxnGeneMat);
0103     hasNoGenes=1:numel(models{i}.rxns);
0104     hasNoGenes(hasGenes)=[];
0105     
0106     models{i}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{i},hasNoGenes,true,true);
0107 <span class="keyword">end</span>
0108 
0109 <span class="comment">%Create a structure that contains all genes used in the blasts in any</span>
0110 <span class="comment">%direction for each of the models in models and for the new organism The</span>
0111 <span class="comment">%first cell is for the new organism and then according to the preferred</span>
0112 <span class="comment">%order. If no such order is supplied, then according to the order in</span>
0113 <span class="comment">%models.</span>
0114 allGenes=cell(numel(models)+1,1);
0115 <span class="keyword">if</span> isempty(preferredOrder)
0116     useOrder=modelNames;
0117 <span class="keyword">else</span>
0118     useOrder=preferredOrder;
0119 <span class="keyword">end</span>
0120 
0121 <span class="comment">%Get the corresponding indexes for those models in the models structure</span>
0122 useOrderIndexes=zeros(numel(models),1);
0123 <span class="keyword">for</span> i=1:numel(models)
0124     [~, index]=ismember(models{i}.id,useOrder);
0125     useOrderIndexes(index)=i;
0126 <span class="keyword">end</span>
0127 
0128 <span class="comment">%Remove all genes from the blast structure that have no genes in the models</span>
0129 <span class="keyword">if</span> onlyGenesInModels==true
0130     modelGenes={};
0131     <span class="keyword">for</span> i=1:numel(models)
0132         modelGenes=[modelGenes;models{i}.genes];
0133     <span class="keyword">end</span>
0134     <span class="keyword">for</span> i=1:numel(blastStructure)
0135         <span class="comment">%Check to see if it should match the toId or fromId</span>
0136         <span class="keyword">if</span> strcmpi(blastStructure(i).fromId,getModelFor)
0137             I=ismember(blastStructure(i).toGenes,modelGenes);
0138         <span class="keyword">else</span>
0139             I=ismember(blastStructure(i).fromGenes,modelGenes);
0140         <span class="keyword">end</span>
0141         blastStructure(i).fromGenes(~indexes)=[];
0142         blastStructure(i).toGenes(~indexes)=[];
0143         blastStructure(i).evalue(~indexes)=[];
0144         blastStructure(i).identity(~indexes)=[];
0145         blastStructure(i).aligLen(~indexes)=[];
0146         blastStructure(i).bitscore(~indexes)=[];
0147         blastStructure(i).ppos(~indexes)=[];
0148         
0149         <span class="comment">%Check that no matching in blastStructure is empty. This happens if</span>
0150         <span class="comment">%no genes in the models are present in the corresponding sheet</span>
0151         <span class="keyword">if</span> isempty(blastStructure(i).fromGenes)
0152             EM=[<span class="string">'No genes in matching from '</span> blastStructure(i).fromId <span class="string">' to '</span> blastStructure(i).toId <span class="string">' are present in the corresponding model'</span>];
0153             <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0154         <span class="keyword">end</span>
0155     <span class="keyword">end</span>
0156 <span class="keyword">end</span>
0157 
0158 <span class="comment">%If only best 1-1 orthologs are to be used then all other measurements are</span>
0159 <span class="comment">%deleted from the blastStructure. All code after this stays the same. This</span>
0160 <span class="comment">%means that preferred order can still matter. The best ortholog scoring is</span>
0161 <span class="comment">%based only on the E-value</span>
0162 <span class="keyword">if</span> strictness==3
0163     <span class="keyword">for</span> i=1:numel(blastStructure)
0164         keep=false(numel(blastStructure(i).toGenes),1);
0165         [allFromGenes, ~, I]=unique(blastStructure(i).fromGenes);
0166         
0167         <span class="comment">%It would be nice to get rid of this loop</span>
0168         <span class="keyword">for</span> j=1:numel(allFromGenes)
0169             allMatches=find(I==j);
0170             bestMatches=allMatches(blastStructure(i).evalue(allMatches)==min(blastStructure(i).evalue(allMatches)));
0171             
0172             <span class="comment">%Keep the best matches</span>
0173             keep(bestMatches)=true;
0174         <span class="keyword">end</span>
0175         
0176         <span class="comment">%Delete all matches that were not best matches</span>
0177         blastStructure(i).fromGenes(~keep)=[];
0178         blastStructure(i).toGenes(~keep)=[];
0179         blastStructure(i).evalue(~keep)=[];
0180         blastStructure(i).identity(~keep)=[];
0181         blastStructure(i).aligLen(~keep)=[];
0182         blastStructure(i).bitscore(~keep)=[];
0183         blastStructure(i).ppos(~keep)=[];
0184     <span class="keyword">end</span>
0185 <span class="keyword">end</span>
0186 
0187 useOrder=[{getModelFor};useOrder];
0188 
0189 <span class="keyword">for</span> i=1:numel(blastStructure)
0190     [~, toIndex]=ismember(blastStructure(i).toId,useOrder);
0191     [~, fromIndex]=ismember(blastStructure(i).fromId,useOrder);
0192     
0193     <span class="comment">%Add all genes to the corresponding list in allGenes</span>
0194     allGenes{toIndex}=[allGenes{toIndex};blastStructure(i).toGenes];
0195     allGenes{fromIndex}=[allGenes{fromIndex};blastStructure(i).fromGenes];
0196 <span class="keyword">end</span>
0197 
0198 <span class="comment">%Keep only the unique gene names</span>
0199 maxOtherGeneNr=0; <span class="comment">%Determines the dimension of the connectivity matrixes</span>
0200 <span class="keyword">for</span> i=1:numel(allGenes)
0201     allGenes{i}=unique(allGenes{i});
0202     <span class="keyword">if</span> i&gt;1
0203         <span class="keyword">if</span> numel(allGenes{i})&gt;maxOtherGeneNr
0204             maxOtherGeneNr=numel(allGenes{i});
0205         <span class="keyword">end</span>
0206     <span class="keyword">end</span>
0207 <span class="keyword">end</span>
0208 
0209 <span class="comment">%Generate a cell array of matrixes that describes how the genes in the new</span>
0210 <span class="comment">%organism map to the models. Each cell matches to the corresponding model</span>
0211 <span class="comment">%in useOrder (starting at 2 of course). First dimension is gene in new</span>
0212 <span class="comment">%organism, second which gene it is in the other organism. The second matrix</span>
0213 <span class="comment">%describes how they map back.</span>
0214 
0215 <span class="comment">%As it is now, a significant match is indicated by a 1. This could be</span>
0216 <span class="comment">%expanded to contain some kind of significance level. The first dimension</span>
0217 <span class="comment">%is still the genes in the new model.</span>
0218 
0219 allTo=cell(numel(useOrder)-1,1);
0220 allFrom=cell(numel(useOrder)-1,1);
0221 
0222 <span class="keyword">for</span> i=1:numel(useOrder)-1
0223     allTo{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0224     allFrom{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0225 <span class="keyword">end</span>
0226 
0227 <span class="comment">%Fill the matches to other species</span>
0228 <span class="keyword">for</span> i=1:numel(blastStructure)
0229     <span class="keyword">if</span> strcmp(blastStructure(i).toId,getModelFor)
0230         <span class="comment">%This was to the new organism They should all match so no checks</span>
0231         <span class="comment">%are being made</span>
0232         [~, a]=ismember(blastStructure(i).toGenes,allGenes{1});
0233         [~, fromModel]=ismember(blastStructure(i).fromId,useOrder);
0234         [~, b]=ismember(blastStructure(i).fromGenes,allGenes{fromModel});
0235         idx = sub2ind(size(allTo{fromModel-1}), a, b);
0236         allTo{fromModel-1}(idx)=1;
0237     <span class="keyword">else</span>
0238         <span class="comment">%This was from the new organism</span>
0239         [~, a]=ismember(blastStructure(i).fromGenes,allGenes{1});
0240         [~, toModel]=ismember(blastStructure(i).toId,useOrder);
0241         [~, b]=ismember(blastStructure(i).toGenes,allGenes{toModel});
0242         idx = sub2ind(size(allFrom{toModel-1}), a, b);
0243         allFrom{toModel-1}(idx)=1;
0244     <span class="keyword">end</span>
0245 <span class="keyword">end</span>
0246 
0247 <span class="comment">%Now we have all the gene matches in a convenient way. For all the genes in</span>
0248 <span class="comment">%the new organism get the genes that should be included from other</span>
0249 <span class="comment">%organisms. If all genes should be included this simply means keep the</span>
0250 <span class="comment">%allFrom matrix as is is. If only 1-1 orthologs are to be included then</span>
0251 <span class="comment">%only those elements are kept.</span>
0252 
0253 finalMappings=cell(numel(useOrder)-1,1);
0254 <span class="keyword">if</span> strictness==1 || strictness==3
0255     <span class="keyword">for</span> j=1:numel(allFrom)
0256         finalMappings{j}=allTo{j}~=0 &amp; allFrom{j}~=0;
0257     <span class="keyword">end</span>
0258 <span class="keyword">else</span>
0259     <span class="keyword">if</span> mapNewGenesToOld==true
0260         finalMappings=allFrom;
0261     <span class="keyword">else</span>
0262         finalMappings=allTo;
0263     <span class="keyword">end</span>
0264 <span class="keyword">end</span>
0265 
0266 <span class="comment">%Remove all genes from the mapping that are not in the models. This doesn't</span>
0267 <span class="comment">%do much if only genes in the models were used for the original mapping.</span>
0268 <span class="comment">%Also simplify the finalMapping and allGenes structures so that they only</span>
0269 <span class="comment">%contain mappings that exist</span>
0270 usedNewGenes=false(numel(allGenes{1}),1);
0271 
0272 <span class="keyword">for</span> i=1:numel(allGenes)-1
0273     <span class="comment">%First remove mappings for those genes that are not in the model</span>
0274     <span class="keyword">if</span> onlyGenesInModels==false
0275         a=ismember(allGenes{i+1},models{useOrderIndexes(i)}.genes);
0276         finalMappings{i}(:,~a)=false;
0277     <span class="keyword">end</span>
0278     
0279     <span class="comment">%Then remove unused ones and simplify</span>
0280     [a, b]=find(finalMappings{i});
0281     usedGenes=false(numel(allGenes{i+1}),1);
0282     usedNewGenes(a)=true;
0283     usedGenes(b)=true;
0284     finalMappings{i}=finalMappings{i}(:,usedGenes);
0285     allGenes{i+1}=allGenes{i+1}(usedGenes);
0286 <span class="keyword">end</span>
0287 
0288 <span class="comment">%Remove all new genes that have not been mapped to anything</span>
0289 allGenes{1}=allGenes{1}(usedNewGenes);
0290 <span class="keyword">for</span> i=1:numel(finalMappings)
0291     finalMappings{i}=finalMappings{i}(usedNewGenes,:);
0292 <span class="keyword">end</span>
0293 
0294 <span class="comment">%Now is it time to choose which reactions should be included from which</span>
0295 <span class="comment">%models. If there is a preferred order specified then each gene can only</span>
0296 <span class="comment">%result in reactions from one model, otherwise they should all be included</span>
0297 
0298 <span class="comment">%Start by simplifying the models by removing genes/reactions that are not</span>
0299 <span class="comment">%used. This is where it gets weird with complexes, especially &quot;or&quot;</span>
0300 <span class="comment">%complexes. In this step only reactions which are encoded by one single</span>
0301 <span class="comment">%gene, or where all genes should be deleted, are deleted. The info on the</span>
0302 <span class="comment">%full complex is still present in the grRules</span>
0303 
0304 <span class="keyword">for</span> i=1:numel(models)
0305     a=ismember(models{useOrderIndexes(i)}.genes,allGenes{i+1});
0306     
0307     <span class="comment">%Don't remove reactions with complexes if not all genes in the complex</span>
0308     <span class="comment">%should be deleted. NOTE: This means that not all the genes in 'a' are</span>
0309     <span class="comment">%guaranteed to be deleted. This approach works fine for 'and'</span>
0310     <span class="comment">%complexes, but there should be a check that it doesn't keep 'or' genes</span>
0311     <span class="comment">%if it doesn't have to!</span>
0312     models{useOrderIndexes(i)}=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},~a,true,true,false);
0313 <span class="keyword">end</span>
0314 
0315 <span class="comment">%Since I want to use mergeModels in the end, I simplify the models further</span>
0316 <span class="comment">%by deleting genes/reactions in the order specified by preferredOrder. This</span>
0317 <span class="comment">%means that the last model will only contain reactions for genes that</span>
0318 <span class="comment">%mapped only to that model</span>
0319 
0320 allUsedGenes=false(numel(allGenes{1}),1);
0321 
0322 <span class="keyword">if</span> ~isempty(preferredOrder) &amp;&amp; numel(models)&gt;1
0323     [usedGenes, ~]=find(finalMappings{1}); <span class="comment">%All that are used in the first model in preferredOrder</span>
0324     allUsedGenes(usedGenes)=true;
0325     <span class="keyword">for</span> i=2:numel(finalMappings)
0326         [usedGenes, ~]=find(finalMappings{i});
0327         usedGenes=unique(usedGenes);
0328         a=ismember(usedGenes,find(allUsedGenes));
0329         
0330         [~, genesToDelete]=find(finalMappings{i}(usedGenes(a),:)); <span class="comment">%IMPORTANT! IS it really correct to remove all genes that map?</span>
0331         genesToDelete=unique(genesToDelete); <span class="comment">%Maybe not needed, but for clarity if nothing else</span>
0332         
0333         <span class="comment">%Remove all the genes that were already found and add the other</span>
0334         <span class="comment">%ones to allUsedGenes</span>
0335         [models{useOrderIndexes(i)}, notDeleted]=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},allGenes{i+1}(genesToDelete),true,false);
0336         allUsedGenes(usedGenes)=true;
0337         
0338         <span class="comment">%Remove the deleted genes from finalMappings and allGenes Don't</span>
0339         <span class="comment">%remove the genes in notDeleted, they are part of complexes with</span>
0340         <span class="comment">%some non-mapped genes</span>
0341         deletedIndexes=~ismember(allGenes{i+1}(genesToDelete),notDeleted);
0342         finalMappings{i}(:,genesToDelete(deletedIndexes))=[];
0343         allGenes{i+1}(genesToDelete(deletedIndexes))=[];
0344     <span class="keyword">end</span>
0345 <span class="keyword">end</span>
0346 
0347 <span class="comment">%Now loop through the models and update the gene associations. Genes not</span>
0348 <span class="comment">%belonging to the new organism will be renamed as 'OLD_MODELID_gene'</span>
0349 <span class="keyword">for</span> i=1:numel(models)
0350     <span class="comment">%Find all the new genes that should be used for this model</span>
0351     [newGenes, oldGenes]=find(finalMappings{i});
0352     
0353     <span class="comment">%Create a new gene list with the genes from the new organism and those</span>
0354     <span class="comment">%genes that could not be removed</span>
0355     replaceableGenes=allGenes{i+1}(unique(oldGenes));
0356     nonReplaceableGenes=setdiff(models{useOrderIndexes(i)}.genes,replaceableGenes);
0357     fullGeneList=[allGenes{1}(unique(newGenes));nonReplaceableGenes];
0358     
0359     <span class="comment">%Just to save some indexing later. This is the LAST index of</span>
0360     <span class="comment">%replaceable ones</span>
0361     nonRepStartIndex=numel(unique(newGenes));
0362     
0363     <span class="comment">%Construct a new rxnGeneMat</span>
0364     newRxnGeneMat=sparse(numel(models{useOrderIndexes(i)}.rxns),numel(fullGeneList));
0365     
0366     <span class="comment">%Now update the rxnGeneMat matrix. This is a little tricky and could</span>
0367     <span class="comment">%probably be done in a more efficient way, but I just loop through the</span>
0368     <span class="comment">%reactions and add them one by one</span>
0369     <span class="keyword">for</span> j=1:numel(models{useOrderIndexes(i)}.rxns)
0370         <span class="comment">%Get the old genes encoding for this reaction</span>
0371         [~, oldGeneIds]=find(models{useOrderIndexes(i)}.rxnGeneMat(j,:));
0372         
0373         <span class="comment">%Update the matrix for each gene. This includes replacing one gene</span>
0374         <span class="comment">%with several new ones if there were several matches</span>
0375         <span class="keyword">for</span> k=1:numel(oldGeneIds)
0376             <span class="comment">%Match the gene to one in the gene list. This is done as a text</span>
0377             <span class="comment">%match. Could probably be done better, but I'm a little lost in</span>
0378             <span class="comment">%the indexing</span>
0379             
0380             geneName=models{useOrderIndexes(i)}.genes(oldGeneIds(k));
0381             
0382             <span class="comment">%First search in the mappable genes</span>
0383             mapIndex=find(ismember(allGenes{i+1},geneName));
0384             
0385             <span class="keyword">if</span> ~isempty(mapIndex)
0386                 <span class="comment">%Get the new genes for that gene</span>
0387                 a=find(finalMappings{i}(:,mapIndex));
0388                 
0389                 <span class="comment">%Find the positions of these genes in the final gene list</span>
0390                 [~, b]=ismember(allGenes{1}(a),fullGeneList);
0391                 
0392                 <span class="comment">%Update the matrix</span>
0393                 newRxnGeneMat(j,b)=1;
0394                 
0395                 <span class="comment">%Update the grRules string. This is tricky, but I hope that</span>
0396                 <span class="comment">%it's ok to replace the old gene name with the new one and</span>
0397                 <span class="comment">%add ') or (' if there were several matches. Be sure of</span>
0398                 <span class="comment">%this!</span>
0399                 repString=fullGeneList{b(1)};
0400                 <span class="keyword">for</span> l=2:numel(b)
0401                     repString=[repString <span class="string">') or ('</span> fullGeneList{b(l)}];
0402                 <span class="keyword">end</span>
0403                 models{useOrderIndexes(i)}.grRules{j}=strrep(models{useOrderIndexes(i)}.grRules{j},geneName{1},repString);
0404             <span class="keyword">else</span>
0405                 <span class="comment">%Then search in the non-replaceable genes. There could only</span>
0406                 <span class="comment">%be one match here</span>
0407                 index=find(ismember(nonReplaceableGenes,geneName));
0408                 
0409                 <span class="comment">%Update the matrix</span>
0410                 newRxnGeneMat(j,nonRepStartIndex+index)=1;
0411                 
0412                 models{useOrderIndexes(i)}.grRules{j}=strrep(models{useOrderIndexes(i)}.grRules{j},geneName{1},strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,geneName{1}));
0413             <span class="keyword">end</span>
0414         <span class="keyword">end</span>
0415     <span class="keyword">end</span>
0416     
0417     <span class="comment">%Add the new list of genes</span>
0418     models{useOrderIndexes(i)}.rxnGeneMat=newRxnGeneMat;
0419     <span class="keyword">if</span> ~isempty(nonReplaceableGenes)
0420         models{useOrderIndexes(i)}.genes=[allGenes{1}(unique(newGenes));strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,nonReplaceableGenes)];
0421     <span class="keyword">else</span>
0422         models{useOrderIndexes(i)}.genes=allGenes{1}(unique(newGenes));
0423     <span class="keyword">end</span>
0424     <span class="keyword">if</span> isfield(models{useOrderIndexes(i)},<span class="string">'geneComps'</span>)
0425         geneComps=models{useOrderIndexes(i)}.geneComps(1);
0426         models{useOrderIndexes(i)}.geneComps=zeros(numel(models{useOrderIndexes(i)}.genes),1);
0427         <span class="comment">%Assume that all genes are in the same compartment, and this</span>
0428         <span class="comment">%compartment is specified for the first gene</span>
0429         models{useOrderIndexes(i)}.geneComps(:)=geneComps;
0430     <span class="keyword">end</span>
0431 <span class="keyword">end</span>
0432 
0433 <span class="comment">%Now merge the models. All information should be correct except for 'or'</span>
0434 <span class="comment">%complexes</span>
0435 draftModel=<a href="mergeModels.html" class="code" title="function model=mergeModels(models,supressWarnings)">mergeModels</a>(models);
0436 
0437 <span class="comment">%Change description of the resulting model</span>
0438 draftModel.id=getModelFor;
0439 description=<span class="string">'Generated by getModelFromHomology using '</span>;
0440 <span class="keyword">for</span> i=1:numel(models)
0441     <span class="keyword">if</span> i&lt;numel(models)
0442         description=[description models{i}.id <span class="string">', '</span>];
0443     <span class="keyword">else</span>
0444         description=[description models{i}.id];
0445     <span class="keyword">end</span>
0446 <span class="keyword">end</span>
0447 draftModel.description=description;
0448 draftModel.rxnNotes=cell(length(draftModel.rxns),1);
0449 draftModel.rxnNotes(:)={<span class="string">'Reaction included by getModelFromHomology'</span>};
0450 draftModel.rxnConfidenceScores=NaN(length(draftModel.rxns),1);
0451 draftModel.rxnConfidenceScores(:)=2;
0452 <span class="comment">%Gene short names are often different between species, safer not to</span>
0453 <span class="comment">%include them.</span>
0454 <span class="keyword">if</span> isfield(draftModel,<span class="string">'geneShortNames'</span>);
0455     draftModel = rmfield(draftModel,<span class="string">'geneShortNames'</span>);
0456 <span class="keyword">end</span>
0457 <span class="comment">%Standardize grRules and notify if problematic grRules are found</span>
0458 [draftModel.grRules,draftModel.rxnGeneMat] = <a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(draftModel,false);
0459 <span class="keyword">end</span></pre></div>
<hr><address>Generated on Mon 07-May-2018 14:25:05 by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>