<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
                "http://www.w3.org/TR/REC-html40/loose.dtd">
<html>
<head>
  <title>Description of getModelFromHomology</title>
  <meta name="keywords" content="getModelFromHomology">
  <meta name="description" content="getModelFromHomology">
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <meta name="generator" content="m2html v1.5 &copy; 2003-2005 Guillaume Flandin">
  <meta name="robots" content="index, follow">
  <link type="text/css" rel="stylesheet" href="../m2html.css">
</head>
<body>
<a name="_top"></a>
<div><a href="../index.html">Home</a> &gt;  <a href="index.html">core</a> &gt; getModelFromHomology.m</div>

<!--<table width="100%"><tr><td align="left"><a href="../index.html"><img alt="<" border="0" src="../left.png">&nbsp;Master index</a></td>
<td align="right"><a href="index.html">Index for core&nbsp;<img alt=">" border="0" src="../right.png"></a></td></tr></table>-->

<h1>getModelFromHomology
</h1>

<h2><a name="_name"></a>PURPOSE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>getModelFromHomology</strong></div>

<h2><a name="_synopsis"></a>SYNOPSIS <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="box"><strong>function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,minLen,minIde,mapNewGenesToOld) </strong></div>

<h2><a name="_description"></a>DESCRIPTION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre class="comment"> getModelFromHomology
   Constructs a new model from a set of existing models and gene homology
   information.

   models            a cell array of model structures to build the model
                     from. These models must be sorted by importance in
                     decreasing order
   blastStructure    a blastStructure as produced by getBlast or
                     getBlastFromExcel
   getModelFor       a three-four letter abbreviation of the organism to
                     build a model for. Must have BLASTP hits in both
                     directions to the organisms in 'models'
   preferredOrder    the order in which reactions should be added from the
                     models. If not supplied, reactions will be included
                     from all models, otherwise one gene will only result
                     in reactions from one model (opt, default {})
   strictness        integer that specifies which reactions should be
                     included:
                     1: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in both directions
                     2: Map new genes to old for all pairs, which have
                     acceptable BLASTP results in correspondent direction
                     (mapping can be done in the opposite direction, see
                     mapNewGenesToOld below)
                     3: Check all BLASTP results and retain only the best
                     results by E-value for all gene pairs in each
                     direction separately. Then map new genes to old for
                     all pairs, which have acceptable BLASTP results in
                     both directions (opt, default 1).
   onlyGenesInModels consider BLASTP results only for genes that exist in
                     the models. This tends to import a larger fraction
                     from the existing models but may give less reliable
                     results. Has effect only if strictness=3 (opt,
                     default false)
   maxE              only look at genes with E-values &lt;= this value (opt,
                     default 10^-30)
   minLen            only look at genes with alignment length &gt;= this
                     value (opt, default 200)
   minIde            only look at genes with identity &gt;= this value
                     (opt, default 40 (%))
   mapNewGenesToOld  determines how to match genes if not looking at only
                     1-1 orthologs. Either map the new genes to the old or
                     old genes to new. The default is to map the new genes
                     (opt, default true)

   draftModel        a model structure for the new organism
   hitGenes          collect the old and new genes

   The models in the 'models' structure should have named the metabolites
   in the same manner, have their reversible reactions in the same
   direction (run sortModel), and use the same compartment names. To avoid
   keeping unneccesary old genes, the models should not have
   'or'-relations in their grRules (use expandModel).

   The resulting draft model contains only reactions associated with
   orthologous genes. The old (original) genes involved in 'and'
   relations in grRules without any orthologs are still included in
   the draft model as OLD_MODELID_geneName.

   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism

   Usage: draftModel=getModelFromHomology(models,blastStructure,...
    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...
    minLen,minIde,mapNewGenesToOld)</pre></div>

<!-- crossreference -->
<h2><a name="_cross"></a>CROSS-REFERENCE INFORMATION <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
This function calls:
<ul style="list-style-image:url(../matlabicon.gif)">
<li><a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>	dispEM</li><li><a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>	mergeModels</li><li><a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>	removeGenes</li><li><a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>	removeReactions</li><li><a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>	standardizeGrRules</li></ul>
This function is called by:
<ul style="list-style-image:url(../matlabicon.gif)">
</ul>
<!-- crossreference -->



<h2><a name="_source"></a>SOURCE CODE <a href="#_top"><img alt="^" border="0" src="../up.png"></a></h2>
<div class="fragment"><pre>0001 <a name="_sub0" href="#_subfunctions" class="code">function [draftModel, hitGenes]=getModelFromHomology(models,blastStructure,</a><span class="keyword">...</span>
0002     getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,<span class="keyword">...</span>
0003     minLen,minIde,mapNewGenesToOld)
0004 <span class="comment">% getModelFromHomology</span>
0005 <span class="comment">%   Constructs a new model from a set of existing models and gene homology</span>
0006 <span class="comment">%   information.</span>
0007 <span class="comment">%</span>
0008 <span class="comment">%   models            a cell array of model structures to build the model</span>
0009 <span class="comment">%                     from. These models must be sorted by importance in</span>
0010 <span class="comment">%                     decreasing order</span>
0011 <span class="comment">%   blastStructure    a blastStructure as produced by getBlast or</span>
0012 <span class="comment">%                     getBlastFromExcel</span>
0013 <span class="comment">%   getModelFor       a three-four letter abbreviation of the organism to</span>
0014 <span class="comment">%                     build a model for. Must have BLASTP hits in both</span>
0015 <span class="comment">%                     directions to the organisms in 'models'</span>
0016 <span class="comment">%   preferredOrder    the order in which reactions should be added from the</span>
0017 <span class="comment">%                     models. If not supplied, reactions will be included</span>
0018 <span class="comment">%                     from all models, otherwise one gene will only result</span>
0019 <span class="comment">%                     in reactions from one model (opt, default {})</span>
0020 <span class="comment">%   strictness        integer that specifies which reactions should be</span>
0021 <span class="comment">%                     included:</span>
0022 <span class="comment">%                     1: Map new genes to old for all pairs, which have</span>
0023 <span class="comment">%                     acceptable BLASTP results in both directions</span>
0024 <span class="comment">%                     2: Map new genes to old for all pairs, which have</span>
0025 <span class="comment">%                     acceptable BLASTP results in correspondent direction</span>
0026 <span class="comment">%                     (mapping can be done in the opposite direction, see</span>
0027 <span class="comment">%                     mapNewGenesToOld below)</span>
0028 <span class="comment">%                     3: Check all BLASTP results and retain only the best</span>
0029 <span class="comment">%                     results by E-value for all gene pairs in each</span>
0030 <span class="comment">%                     direction separately. Then map new genes to old for</span>
0031 <span class="comment">%                     all pairs, which have acceptable BLASTP results in</span>
0032 <span class="comment">%                     both directions (opt, default 1).</span>
0033 <span class="comment">%   onlyGenesInModels consider BLASTP results only for genes that exist in</span>
0034 <span class="comment">%                     the models. This tends to import a larger fraction</span>
0035 <span class="comment">%                     from the existing models but may give less reliable</span>
0036 <span class="comment">%                     results. Has effect only if strictness=3 (opt,</span>
0037 <span class="comment">%                     default false)</span>
0038 <span class="comment">%   maxE              only look at genes with E-values &lt;= this value (opt,</span>
0039 <span class="comment">%                     default 10^-30)</span>
0040 <span class="comment">%   minLen            only look at genes with alignment length &gt;= this</span>
0041 <span class="comment">%                     value (opt, default 200)</span>
0042 <span class="comment">%   minIde            only look at genes with identity &gt;= this value</span>
0043 <span class="comment">%                     (opt, default 40 (%))</span>
0044 <span class="comment">%   mapNewGenesToOld  determines how to match genes if not looking at only</span>
0045 <span class="comment">%                     1-1 orthologs. Either map the new genes to the old or</span>
0046 <span class="comment">%                     old genes to new. The default is to map the new genes</span>
0047 <span class="comment">%                     (opt, default true)</span>
0048 <span class="comment">%</span>
0049 <span class="comment">%   draftModel        a model structure for the new organism</span>
0050 <span class="comment">%   hitGenes          collect the old and new genes</span>
0051 <span class="comment">%</span>
0052 <span class="comment">%   The models in the 'models' structure should have named the metabolites</span>
0053 <span class="comment">%   in the same manner, have their reversible reactions in the same</span>
0054 <span class="comment">%   direction (run sortModel), and use the same compartment names. To avoid</span>
0055 <span class="comment">%   keeping unneccesary old genes, the models should not have</span>
0056 <span class="comment">%   'or'-relations in their grRules (use expandModel).</span>
0057 <span class="comment">%</span>
0058 <span class="comment">%   The resulting draft model contains only reactions associated with</span>
0059 <span class="comment">%   orthologous genes. The old (original) genes involved in 'and'</span>
0060 <span class="comment">%   relations in grRules without any orthologs are still included in</span>
0061 <span class="comment">%   the draft model as OLD_MODELID_geneName.</span>
0062 <span class="comment">%</span>
0063 <span class="comment">%   NOTE: &quot;to&quot; and &quot;from&quot; means relative to the new organism</span>
0064 <span class="comment">%</span>
0065 <span class="comment">%   Usage: draftModel=getModelFromHomology(models,blastStructure,...</span>
0066 <span class="comment">%    getModelFor,preferredOrder,strictness,onlyGenesInModels,maxE,...</span>
0067 <span class="comment">%    minLen,minIde,mapNewGenesToOld)</span>
0068 
0069 hitGenes.oldGenes = [];  <span class="comment">% collect the old genes from the template model (organism)</span>
0070 hitGenes.newGenes = [];  <span class="comment">% collect the new genes of the draft model (target organism)</span>
0071 
0072 <span class="keyword">if</span> nargin&lt;4
0073     preferredOrder=[];
0074 <span class="keyword">end</span>
0075 <span class="keyword">if</span> nargin&lt;5
0076     strictness=1;
0077 <span class="keyword">end</span>
0078 <span class="keyword">if</span> nargin&lt;6
0079     onlyGenesInModels=false;
0080 <span class="keyword">end</span>
0081 <span class="keyword">if</span> nargin&lt;7
0082     maxE=10^-30;
0083 <span class="keyword">end</span>
0084 <span class="keyword">if</span> nargin&lt;8
0085     minLen=200;
0086 <span class="keyword">end</span>
0087 <span class="keyword">if</span> nargin&lt;9
0088     minIde=40;
0089 <span class="keyword">end</span>
0090 <span class="keyword">if</span> nargin&lt;10
0091     mapNewGenesToOld=true;
0092 <span class="keyword">end</span>
0093 
0094 preferredOrder=preferredOrder(:);
0095 
0096 <span class="keyword">if</span> isfield(models,<span class="string">'S'</span>)
0097     models={models};
0098 <span class="keyword">end</span>
0099     
0100 <span class="comment">%Check that all the information is in the blast structure</span>
0101 modelNames=cell(numel(models),1);
0102 <span class="keyword">for</span> i=1:numel(models)
0103     modelNames{i}=models{i}.id;
0104     <span class="comment">%Gene short names and geneMiriams are often different between species,</span>
0105     <span class="comment">%safer not to include them</span>
0106     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneShortNames'</span>)
0107         models{i}=rmfield(models{i},<span class="string">'geneShortNames'</span>);
0108     <span class="keyword">end</span>
0109     <span class="keyword">if</span> isfield(models{i},<span class="string">'geneMiriams'</span>)
0110         models{i}=rmfield(models{i},<span class="string">'geneMiriams'</span>);
0111     <span class="keyword">end</span>
0112 <span class="keyword">end</span>
0113 
0114 <span class="comment">%Assume for now that all information is there and that it's correct. This</span>
0115 <span class="comment">%is important to fix since no further checks are being made!</span>
0116 
0117 <span class="comment">%Check whether provided fasta files use the same gene identifiers as</span>
0118 <span class="comment">%provided template models</span>
0119 <span class="keyword">for</span> i=1:numel(blastStructure)
0120     <span class="keyword">if</span> ~strcmp(blastStructure(i).fromId,getModelFor)
0121         j=strcmpi(blastStructure(i).fromId,modelNames);
0122         <span class="keyword">if</span> j==0
0123             error([<span class="string">'While the blastStructure contains sequences from '</span><span class="keyword">...</span>
0124                 <span class="string">'organismID &quot;%s&quot; (as\nprovided in getBlast), none of '</span><span class="keyword">...</span>
0125                 <span class="string">'template models have this id (as model.id)'</span>],<span class="keyword">...</span>
0126                 string(blastStructure(i).fromId));
0127         <span class="keyword">end</span>
0128         k=sum(ismember(blastStructure(i).fromGenes,models{j}.genes));
0129         <span class="keyword">if</span> k&lt;(numel(models{j}.genes)*0.05)
0130             error([<span class="string">'Less than 5%% of the genes in the template model '</span><span class="keyword">...</span>
0131                 <span class="string">'with model.id &quot;%s&quot;\ncan be found in the blastStructure. '</span><span class="keyword">...</span>
0132                 <span class="string">'Ensure that the protein FASTA\nused in getBlast and '</span><span class="keyword">...</span>
0133                 <span class="string">'the template model used in getModelFromHomology\nuse '</span><span class="keyword">...</span>
0134                 <span class="string">'the same style of gene identifiers'</span>],models{j}.id)
0135         <span class="keyword">end</span>
0136     <span class="keyword">end</span>
0137 <span class="keyword">end</span>
0138 
0139 <span class="comment">%Standardize grRules of template models</span>
0140 <span class="keyword">for</span> i=1:length(models)
0141     fprintf(<span class="string">'\nStandardizing grRules of template model with ID &quot;%s&quot; ...'</span>,models{i}.id);
0142     [models{i}.grRules,models{i}.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(models{i},false);
0143 <span class="keyword">end</span>
0144 fprintf(<span class="string">' done\n'</span>);
0145 
0146 <span class="comment">%Remove all gene matches that are below the cutoffs</span>
0147 <span class="keyword">for</span> i=1:numel(blastStructure)
0148     indexes=blastStructure(i).evalue&lt;maxE &amp; blastStructure(i).aligLen&gt;=minLen &amp; blastStructure(i).identity&gt;=minIde; <span class="comment">%Do it in this direction to lose NaNs</span>
0149     blastStructure(i).fromGenes(~indexes)=[];
0150     blastStructure(i).toGenes(~indexes)=[];
0151     blastStructure(i).evalue(~indexes)=[];
0152     blastStructure(i).identity(~indexes)=[];
0153     blastStructure(i).aligLen(~indexes)=[];
0154     blastStructure(i).bitscore(~indexes)=[];
0155     blastStructure(i).ppos(~indexes)=[];
0156 <span class="keyword">end</span>
0157 
0158 <span class="comment">%Remove all reactions from the models that have no genes encoding for them.</span>
0159 <span class="comment">%Also remove all genes that encode for no reactions. There shouldn't be any</span>
0160 <span class="comment">%but there might be mistakes</span>
0161 <span class="keyword">for</span> i=1:numel(models)
0162     [hasGenes, ~]=find(models{i}.rxnGeneMat);
0163     hasNoGenes=1:numel(models{i}.rxns);
0164     hasNoGenes(hasGenes)=[];
0165     models{i}=<a href="removeReactions.html" class="code" title="function reducedModel=removeReactions(model,rxnsToRemove,removeUnusedMets,removeUnusedGenes,removeUnusedComps)">removeReactions</a>(models{i},hasNoGenes,true,true);
0166 <span class="keyword">end</span>
0167 
0168 <span class="comment">%Create a structure that contains all genes used in the blasts in any</span>
0169 <span class="comment">%direction for each of the models in 'models' and for the new organism. The</span>
0170 <span class="comment">%first cell is for the new organism and then according to the preferred</span>
0171 <span class="comment">%order. If no such order is supplied, then according to the order in</span>
0172 <span class="comment">%'models'</span>
0173 allGenes=cell(numel(models)+1,1);
0174 <span class="keyword">if</span> isempty(preferredOrder)
0175     useOrder=modelNames;
0176 <span class="keyword">else</span>
0177     useOrder=preferredOrder;
0178 <span class="keyword">end</span>
0179 
0180 <span class="comment">%Get the corresponding indexes for those models in the 'models' structure</span>
0181 useOrderIndexes=zeros(numel(models),1);
0182 <span class="keyword">for</span> i=1:numel(models)
0183     [~, index]=ismember(models{i}.id,useOrder);
0184     useOrderIndexes(index)=i;
0185 <span class="keyword">end</span>
0186 
0187 <span class="comment">%Remove all genes from the blast structure that have no genes in the models</span>
0188 <span class="keyword">if</span> onlyGenesInModels==true
0189     modelGenes={};
0190     <span class="keyword">for</span> i=1:numel(models)
0191         modelGenes=[modelGenes;models{i}.genes];
0192     <span class="keyword">end</span>
0193     <span class="keyword">for</span> i=1:numel(blastStructure)
0194         <span class="comment">%Check to see if it should match the toId or fromId</span>
0195         <span class="keyword">if</span> strcmpi(blastStructure(i).fromId,getModelFor)
0196             I=ismember(blastStructure(i).toGenes,modelGenes);
0197         <span class="keyword">else</span>
0198             I=ismember(blastStructure(i).fromGenes,modelGenes);
0199         <span class="keyword">end</span>
0200         blastStructure(i).fromGenes(~I)=[];
0201         blastStructure(i).toGenes(~I)=[];
0202         blastStructure(i).evalue(~I)=[];
0203         blastStructure(i).identity(~I)=[];
0204         blastStructure(i).aligLen(~I)=[];
0205         blastStructure(i).bitscore(~I)=[];
0206         blastStructure(i).ppos(~I)=[];
0207         
0208         <span class="comment">%Check that no matching in blastStructure is empty. This happens if</span>
0209         <span class="comment">%no genes in the models are present in the corresponding sheet</span>
0210         <span class="keyword">if</span> isempty(blastStructure(i).fromGenes)
0211             EM=[<span class="string">'No genes in matching from '</span> blastStructure(i).fromId <span class="string">' to '</span> blastStructure(i).toId <span class="string">' are present in the corresponding model'</span>];
0212             <a href="dispEM.html" class="code" title="function dispEM(string,throwErrors,toList,trimWarnings)">dispEM</a>(EM);
0213         <span class="keyword">end</span>
0214     <span class="keyword">end</span>
0215 <span class="keyword">end</span>
0216 
0217 <span class="comment">%If only best orthologs are to be used then all other measurements are</span>
0218 <span class="comment">%deleted from the blastStructure. All code after this stays the same. This</span>
0219 <span class="comment">%means that preferred order can still matter. The best ortholog scoring is</span>
0220 <span class="comment">%based only on the E-value</span>
0221 <span class="keyword">if</span> strictness==3
0222     <span class="keyword">for</span> i=1:numel(blastStructure)
0223         keep=false(numel(blastStructure(i).toGenes),1);
0224         [allFromGenes, ~, I]=unique(blastStructure(i).fromGenes);
0225         
0226         <span class="comment">%It would be nice to get rid of this loop</span>
0227         <span class="keyword">for</span> j=1:numel(allFromGenes)
0228             allMatches=find(I==j);
0229             bestMatches=allMatches(blastStructure(i).evalue(allMatches)==min(blastStructure(i).evalue(allMatches)));
0230             
0231             <span class="comment">%Keep the best matches</span>
0232             keep(bestMatches)=true;
0233         <span class="keyword">end</span>
0234         
0235         <span class="comment">%Delete all matches that were not best matches</span>
0236         blastStructure(i).fromGenes(~keep)=[];
0237         blastStructure(i).toGenes(~keep)=[];
0238         blastStructure(i).evalue(~keep)=[];
0239         blastStructure(i).identity(~keep)=[];
0240         blastStructure(i).aligLen(~keep)=[];
0241         blastStructure(i).bitscore(~keep)=[];
0242         blastStructure(i).ppos(~keep)=[];
0243     <span class="keyword">end</span>
0244 <span class="keyword">end</span>
0245 
0246 useOrder=[{getModelFor};useOrder];
0247 
0248 <span class="keyword">for</span> i=1:numel(blastStructure)
0249     [~, toIndex]=ismember(blastStructure(i).toId,useOrder);
0250     [~, fromIndex]=ismember(blastStructure(i).fromId,useOrder);
0251     
0252     <span class="comment">%Add all genes to the corresponding list in allGenes</span>
0253     allGenes{toIndex}=[allGenes{toIndex};blastStructure(i).toGenes];
0254     allGenes{fromIndex}=[allGenes{fromIndex};blastStructure(i).fromGenes];
0255 <span class="keyword">end</span>
0256 
0257 <span class="comment">%Keep only the unique gene names</span>
0258 maxOtherGeneNr=0; <span class="comment">%Determines the dimension of the connectivity matrixes</span>
0259 <span class="keyword">for</span> i=1:numel(allGenes)
0260     allGenes{i}=unique(allGenes{i});
0261     <span class="keyword">if</span> i&gt;1
0262         <span class="keyword">if</span> numel(allGenes{i})&gt;maxOtherGeneNr
0263             maxOtherGeneNr=numel(allGenes{i});
0264         <span class="keyword">end</span>
0265     <span class="keyword">end</span>
0266 <span class="keyword">end</span>
0267 
0268 <span class="comment">%Generate a cell array of matrixes that describes how the genes in the new</span>
0269 <span class="comment">%organism map to the models. Each cell matches to the corresponding model</span>
0270 <span class="comment">%in useOrder (starting at 2 of course). First dimension is gene in new</span>
0271 <span class="comment">%organism, second which gene it is in the other organism. The second matrix</span>
0272 <span class="comment">%describes how they map back.</span>
0273 
0274 <span class="comment">%As it is now, a significant match is indicated by a 1. This could be</span>
0275 <span class="comment">%expanded to contain some kind of significance level. The first dimension</span>
0276 <span class="comment">%is still the genes in the new model.</span>
0277 
0278 allTo=cell(numel(useOrder)-1,1);
0279 allFrom=cell(numel(useOrder)-1,1);
0280 
0281 <span class="keyword">for</span> i=1:numel(useOrder)-1
0282     allTo{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0283     allFrom{i}=sparse(numel(allGenes{1}),numel(allGenes{i+1}));
0284 <span class="keyword">end</span>
0285 
0286 <span class="comment">%Fill the matches to other species</span>
0287 <span class="keyword">for</span> i=1:numel(blastStructure)
0288     <span class="keyword">if</span> strcmp(blastStructure(i).toId,getModelFor)
0289         <span class="comment">%This was 'to' the new organism. They should all match so no checks</span>
0290         <span class="comment">%are being made</span>
0291         [~, a]=ismember(blastStructure(i).toGenes,allGenes{1});
0292         [~, fromModel]=ismember(blastStructure(i).fromId,useOrder);
0293         [~, b]=ismember(blastStructure(i).fromGenes,allGenes{fromModel});
0294         idx = sub2ind(size(allTo{fromModel-1}), a, b);
0295         allTo{fromModel-1}(idx)=1;
0296     <span class="keyword">else</span>
0297         <span class="comment">%This was 'from' the new organism</span>
0298         [~, a]=ismember(blastStructure(i).fromGenes,allGenes{1});
0299         [~, toModel]=ismember(blastStructure(i).toId,useOrder);
0300         [~, b]=ismember(blastStructure(i).toGenes,allGenes{toModel});
0301         idx = sub2ind(size(allFrom{toModel-1}), a, b);
0302         allFrom{toModel-1}(idx)=1;
0303     <span class="keyword">end</span>
0304 <span class="keyword">end</span>
0305 
0306 <span class="comment">%Now we have all the gene matches in a convenient way. For all the genes in</span>
0307 <span class="comment">%the new organism get the genes that should be included from other</span>
0308 <span class="comment">%organisms. If all genes should be included this simply means keep the</span>
0309 <span class="comment">%allFrom matrix as it is. If only orthologs which could be mapped in both</span>
0310 <span class="comment">%BLAST directions are to be included then only those elements are kept.</span>
0311 
0312 finalMappings=cell(numel(useOrder)-1,1);
0313 <span class="keyword">if</span> strictness==1 || strictness==3
0314     <span class="keyword">for</span> j=1:numel(allFrom)
0315         finalMappings{j}=allTo{j}~=0 &amp; allFrom{j}~=0;
0316     <span class="keyword">end</span>
0317 <span class="keyword">else</span>
0318     <span class="keyword">if</span> mapNewGenesToOld==true
0319         finalMappings=allFrom;
0320     <span class="keyword">else</span>
0321         finalMappings=allTo;
0322     <span class="keyword">end</span>
0323 <span class="keyword">end</span>
0324 
0325 <span class="comment">%Remove all genes from the mapping that are not in the models. This doesn't</span>
0326 <span class="comment">%do much if only genes in the models were used for the original mapping.</span>
0327 <span class="comment">%Also simplify the finalMapping and allGenes structures so that they only</span>
0328 <span class="comment">%contain mappings that exist</span>
0329 usedNewGenes=false(numel(allGenes{1}),1);
0330 
0331 <span class="keyword">for</span> i=1:numel(allGenes)-1
0332     <span class="comment">%First remove mappings for those genes that are not in the model</span>
0333     <span class="keyword">if</span> onlyGenesInModels==false
0334         a=ismember(allGenes{i+1},models{useOrderIndexes(i)}.genes);
0335         finalMappings{i}(:,~a)=false;
0336     <span class="keyword">end</span>
0337     
0338     <span class="comment">%Then remove unused ones and simplify</span>
0339     [a, b]=find(finalMappings{i});
0340     usedGenes=false(numel(allGenes{i+1}),1);
0341     usedNewGenes(a)=true;
0342     usedGenes(b)=true;
0343     finalMappings{i}=finalMappings{i}(:,usedGenes);
0344     allGenes{i+1}=allGenes{i+1}(usedGenes);
0345 <span class="keyword">end</span>
0346 
0347 <span class="comment">%Remove all new genes that have not been mapped to anything</span>
0348 allGenes{1}=allGenes{1}(usedNewGenes);
0349 <span class="keyword">for</span> i=1:numel(finalMappings)
0350     finalMappings{i}=finalMappings{i}(usedNewGenes,:);
0351 <span class="keyword">end</span>
0352 
0353 <span class="comment">%Now is it time to choose which reactions should be included from which</span>
0354 <span class="comment">%models. If there is a preferred order specified then each gene can only</span>
0355 <span class="comment">%result in reactions from one model, otherwise they should all be included</span>
0356 
0357 <span class="comment">%Start by simplifying the models by removing genes/reactions that are not</span>
0358 <span class="comment">%used. This is where it gets weird with complexes, especially &quot;or&quot;</span>
0359 <span class="comment">%complexes. In this step only reactions which are encoded by one single</span>
0360 <span class="comment">%gene, or where all genes should be deleted, are deleted. The info on the</span>
0361 <span class="comment">%full complex is still present in the grRules</span>
0362 
0363 <span class="keyword">for</span> i=1:numel(models)
0364     a=ismember(models{useOrderIndexes(i)}.genes,allGenes{i+1});
0365     
0366     <span class="comment">%Don't remove reactions with complexes if not all genes in the complex</span>
0367     <span class="comment">%should be deleted. NOTE: This means that not all the genes in 'a' are</span>
0368     <span class="comment">%guaranteed to be deleted. This approach works fine for 'and'</span>
0369     <span class="comment">%complexes, but there should be a check that it doesn't keep 'or' genes</span>
0370     <span class="comment">%if it doesn't have to!</span>
0371     models{useOrderIndexes(i)}=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},~a,true,true,false);
0372 <span class="keyword">end</span>
0373 
0374 <span class="comment">%Since mergeModels function will be used in the end, the models are</span>
0375 <span class="comment">%simplified further by deleting genes/reactions in the order specified by</span>
0376 <span class="comment">%preferredOrder. This means that the last model will only contain reactions</span>
0377 <span class="comment">%for genes that mapped only to that model</span>
0378 
0379 allUsedGenes=false(numel(allGenes{1}),1);
0380 
0381 <span class="keyword">if</span> ~isempty(preferredOrder) &amp;&amp; numel(models)&gt;1
0382     [usedGenes, ~]=find(finalMappings{1}); <span class="comment">%All that are used in the first model in preferredOrder</span>
0383     allUsedGenes(usedGenes)=true;
0384     <span class="keyword">for</span> i=2:numel(finalMappings)
0385         [usedGenes, ~]=find(finalMappings{i});
0386         usedGenes=unique(usedGenes);
0387         a=ismember(usedGenes,find(allUsedGenes));
0388         
0389         [~, genesToDelete]=find(finalMappings{i}(usedGenes(a),:)); <span class="comment">%IMPORTANT! IS it really correct to remove all genes that map?</span>
0390         genesToDelete=unique(genesToDelete); <span class="comment">%Maybe not needed, but for clarity if nothing else</span>
0391         
0392         <span class="comment">%Remove all the genes that were already found and add the other</span>
0393         <span class="comment">%ones to allUsedGenes</span>
0394         models{useOrderIndexes(i)}=<a href="removeGenes.html" class="code" title="function reducedModel = removeGenes(model,genesToRemove,removeUnusedMets,removeBlockedRxns,standardizeRules)">removeGenes</a>(models{useOrderIndexes(i)},allGenes{i+1}(genesToDelete),true,true,false);
0395         allUsedGenes(usedGenes)=true;
0396         
0397         <span class="comment">%Remove the deleted genes from finalMappings and allGenes.</span>
0398         finalMappings{i}(:,genesToDelete)=[];
0399         allGenes{i+1}(genesToDelete)=[];
0400     <span class="keyword">end</span>
0401 <span class="keyword">end</span>
0402 
0403 <span class="comment">%Now loop through the models and update the gene associations. Genes not</span>
0404 <span class="comment">%belonging to the new organism will be renamed as 'OLD_MODELID_gene'</span>
0405 <span class="keyword">for</span> i=1:numel(models)
0406     <span class="comment">%Find all the new genes that should be used for this model</span>
0407     [newGenes, oldGenes]=find(finalMappings{i});
0408     
0409     <span class="comment">%Create a new gene list with the genes from the new organism and those</span>
0410     <span class="comment">%genes that could not be removed</span>
0411     replaceableGenes=allGenes{i+1}(unique(oldGenes));
0412     nonReplaceableGenes=setdiff(models{useOrderIndexes(i)}.genes,replaceableGenes);
0413     fullGeneList=[allGenes{1}(unique(newGenes));nonReplaceableGenes];
0414     
0415     <span class="comment">%Just to save some indexing later. This is the LAST index of</span>
0416     <span class="comment">%replaceable ones</span>
0417     nonRepStartIndex=numel(unique(newGenes));
0418     
0419     <span class="comment">%Construct a new rxnGeneMat</span>
0420     newRxnGeneMat=sparse(numel(models{useOrderIndexes(i)}.rxns),numel(fullGeneList));
0421     
0422     <span class="comment">%Now update the rxnGeneMat. This is a little tricky and could</span>
0423     <span class="comment">%probably be done in a more efficient way, but I just loop through the</span>
0424     <span class="comment">%reactions and add them one by one</span>
0425     <span class="keyword">for</span> j=1:numel(models{useOrderIndexes(i)}.rxns)
0426         <span class="comment">%Get the old genes encoding for this reaction</span>
0427         [~, oldGeneIds]=find(models{useOrderIndexes(i)}.rxnGeneMat(j,:));
0428         
0429         <span class="comment">%Update the matrix for each gene. This includes replacing one gene</span>
0430         <span class="comment">%with several new ones if there were several matches</span>
0431         <span class="keyword">for</span> k=1:numel(oldGeneIds)
0432             <span class="comment">%Match the gene to one in the gene list. This is done as a text</span>
0433             <span class="comment">%match. Could probably be done better, but I'm a little lost in</span>
0434             <span class="comment">%the indexing</span>
0435             
0436             geneName=models{useOrderIndexes(i)}.genes(oldGeneIds(k));
0437             
0438             <span class="comment">%First search in the mappable genes</span>
0439             mapIndex=find(ismember(allGenes{i+1},geneName));
0440             
0441             <span class="keyword">if</span> ~isempty(mapIndex)
0442                 <span class="comment">% add the old genes</span>
0443                 hitGenes.oldGenes = [hitGenes.oldGenes, {geneName}];
0444                 
0445                 <span class="comment">%Get the new genes for that gene</span>
0446                 a=find(finalMappings{i}(:,mapIndex));
0447                 
0448                 <span class="comment">%Find the positions of these genes in the final gene list</span>
0449                 [~, b]=ismember(allGenes{1}(a),fullGeneList);
0450                 
0451                 <span class="comment">%Update the matrix</span>
0452                 newRxnGeneMat(j,b)=1;
0453                 
0454                 <span class="comment">%Update the grRules string. This is tricky, but I hope that</span>
0455                 <span class="comment">%it's ok to replace the old gene name with the new one and</span>
0456                 <span class="comment">%add ') or (' if there were several matches. Be sure of</span>
0457                 <span class="comment">%this!</span>
0458                 repString=fullGeneList{b(1)};
0459                 <span class="keyword">for</span> l=2:numel(b)
0460                     repString=[repString <span class="string">') or ('</span> fullGeneList{b(l)}];
0461                 <span class="keyword">end</span>
0462                 
0463                 <span class="comment">% add the new matched genes</span>
0464                 hitGenes.newGenes = [hitGenes.newGenes, {repString}];
0465                 
0466                 <span class="comment">%Use regexprep instead of strrep to prevent partial matches</span>
0467                 models{useOrderIndexes(i)}.grRules{j}=regexprep(models{useOrderIndexes(i)}.grRules{j},[<span class="string">'(^|\s|\()'</span> geneName{1} <span class="string">'($|\s|\))'</span>],[<span class="string">'$1'</span> repString <span class="string">'$2'</span>]);
0468             <span class="keyword">else</span>
0469                 <span class="comment">%Then search in the non-replaceable genes. There could only</span>
0470                 <span class="comment">%be one match here</span>
0471                 index=find(ismember(nonReplaceableGenes,geneName));
0472                 
0473                 <span class="comment">%Update the matrix</span>
0474                 newRxnGeneMat(j,nonRepStartIndex+index)=1;
0475                 
0476                 models{useOrderIndexes(i)}.grRules{j}=strrep(models{useOrderIndexes(i)}.grRules{j},geneName{1},strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,geneName{1}));
0477             <span class="keyword">end</span>
0478         <span class="keyword">end</span>
0479     <span class="keyword">end</span>
0480     
0481     <span class="comment">%Add the new list of genes</span>
0482     models{useOrderIndexes(i)}.rxnGeneMat=newRxnGeneMat;
0483     <span class="keyword">if</span> ~isempty(nonReplaceableGenes)
0484         models{useOrderIndexes(i)}.genes=[allGenes{1}(unique(newGenes));strcat(<span class="string">'OLD_'</span>,models{useOrderIndexes(i)}.id,<span class="string">'_'</span>,nonReplaceableGenes)];
0485     <span class="keyword">else</span>
0486         models{useOrderIndexes(i)}.genes=allGenes{1}(unique(newGenes));
0487     <span class="keyword">end</span>
0488     <span class="keyword">if</span> isfield(models{useOrderIndexes(i)},<span class="string">'geneComps'</span>)
0489         geneComps=models{useOrderIndexes(i)}.geneComps(1);
0490         models{useOrderIndexes(i)}.geneComps=zeros(numel(models{useOrderIndexes(i)}.genes),1);
0491         <span class="comment">%Assume that all genes are in the same compartment, and this</span>
0492         <span class="comment">%compartment is specified for the first gene</span>
0493         models{useOrderIndexes(i)}.geneComps(:)=geneComps;
0494     <span class="keyword">end</span>
0495 <span class="keyword">end</span>
0496 
0497 <span class="comment">%Now merge the models. All information should be correct except for 'or'</span>
0498 <span class="comment">%complexes</span>
0499 draftModel=<a href="mergeModels.html" class="code" title="function model=mergeModels(models,metParam,supressWarnings)">mergeModels</a>(models,<span class="string">'metNames'</span>);
0500 
0501 <span class="comment">%Change description of the resulting model</span>
0502 draftModel.id=getModelFor;
0503 description=<span class="string">'Generated by getModelFromHomology using '</span>;
0504 <span class="keyword">for</span> i=1:numel(models)
0505     <span class="keyword">if</span> i&lt;numel(models)
0506         description=[description models{i}.id <span class="string">', '</span>];
0507     <span class="keyword">else</span>
0508         description=[description models{i}.id];
0509     <span class="keyword">end</span>
0510 <span class="keyword">end</span>
0511 draftModel.description=description;
0512 draftModel.rxnNotes=cell(length(draftModel.rxns),1);
0513 draftModel.rxnNotes(:)={<span class="string">'Included by getModelFromHomology'</span>};
0514 draftModel.rxnConfidenceScores=NaN(length(draftModel.rxns),1);
0515 draftModel.rxnConfidenceScores(:)=2;
0516 <span class="comment">%Standardize grRules and notify if problematic grRules are found</span>
0517 [draftModel.grRules,draftModel.rxnGeneMat]=<a href="standardizeGrRules.html" class="code" title="function [grRules,rxnGeneMat,indexes2check] = standardizeGrRules(model,embedded)">standardizeGrRules</a>(draftModel,false);
0518 <span class="keyword">end</span></pre></div>
<hr><address>Generated by <strong><a href="http://www.artefact.tk/software/matlab/m2html/" title="Matlab Documentation in HTML">m2html</a></strong> &copy; 2005</address>
</body>
</html>